<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Unlocking single cell spatial omics analyses with scdney • ScdneySpatial</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney">
<meta property="og:description" content="ScdneySpatial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ScdneySpatial</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/HOWTO_BUILD_WORKSHOP.html">How To Build A Workshop Package</a>
    </li>
    <li>
      <a href="../articles/workshop_example.html">An Example Workshop</a>
    </li>
    <li>
      <a href="../articles/workshop_material.html">Unlocking single cell spatial omics analyses with scdney</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/ScdneySpatial" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="workshop_material_files/htmlwidgets-1.6.3/htmlwidgets.js"></script><script src="workshop_material_files/plotly-binding-4.10.3/plotly.js"></script><script src="workshop_material_files/typedarray-0.1/typedarray.min.js"></script><link href="workshop_material_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="workshop_material_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><link href="workshop_material_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="workshop_material_files/plotly-main-2.11.1/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Unlocking single cell spatial omics analyses
with scdney</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/ScdneySpatial/blob/HEAD/vignettes/workshop_material.Rmd" class="external-link"><code>vignettes/workshop_material.Rmd</code></a></small>
      <div class="hidden name"><code>workshop_material.Rmd</code></div>

    </div>

    
    
<style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>
<p>Farhan Ameen<span class="math inline">\(^{1,2,3}\)</span>, Shila
Ghazanfar<span class="math inline">\(^{2,3}\)</span>, Ellis Patrick<span class="math inline">\(^{1,2,3}\)</span>.</p>
<p><span class="math inline">\(^1\)</span> Westmead Institute for
Medical Research, University of Sydney, Australia<br><span class="math inline">\(^2\)</span> Sydney Precision Data Science
Centre, University of Sydney, Australia<br><span class="math inline">\(^3\)</span> School of Mathematics and
Statistics, University of Sydney, Australia</p>
<p><br> Contact: ellis.patrick@sydney.edu.au</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>Understanding the interplay between different types of cells and
their immediate environment is critical for understanding the mechanisms
of cells themselves and their function in the context of human diseases.
Recent advances in high dimensional in situ cytometry technologies have
fundamentally revolutionized our ability to observe these complex
cellular relationships providing an unprecedented characterisation of
cellular heterogeneity in a tissue environment.</p>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>In this tutorial we will introduce an analytical framework for
analysing data from high dimensional spatial omics technologies such as,
CODEX, CycIF, IMC and High Definition Spatial Transcriptomics. This
framework makes use of functionality from our Bioconductor packages
simpleSeg, FuseSOM, scClassify, scHot, spicyR, listClust, statial,
scFeatures and ClassifyR. By the end of this tutorial attendees will be
able to implement and assess some of the key steps of a spatial analysis
pipeline including cell segmentation, feature normalisation, cell type
identification, microenvironment and cell-state characterisation,
spatial hypothesis testing and patient classification. Understanding
these key steps will provide attendees with the core skills needed to
interrogate the comprehensive spatial information generated by these
exciting new technologies.</p>
</div>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<p>It is expected that students will have:</p>
<ul>
<li>basic knowledge of R syntax,</li>
<li>familiarity with SingleCellExperiment and/or SpatialExperiment
objects, and</li>
<li>this workshop will not provide an in-depth description of
cell-resolution spatial omics technologies.</li>
</ul>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<p>Several single cell R packages will be used from the scdney package,
for more information visit: <a href="https://sydneybiox.github.io/scdney/" class="external-link uri">https://sydneybiox.github.io/scdney/</a></p>
</div>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Data visualisation &amp; cell segmentation</td>
<td>1h 30m</td>
</tr>
<tr class="even">
<td>Cell type clustering &amp; classification</td>
<td>1h 30m</td>
</tr>
<tr class="odd">
<td>Spatial analysis &amp; feature generation</td>
<td>2h</td>
</tr>
<tr class="even">
<td>Patient Classification</td>
<td>1h 30m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>Understand and visualise spatial omics datasets.</li>
<li>Identify key biological questions that can be addressed with these
technologies and spatial analysis.</li>
<li>Understand the key analytical steps involved in spatial omics
analysis, and perform these steps using R.</li>
<li>Evaluate the performance of data normalisation and cell
segmentation.</li>
<li>Understand and generate individual feature representations from
spatial omics data.</li>
<li>Develop appreciation on how to assess performance of classification
models.</li>
<li>Perform disease outcome prediction using the feature representation
and robust classification framework.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"simpleSeg"</span>, <span class="st">"cytomapper"</span>, <span class="st">"scClassify"</span>, <span class="st">"scHot"</span>, <span class="st">"FuseSOM"</span>, <span class="st">"spicyR"</span>, <span class="st">"lisaClust"</span>,<span class="st">"Statial"</span>, <span class="st">"scFeatures"</span>, <span class="st">"ClassifyR"</span>, <span class="st">"tidyverse"</span>, <span class="st">"scater"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"STexampleData"</span>, <span class="st">"SpatialDatasets"</span>, <span class="st">"tidySingleCellExperiment"</span>, <span class="st">"scuttle"</span>, <span class="st">"batchelor"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#library(ScdneySpatial)</span></span>
<span></span>
<span><span class="co"># packages from scdney</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/simpleSeg/" class="external-link">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scClassify</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scHOT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/" class="external-link">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/" class="external-link">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial" class="external-link">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scFeatures</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/SpatialDatasets" class="external-link">SpatialDatasets</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Other required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lmweber/STexampleData" class="external-link">STexampleData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/BumpyMatrix" class="external-link">BumpyMatrix</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">1</span>  <span class="co"># Feel free to parallelise things if you have the cores to spare.</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/generateBPParam.html" class="external-link">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"celltype_colours.R"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_SingleCellExperiment_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-data">The data<a class="anchor" aria-label="anchor" href="#the-data"></a>
</h3>
<p>We will use two motivating datasets:</p>
<ul>
<li>
<a href="https://www.cell.com/fulltext/S0092-8674(18)31100-0" class="external-link">Keren
et al, 2018</a>: A multiplexed ion beam imaging by time-of-flight
(MIBI-TOF) dataset profilining tissue from triple-negative breast cancer
patients.</li>
<li>
<a href="https://www.nature.com/articles/s41587-021-01006-2" class="external-link">Lohoff
et al, 2022</a>: A seqFISH study of early mouse organogenesis. We will
use a subset of data that is made available from the STExampleData
package.</li>
</ul>
</div>
<div class="section level3">
<h3 id="data-visualisation-and-exploration">Data visualisation and exploration<a class="anchor" aria-label="anchor" href="#data-visualisation-and-exploration"></a>
</h3>
<p>Here we will download the datasets, examine the structure, visualise
the data and perform some exploratory analyses.</p>
<div class="section level4">
<h4 id="seqfish-mouse-embryo">SeqFISH mouse embryo<a class="anchor" aria-label="anchor" href="#seqfish-mouse-embryo"></a>
</h4>
<p>Here we download the seqFISH mouse embryo data. This is a
<code>SpatialExperiment</code> object, which extends the
<code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">STexampleData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/STexampleData/man/STexampleData.html" class="external-link">seqFISH_mouseEmbryo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 351 11026 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts molecules</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(11026): embryo1_Pos0_cell10_z2 embryo1_Pos0_cell100_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos28_cell97_z2 embryo1_Pos28_cell98_z2</span></span>
<span><span class="co">#&gt; colData names(14): cell_id embryo ... segmentation_vertices sample_id</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span></code></pre>
<p>We can use functions designed for <code>SingleCellExperiment</code>
objects in the <code>scater</code> package for plotting via the
<code>reducedDim</code> slot. We multiply the spatial coordinates by a
matrix to flip the y-axis and ensure we fix the aspect ratio.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span><span class="va">coord_transform</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">coord_transform</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sox2"</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/visualiseSeqFISH-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How many cells are in this data?</li>
<li>How many genes?</li>
<li>Plot gene expression mapping point size to the cell area.</li>
</ol>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># try to answer the above question using the spe object. </span></span>
<span><span class="co"># you may want to check the SingleCellExperiment vignette.</span></span>
<span><span class="co"># https://bioconductor.org/packages/3.17/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html</span></span></code></pre></div>
<p>We can perform a typical gene-expression based analysis for this
data. Later in part two we will perform some specific analytical
techniques, but for now let’s explore the dataset and use methods
designed for single cell data.</p>
<p>Dimensionality reduction using PCA, batch correction across tiles
using the <code>batchelor</code> package, followed by UMAP and
plotting.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b.out</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/batchCorrect.html" class="external-link">batchCorrect</a></span><span class="op">(</span><span class="va">spe</span>, batch <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">pos</span>, assay.type <span class="op">=</span> <span class="st">"logcounts"</span>, PARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">FastMnnParam</a></span><span class="op">(</span>d<span class="op">=</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"FastMnn"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">b.out</span>, <span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"FastMnn"</span><span class="op">)</span></span>
<span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 351 11026 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts molecules logcounts</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(11026): embryo1_Pos0_cell10_z2 embryo1_Pos0_cell100_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos28_cell97_z2 embryo1_Pos28_cell98_z2</span></span>
<span><span class="co">#&gt; colData names(15): cell_id embryo ... sample_id sizeFactor</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_celltype_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"UMAP"</span>, colour_by <span class="op">=</span> <span class="st">"celltype_mapped_refined"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">celltype_colours</span><span class="op">)</span></span>
<span><span class="va">g_celltype_umap</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/seqFISHEDA-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"UMAP"</span>, colour_by <span class="op">=</span> <span class="st">"Sox2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/seqFISHEDA-2.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_celltype_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"celltype_mapped_refined"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">celltype_colours</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_all</span> <span class="op">&lt;-</span> <span class="va">g_celltype_spatial</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> <span class="va">g_celltype_umap</span></span>
<span><span class="va">g_all</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/seqFISHEDA-3.png" width="700"></p>
<div class="question">
<p><strong>Advanced/Extension Question</strong></p>
<ol style="list-style-type: decimal">
<li>What considerations need to be made for batch correction of spatial
data? What assumptions are being made and/or broken? How could you check
this?</li>
<li>Check out the <a href="https://davidgohel.github.io/ggiraph/index.html" class="external-link"><code>ggiraph</code></a>
package for extending the <code>g_all</code> object to an interactive
plot with a tooltip that links the spatial and UMAP coordinate systems.
(Hint: This may involve generating a new ggplot object outside of the
<code>plotReducedDim</code> function.)</li>
</ol>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># try to examine answer the above questions using the spe object. </span></span>
<span><span class="co"># you may want to set up some small simulation..</span></span></code></pre></div>
<p>At this point we will pause our examination of the seqFISH dataset
that is in the object <code>spe</code>, and turn over to the second
example dataset. In the second part we will revisit this data for
performing <code>scHOT</code> testing.</p>
</div>
<div class="section level4">
<h4 id="mibi-tof-breast-cancer">MIBI-TOF breast cancer<a class="anchor" aria-label="anchor" href="#mibi-tof-breast-cancer"></a>
</h4>
<p>Here we explore the MIBI-TOF breast cancer data, this is also a
<code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html" class="external-link">spe_Keren_2018</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tumour_type</span> <span class="op">!=</span> <span class="st">"cold"</span>,</span>
<span>         <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">`Survival_days_capped.`</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>event <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">Censored</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">kerenSPE</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 48 170171 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): intensities</span></span>
<span><span class="co">#&gt; rownames(48): Na Si ... Ta Au</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(170171): 1 2 ... 197677 197678</span></span>
<span><span class="co">#&gt; colData names(41): CellID imageID ... sample_id event</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How many cells are in this data?</li>
<li>How many markers? How many images?</li>
</ol>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># try to answer the above question using the imc object. </span></span>
<span><span class="co"># you may want to check the SingleCellExperiment vignette.</span></span>
<span><span class="co"># https://www.bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html</span></span></code></pre></div>
<p>Use the following code to run a UMAP <code>intensities</code>
assay.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Visualise the UMAP using the <code>plotReducedDim</code> function
and colour the UMAP by <code>cellType</code>. What does this UMAP tell
us?</li>
<li>What are some observations we could make if we coloured by
<code>imageID</code>?</li>
</ol>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">kerenSPE</span>, exprs_values <span class="op">=</span> <span class="st">"intensities"</span>, name <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># try to answer the question here!</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="cell-segmentation-evaluation">Cell segmentation &amp; evaluation<a class="anchor" aria-label="anchor" href="#cell-segmentation-evaluation"></a>
</h3>
<div class="section level4">
<h4 id="reading-images">Reading images<a class="anchor" aria-label="anchor" href="#reading-images"></a>
</h4>
<p>To load in our images we use the <code>loadImages</code> function
from <code>cytomapper</code>, here we use the patient 5 image from Keren
et al. as an example.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">image5</span> <span class="op">=</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kerenPatient5.tiff"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">image5</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"imageID"</span> <span class="op">=</span> <span class="st">"kerenPatient5"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Setting the channel names according to orginal paper. </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html" class="external-link">channelNames</a></span><span class="op">(</span><span class="va">image5</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Au"</span>, <span class="st">"Background"</span>, <span class="st">"Beta catenin"</span>, <span class="st">"Ca"</span>, <span class="st">"CD11b"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD138"</span>, <span class="st">"CD16"</span>, <span class="st">"CD20"</span>, <span class="st">"CD209"</span>, <span class="st">"CD3"</span>, <span class="st">"CD31"</span>, <span class="st">"CD4"</span>, <span class="st">"CD45"</span>, <span class="st">"CD45RO"</span>, <span class="st">"CD56"</span>, <span class="st">"CD63"</span>, <span class="st">"CD68"</span>, <span class="st">"CD8"</span>, <span class="st">"dsDNA"</span>, <span class="st">"EGFR"</span>, <span class="st">"Fe"</span>, <span class="st">"FoxP3"</span>, <span class="st">"H3K27me3"</span>, <span class="st">"H3K9ac"</span>, <span class="st">"HLA_Class_1"</span>, <span class="st">"HLA_DR"</span>, <span class="st">"IDO"</span>, <span class="st">"Keratin17"</span>, <span class="st">"Keratin6"</span>, <span class="st">"Ki67"</span>, <span class="st">"Lag3"</span>, <span class="st">"MPO"</span>, <span class="st">"Na"</span>, <span class="st">"P"</span>, <span class="st">"p53"</span>, <span class="st">"Pan-Keratin"</span>, <span class="st">"PD-L1"</span>, <span class="st">"PD1"</span>, <span class="st">"phospho-S6"</span>, <span class="st">"Si"</span>, <span class="st">"SMA"</span>, <span class="st">"Ta"</span>, <span class="st">"Vimentin"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="how-do-i-perform-segmentation-in-r">How do I perform segmentation in R?<a class="anchor" aria-label="anchor" href="#how-do-i-perform-segmentation-in-r"></a>
</h4>
<p>Images stored in a <code>list</code> or <code>CytoImageList</code>
can be segmented using <code>simpleSeg</code>. Below
<code>simpleSeg</code> will identify the nuclei in the image using the
<code>dsDNA</code>, <code>H3K27me3</code> and <code>H3K9ac</code>
channel. To estimate the cell body of the cells we will simply dilate
out from the nuclei by 3 pixels. We also have specified that the
channels be sqrt transformed, and the 99th quantile of values removed to
ensure our segmentation is not affected by outliers.</p>
<p>We can visualise the segmentations using the <code>display</code> and
<code>colorLabels</code> functions in <code>EBImage</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate segmentation masks</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg</a></span><span class="op">(</span></span>
<span>  <span class="va">image5</span>,</span>
<span>  nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dsDNA"</span>, <span class="st">"H3K27me3"</span>, <span class="st">"H3K9ac"</span><span class="op">)</span>,</span>
<span>  cellBody <span class="op">=</span> <span class="st">"dilate"</span>, </span>
<span>  transform <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sqrt"</span>, <span class="st">"norm99"</span><span class="op">)</span>,</span>
<span>  sizeSelection <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  smooth <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  discSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Visualise segmentation performance one way.</span></span>
<span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html" class="external-link">colorLabels</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/simpleSeg-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="is-this-segmentation-appropriate">Is this segmentation appropriate?<a class="anchor" aria-label="anchor" href="#is-this-segmentation-appropriate"></a>
</h4>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> makes
it easy to overlay the masks on top of the intensities of markers in the
image. Here we can s</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance another way.</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image <span class="op">=</span> <span class="va">image5</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45"</span>, <span class="st">"Pan-Keratin"</span>, <span class="st">"SMA"</span>, <span class="st">"dsDNA"</span><span class="op">)</span>,</span>
<span>  display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="co"># Adjust the brightness, contrast and gamma of each channel.</span></span>
<span>  bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.8</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/plotSegmentation-1.png" width="700"></p>
<p>Using the <code>table</code> function in R we can measure the pixel
area of each cell and plot this on a histogram. In this dataset 1 µm is
equivalent to <span class="math inline">\(2.56\)</span> pixels, we can
transform the pixel area of a cell to physical area by dividing the
pixel area by <span class="math inline">\(2.56^2\)</span>. We can see
below the majority of our cells are just below <span class="math inline">\(100µm^2\)</span>, you should decide whether or not
this is an appropriate size based on your knowledge of the cell types
being imaged.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get area of every cell</span></span>
<span><span class="va">cellSize</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fl">2.56</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter out first index, which represents background</span></span>
<span><span class="va">cellSize</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cellSize</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">30</span>,</span>
<span>       xlab <span class="op">=</span> <span class="st">"Pixel area of cell (µm^2)"</span>,</span>
<span>       main <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> </span></code></pre></div>
<p><img src="workshop_material_files/figure-html/pixelArea-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Do our segmentations look better if other parameters are used?
<em>Hint</em>: use the help function to see what other parameters are
available in <code>simpleSeg</code>.</li>
</ol>
</div>
</div>
<div class="section level4">
<h4 id="how-to-i-convert-my-image-and-mask-to-an-object-i-can-use-in-r">How to I convert my image and mask to an object I can use in R?<a class="anchor" aria-label="anchor" href="#how-to-i-convert-my-image-and-mask-to-an-object-i-can-use-in-r"></a>
</h4>
<p>With our segmentation <code>masks</code> we can convert our image to
a <code>SpatialExperiment</code> object using the
<code>measureObjects</code> function in <code>cytomapper</code>.
<code>measureObjects</code> will find the center of each cell and
represent the cell as an x and y coordinate. In addition the average
marker expression within each cell is summarised and stored in the
assays object of the <code>SpatialExperiment</code>. <em>NOTE:</em> If
you are using other segmentation tools or softwares, you can import the
masks as tiffs into R and use <code>measureObjects</code> to create a
<code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarise the expression of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects</a></span><span class="op">(</span></span>
<span>  mask <span class="op">=</span> <span class="va">masks</span>,</span>
<span>  image <span class="op">=</span> <span class="va">image5</span>,</span>
<span>  feature_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"basic"</span>, <span class="st">"moment"</span><span class="op">)</span>,</span>
<span>  basic_feature <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>  moment_feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cx"</span>, <span class="st">"cy"</span><span class="op">)</span>,</span>
<span>  img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">cells</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 44 3739 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(44): Au Background ... Ta Vimentin</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(5): imageID object_id m.cx m.cy objectNum</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="feature-normalisation">Feature normalisation<a class="anchor" aria-label="anchor" href="#feature-normalisation"></a>
</h3>
<p>Before moving onto annotating the cell types, we should check if the
marker intensities of each cell require any transformation or
normalisation. Here we examine the distribution of CD45 in each cell
across all the images, which is a general marker for immune cells. Below
we can see that the intensity of CD45 looks very skewed.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Joining the marker information with the cell information</span></span>
<span><span class="va">proteinIntensities</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"intensities"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the density of CD45 across all images.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">proteinIntensities</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD45</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewingProteinDensity-1.png" width="700"></p>
<p>We can transform and normalise our data using the
<code>normalizeCells</code> function in <code>simpleSeg</code>. Here we
have taken the intensities from the <code>intensities</code> assay,
performed a asinh transform, then for each image trimmed the 99 quantile
and min-max scaled to 0-1. In addition to this we have performed
principle component analysis, and regressed out the first PC. This
modified data is then stored in the <code>normIntensities</code> assay.
We can see that the distribution of CD45 looks much less skewed than
before.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/normalizeCells.html" class="external-link">normalizeCells</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  transformation <span class="op">=</span> <span class="st">"asinh"</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"minMax"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>  assayIn <span class="op">=</span> <span class="st">"intensities"</span>,</span>
<span>  assayOut <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">normProteinIntensities</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"normIntensities"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">normProteinIntensities</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD45</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/kerenNormalise-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>CD3 is a marker which characterises T cells, plot the distribution
of CD3. What does it look like? What does this tell us about T cells in
our dataset.</li>
</ol>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Answer question here</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h3>
<div class="section level4">
<h4 id="cell-type-clustering">Cell type clustering<a class="anchor" aria-label="anchor" href="#cell-type-clustering"></a>
</h4>
<div class="section level5">
<h5 id="clustering-cell-types">Clustering cell types<a class="anchor" aria-label="anchor" href="#clustering-cell-types"></a>
</h5>
<p>Here we cluster using the <code>runFuseSOM</code> function from
<code>FuseSOM</code>. We have chosen to specify the same subset of
markers used in the original manuscript for gating cell types. We have
also specified the number of clusters to identify to be
<code>numClusters = 17</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">useMarkers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45"</span>, <span class="st">"FoxP3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"CD3"</span>, <span class="st">"CD20"</span>, <span class="st">"CD16"</span>, <span class="st">"CD68"</span>, <span class="st">"MPO"</span>, <span class="st">"HLA.DR"</span>, <span class="st">"Pan.Keratin"</span>, <span class="st">"Keratin17"</span>, <span class="st">"Keratin6"</span>, <span class="st">"p53"</span>, <span class="st">"Beta.catenin"</span>, <span class="st">"EGFR"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html" class="external-link">runFuseSOM</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  numClusters <span class="op">=</span> <span class="fl">17</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="cluster-interpretation">Cluster Interpretation<a class="anchor" aria-label="anchor" href="#cluster-interpretation"></a>
</h5>
<p>We can begin the process of understanding what each of these cell
clusters are by using the <code>plotGroupedHeatmap</code> function from
<code>scater</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  features <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/heatmap-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Have we captured distinct cell populations? Can you identify any of
the clusters?</li>
<li>Run the code below to compare the cell types in the dataset to the
clusters we have identified. How do our clusters compare? Are there any
clusters that should be combined?</li>
</ol>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Type your answer here</span></span>
<span></span>
<span></span>
<span><span class="co"># Question 2 code</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">kerenSPE</span>, cellType <span class="op">=</span> <span class="st">"clusters"</span>, region <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell types in data"</span>, y <span class="op">=</span> <span class="st">"Our clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/clusteringQ1-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="how-do-i-determine-an-appropriate-number-of-clusters">How do I determine an appropriate number of clusters?<a class="anchor" aria-label="anchor" href="#how-do-i-determine-an-appropriate-number-of-clusters"></a>
</h5>
<p>Great question! There is no right answer here, however there are
several statistics which can be used to estimate the “true” number of
clusters. We can use the <code>estimateNumCluster</code> and
<code>optiPlot</code> functions from <code>FuseSOM</code> to examine if
our choice of 17 clusters is reasonable. Here we use the gap statistic,
other methods such as jump, slope, silhouette and within cluster
distance (wcd) are also available.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html" class="external-link">estimateNumCluster</a></span><span class="op">(</span><span class="va">kerenSPE</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html" class="external-link">optiPlot</a></span><span class="op">(</span><span class="va">kerenSPE</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/optiPlot-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">clusterEstimation</span><span class="op">$</span><span class="va">Discriminant</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 10</span></span></code></pre>
<p>Finally we can run a UMAP to examine how distinct our clusters are
from one another.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMP.</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"normUMAP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># UMAP by cell type cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/normUMAP-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How does this UMAP compare to the original UMAP.</li>
</ol>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Answer question here.</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="cell-type-classification">Cell type classification<a class="anchor" aria-label="anchor" href="#cell-type-classification"></a>
</h4>
<p>Cell type clustering may difficult without the aid of domain experts
who know how many cell types to expect and are able to annotate the
clusters. Cell type classification is an alternative approach to
clustering which annotates cell types based on an expert labeled
reference dataset.</p>
<p><em>TO FILL</em> - Need a reference dataset with similar markers to
keren: [1] “Na” “Si” “P” “Ca” “Fe” “dsDNA”<br>
[7] “Vimentin” “SMA” “Background” “B7H3” “FoxP3” “Lag3”<br>
[13] “CD4” “CD16” “CD56” “OX40” “PD1” “CD31”<br>
[19] “PD.L1” “EGFR” “Ki67” “CD209” “CD11c” “CD138”<br>
[25] “CD163” “CD68” “CSF.1R” “CD8” “CD3” “IDO”<br>
[31] “Keratin17” “CD63” “CD45RO” “CD20” “p53” “Beta.catenin” [37]
“HLA.DR” “CD11b” “CD45” “H3K9ac” “Pan.Keratin” “H3K27me3”<br>
[43] “phospho.S6” “MPO” “Keratin6” “HLA_Class_1” “Ta” “Au”</p>
</div>
<div class="section level4">
<h4 id="viewing-images">Viewing images<a class="anchor" aria-label="anchor" href="#viewing-images"></a>
</h4>
<p>We can look at the distribution of cells in an image. Here we compare
our clusters to the cell types in the dataset.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Our clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-1.png" width="700"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Dataset clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-2.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="analytical-techniques">Analytical techniques<a class="anchor" aria-label="anchor" href="#analytical-techniques"></a>
</h3>
<div class="section level4">
<h4 id="schot-analysis-of-the-developing-brain">scHOT analysis of the developing brain<a class="anchor" aria-label="anchor" href="#schot-analysis-of-the-developing-brain"></a>
</h4>
<p>Here we will ask which gene patterns we observe to be changing across
the <code>spe$gutRegion</code> cell type in space. Note that we want to
assess the anatomical region corresponding to the anterior end of the
developing gut developing brain so we will first subset the cells using
the spatial coordinates. We can check what we have selected by
plotting.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter cells which are labeled gut tub, and in the anterior section.</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">gutRegion</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">$</span><span class="va">celltype_mapped_refined</span> <span class="op">==</span> <span class="st">"Gut tube"</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.5</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"gutRegion"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Let’s subset the data to only these cells and continue with our scHOT
analysis.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_gut</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>,<span class="va">spe</span><span class="op">$</span><span class="va">gutRegion</span><span class="op">]</span></span>
<span><span class="va">spe_gut</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 351 472 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts molecules logcounts</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(472): embryo1_Pos3_cell377_z2 embryo1_Pos3_cell388_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos27_cell74_z2 embryo1_Pos28_cell373_z2</span></span>
<span><span class="co">#&gt; colData names(16): cell_id embryo ... sizeFactor gutRegion</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre>
<p>We select genes with at least some proportion of expressed cells for
testing, and create the <code>scHOT</code> object.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_gut</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span>, <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_to_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe_gut</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_gut</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gene_to_test</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 165</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_to_test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">gene_to_test</span>, <span class="fl">1</span>, <span class="va">paste0</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_to_test</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;         [,1]     </span></span>
<span><span class="co">#&gt; Acvr1   "Acvr1"  </span></span>
<span><span class="co">#&gt; Acvr2a  "Acvr2a" </span></span>
<span><span class="co">#&gt; Ahnak   "Ahnak"  </span></span>
<span><span class="co">#&gt; Akr1c19 "Akr1c19"</span></span>
<span><span class="co">#&gt; Aldh1a2 "Aldh1a2"</span></span>
<span><span class="co">#&gt; Aldh2   "Aldh2"</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_buildFromSCE.html" class="external-link">scHOT_buildFromSCE</a></span><span class="op">(</span><span class="va">spe_gut</span>,</span>
<span>                                    assayName <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                                    positionType <span class="op">=</span> <span class="st">"spatial"</span>,</span>
<span>                                    positionColData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x_global_affine"</span>, <span class="st">"y_global_affine"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scHOT_spatial</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: scHOT </span></span>
<span><span class="co">#&gt; dim: 351 472 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): expression</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(472): embryo1_Pos3_cell377_z2 embryo1_Pos3_cell388_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos27_cell74_z2 embryo1_Pos28_cell373_z2</span></span>
<span><span class="co">#&gt; colData names(16): cell_id embryo ... sizeFactor gutRegion</span></span>
<span><span class="co">#&gt; testingScaffold dim: 0 0 </span></span>
<span><span class="co">#&gt; weightMatrix dim: 0 0 </span></span>
<span><span class="co">#&gt; scHOT_output colnames (0):</span></span>
<span><span class="co">#&gt; param names (0):</span></span>
<span><span class="co">#&gt; position type: spatial</span></span></code></pre>
<p>We now add the testing scaffold to the <code>scHOT</code> object, and
set the local weight matrix for testing, with a choice of span of 0.1
(the proportion of cells to weight around each cell). We can speed up
computation by not requiring the weight matrix correspond to every
individual cell, but instead a random selection among all the cells
using the <code>thin</code> function.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_addTestingScaffold.html" class="external-link">scHOT_addTestingScaffold</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="va">gene_to_test</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">testingScaffold</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;         gene_1   </span></span>
<span><span class="co">#&gt; Acvr1   "Acvr1"  </span></span>
<span><span class="co">#&gt; Acvr2a  "Acvr2a" </span></span>
<span><span class="co">#&gt; Ahnak   "Ahnak"  </span></span>
<span><span class="co">#&gt; Akr1c19 "Akr1c19"</span></span>
<span><span class="co">#&gt; Aldh1a2 "Aldh1a2"</span></span>
<span><span class="co">#&gt; Aldh2   "Aldh2"</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_setWeightMatrix.html" class="external-link">scHOT_setWeightMatrix</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, span <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">weightMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/thin.html" class="external-link">thin</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">weightMatrix</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"weightMatrix"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1]  53 472</span></span></code></pre>
<p>For a given cell we can visually examine the local weight given by
the span parameter.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellID</span> <span class="op">=</span> <span class="fl">10</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      W <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"weightMatrix"</span><span class="op">)</span><span class="op">[</span><span class="va">cellID</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x_global_affine</span>, y <span class="op">=</span> <span class="op">-</span><span class="va">y_global_affine</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">W</span>, size <span class="op">=</span> <span class="va">W</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"black"</span>, high <span class="op">=</span> <span class="st">"purple"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html" class="external-link">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">2.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spatial Weight"</span><span class="op">)</span>,</span>
<span>         size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spatial Weight"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Central cell: "</span>, <span class="va">cellID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="cn">NULL</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-5-1.png" width="700">
::: question <strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>How will the results change if the span is increased/decreased?</li>
</ol>
<p>:::</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Make associated changes to the code to test out the question above.</span></span></code></pre></div>
<p>We set the higher order function as the weighted mean function, and
then calculate the observed higher order test statistics. This may take
around 10 seconds.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_calculateGlobalHigherOrderFunction.html" class="external-link">scHOT_calculateGlobalHigherOrderFunction</a></span><span class="op">(</span></span>
<span>    <span class="va">scHOT_spatial</span>,</span>
<span>    higherOrderFunction <span class="op">=</span> <span class="va">weightedMean</span>,</span>
<span>    higherOrderFunctionType <span class="op">=</span> <span class="st">"weighted"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; DataFrame with 165 rows and 2 columns</span></span>
<span><span class="co">#&gt;              gene_1 globalHigherOrderFunction</span></span>
<span><span class="co">#&gt;         &lt;character&gt;                  &lt;matrix&gt;</span></span>
<span><span class="co">#&gt; Acvr1         Acvr1                  0.216666</span></span>
<span><span class="co">#&gt; Acvr2a       Acvr2a                  0.375776</span></span>
<span><span class="co">#&gt; Ahnak         Ahnak                  0.976418</span></span>
<span><span class="co">#&gt; Akr1c19     Akr1c19                  0.744070</span></span>
<span><span class="co">#&gt; Aldh1a2     Aldh1a2                  0.245981</span></span>
<span><span class="co">#&gt; ...             ...                       ...</span></span>
<span><span class="co">#&gt; Wnt5a         Wnt5a                  0.335820</span></span>
<span><span class="co">#&gt; Wnt5b         Wnt5b                  0.220300</span></span>
<span><span class="co">#&gt; Xist           Xist                  1.162241</span></span>
<span><span class="co">#&gt; Zfp444       Zfp444                  0.744082</span></span>
<span><span class="co">#&gt; Zfp57         Zfp57                  0.595519</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_calculateHigherOrderTestStatistics.html" class="external-link">scHOT_calculateHigherOrderTestStatistics</a></span><span class="op">(</span></span>
<span>    <span class="va">scHOT_spatial</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now we can plot the overall mean versus the scHOT statistic to
observe any relationship. Labels can be interactively visualised using
<code>ggplotly</code>. Some genes may have different distributions so we
turn to permutation testing to assess statistical significance.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">scHOT_output</span><span class="op">)</span>, </span>
<span>           <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">globalHigherOrderFunction</span>, y <span class="op">=</span> <span class="va">higherOrderStatistic</span>, label <span class="op">=</span> <span class="va">gene_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mean across all cells"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"scHOT statistic for local weightedMean"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html" class="external-link">ggplotly</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<div class="plotly html-widget html-fill-item" id="htmlwidget-d10f9e6383278f5ad4f8" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-d10f9e6383278f5ad4f8">{"x":{"data":[{"x":[0.21666552735611611,0.37577624073907062,0.97641826897498996,0.74407017030072786,0.24598113331792176,0.57621688385637004,0.23714015727416829,0.21311146138673961,0.57203612263584536,0.21501613520732421,0.51264320790253159,0.78364174184367774,0.61137793088200598,0.23908786096561196,0.57331002310046431,0.79441922973015022,0.29529448321734508,0.22554474469973529,1.3034118346505226,0.65843819340434984,0.36069675428384212,0.50167467078331296,1.6982331364217238,0.6169600807730351,0.31478264601638573,0.44608821671678767,0.34208291118085427,1.9602204845818796,0.63466372607287114,0.72354119106741821,0.48109400065677832,0.2782337364534731,0.71526027492151278,0.92291201508114784,0.6432469482317259,0.21184593402623622,0.82435111277727369,0.41624493839071014,0.60440035919341417,1.4422476150058909,1.075913614761882,0.20042244573595483,0.27028744249370973,0.37269680194193966,0.74477315065749117,0.7866580914135477,0.22107122298289217,1.2763002780546775,1.1635097126550837,0.4120981727395866,0.25732306656694154,2.0492375226753303,1.3820831684358559,0.29890041870737799,0.38018117956576419,0.20808104326229426,1.0954940816795466,1.9601907765655602,0.2699867563461546,0.66824020210897617,0.32522575230510742,0.58927391005327978,0.58682333818527499,0.29647986944199978,0.19235173147853118,0.50884635454788629,0.24722469804609015,0.76766056339513056,0.55574180080135738,0.24864530189996081,0.25388676965681539,0.62470373987410555,0.21718656392817837,0.3540037183500146,0.58266002463937372,0.7884118671369108,0.57216715914481908,0.55700349034140129,0.58223625859697292,1.6139119096047625,1.0926070950474727,0.25316502418075132,0.24402709287273733,1.2309838380385743,1.4927254753415129,3.1461735946697611,0.54584533525953693,0.58728140567460896,2.7900368714686992,1.121406291147095,0.67470063305624151,0.5950188928353165,1.2933786162089944,0.21372117903023019,0.39291396306236454,0.80984277898690182,0.25062482071371628,0.28616762990566519,3.1714201156286199,0.27637731585371667,0.55632613598057212,0.60108958405887536,0.32671818851712797,0.98607793273384281,0.42119090449842173,0.32815258332994046,1.1472873701564217,0.71181135115499838,0.21488565244537938,0.23273347444793882,0.85701652609813805,0.26772453783279287,0.30990857893632007,0.33575352309027001,0.20264373803268534,0.45420800833416286,0.23978661944540236,0.21479506920752414,2.4988430121508967,0.24600963608570298,0.41973395792336066,0.23305094778343419,0.41605637457964789,0.27926233014940444,0.4537272391422531,0.31457161989425569,0.22211894233092785,1.1842563007397477,1.0084242460794188,0.75690504423769356,0.3426839327056237,0.23998577481502317,1.6203129561432765,1.1630577214433306,0.43160490248648653,0.4237788706716783,0.28103994379488711,0.7860524133034128,0.36636472198663073,0.23885545135155606,1.414629844358563,2.1829706132926536,0.54676314117002078,0.47472551591541745,0.47808969273253948,0.58594082534803971,0.87386089471660489,0.49776455063325081,0.20957263415951197,1.6886508256146189,2.3593471668159869,0.32030042961685501,0.2611588156308331,1.2984821828822615,0.50716608578160127,0.29612081131710122,1.30846630033652,0.27344062006270492,0.31903210675918403,0.60523576725855821,0.33581994971876239,0.22030035859149219,1.1622405164885681,0.74408203349418633,0.59551893573281622],"y":[0.075095380214497021,0.066514261705854574,0.33198974116788166,0.16733424713635042,0.18278360718917858,0.047671537659420338,0.07186424187938692,0.072962447674642716,0.073181758790845611,0.075120169081686533,0.30177676403587872,0.1360275037983418,0.49699891953758263,0.056942698676209873,0.49364457413495111,0.41880335773158406,0.20077560847361331,0.11467506362167465,0.40275597064761215,0.35314778827094068,0.19625521212022629,0.29492911074680683,0.45514740322789654,0.4539109990827444,0.15886662220312905,0.31463682633120993,0.096375648802461453,0.28731276477496509,0.45716024335599253,0.37270448722092175,0.14208552431771437,0.19677325129049691,0.40613789875267448,0.25051524280204962,0.19710074006945463,0.049166249684574538,0.45953594375934403,0.092989995065445849,0.14184919566747706,0.092563539791341451,0.22640012377923538,0.071806531265627305,0.074721694907642425,0.067030775412973254,0.14743939925476779,0.1734723099137484,0.10109092167971297,0.18179746784504919,0.18523524547279363,0.1175773883248421,0.16478510133758065,0.67492003058760197,0.366516341294432,0.10155142974125538,0.19711678094904198,0.10250082750830299,0.18832169593096001,0.18886922151675242,0.19659512793646761,0.28064129449400821,0.23524277936887625,0.48264343713798979,0.57536942841077399,0.12822926457478565,0.08358813994269669,0.45639989147226695,0.053989331141883021,0.55655937606082417,0.4822046668653534,0.058089447246819546,0.13102669292250885,0.12022546604755678,0.10148233737483212,0.27844114183633184,0.34495831689211026,0.28451211012487748,0.29033668050777695,0.24716736575919387,0.25272480409698705,0.41528830442036102,0.26209008279250512,0.045907743179835855,0.09085191586359348,0.19727984475908225,0.16053793632899108,0.28946312010240488,0.17750889579707813,0.24099351762276219,0.36913043172168658,0.33793229874715225,0.13422303963379703,0.35986322791147862,0.25160004707783945,0.072702929914132353,0.066365839569215351,0.16501604172984119,0.11689242051874478,0.11761442755781423,0.31131567935608212,0.093249754746791064,0.65257377436243924,0.14392269285478268,0.13614020147214193,0.69828735622624694,0.49888120969141392,0.093675322304645522,0.16645054123515643,0.10964026189916989,0.083969838940518488,0.053648789168821397,0.27369883421291136,0.077623141162213583,0.12466191712904855,0.29624618414932713,0.073048534054927158,0.56965653768002134,0.054245878043803236,0.058404339597059228,0.40998662475795533,0.16397672470283373,0.22379415253303789,0.083410448705932005,0.17780335910001122,0.10058222373924668,0.099150618043280792,0.13081957061787233,0.18269142698974802,0.12340876027492466,0.16965359159267754,0.092757739866083044,0.14286592514491342,0.091742916407448818,0.67710108943687441,0.75174986496642204,0.34742425041077118,0.10223663225145126,0.10097882441763696,0.34652476752963296,0.076266557294053891,0.11384079137320724,0.39927390801112272,0.41994711407553248,0.42476491099519398,0.1063925332329429,0.56488051773448922,0.34040256111296369,0.46242377876376256,0.45279942151473102,0.083654718276944068,0.14457351633888255,0.12042511501795565,0.051474315762060699,0.062109339514151464,0.095456549955225886,0.26578096419295338,0.11300255828680771,0.26947465570342066,0.12077106508792358,0.067784020669587802,0.094066179765152511,0.17229957637974883,0.10492409670964814,0.12082808057197614,0.1189296002916481,0.13012771733103512],"text":["globalHigherOrderFunction: 0.2166655<br />higherOrderStatistic: 0.07509538<br />gene_1: Acvr1","globalHigherOrderFunction: 0.3757762<br />higherOrderStatistic: 0.06651426<br />gene_1: Acvr2a","globalHigherOrderFunction: 0.9764183<br />higherOrderStatistic: 0.33198974<br />gene_1: Ahnak","globalHigherOrderFunction: 0.7440702<br />higherOrderStatistic: 0.16733425<br />gene_1: Akr1c19","globalHigherOrderFunction: 0.2459811<br />higherOrderStatistic: 0.18278361<br />gene_1: Aldh1a2","globalHigherOrderFunction: 0.5762169<br />higherOrderStatistic: 0.04767154<br />gene_1: Aldh2","globalHigherOrderFunction: 0.2371402<br />higherOrderStatistic: 0.07186424<br />gene_1: Apln","globalHigherOrderFunction: 0.2131115<br />higherOrderStatistic: 0.07296245<br />gene_1: Apoa4","globalHigherOrderFunction: 0.5720361<br />higherOrderStatistic: 0.07318176<br />gene_1: Ash2l","globalHigherOrderFunction: 0.2150161<br />higherOrderStatistic: 0.07512017<br />gene_1: Atp1b1","globalHigherOrderFunction: 0.5126432<br />higherOrderStatistic: 0.30177676<br />gene_1: Axin2","globalHigherOrderFunction: 0.7836417<br />higherOrderStatistic: 0.13602750<br />gene_1: Bak1","globalHigherOrderFunction: 0.6113779<br />higherOrderStatistic: 0.49699892<br />gene_1: Bambi","globalHigherOrderFunction: 0.2390879<br />higherOrderStatistic: 0.05694270<br />gene_1: Bid","globalHigherOrderFunction: 0.5733100<br />higherOrderStatistic: 0.49364457<br />gene_1: Bmp4","globalHigherOrderFunction: 0.7944192<br />higherOrderStatistic: 0.41880336<br />gene_1: Bmp7","globalHigherOrderFunction: 0.2952945<br />higherOrderStatistic: 0.20077561<br />gene_1: Car7","globalHigherOrderFunction: 0.2255447<br />higherOrderStatistic: 0.11467506<br />gene_1: Cavin3","globalHigherOrderFunction: 1.3034118<br />higherOrderStatistic: 0.40275597<br />gene_1: Cdh1","globalHigherOrderFunction: 0.6584382<br />higherOrderStatistic: 0.35314779<br />gene_1: Cdh2","globalHigherOrderFunction: 0.3606968<br />higherOrderStatistic: 0.19625521<br />gene_1: Cers4","globalHigherOrderFunction: 0.5016747<br />higherOrderStatistic: 0.29492911<br />gene_1: Chrd","globalHigherOrderFunction: 1.6982331<br />higherOrderStatistic: 0.45514740<br />gene_1: Cldn4","globalHigherOrderFunction: 0.6169601<br />higherOrderStatistic: 0.45391100<br />gene_1: Clic6","globalHigherOrderFunction: 0.3147826<br />higherOrderStatistic: 0.15886662<br />gene_1: Cntfr","globalHigherOrderFunction: 0.4460882<br />higherOrderStatistic: 0.31463683<br />gene_1: Col1a2","globalHigherOrderFunction: 0.3420829<br />higherOrderStatistic: 0.09637565<br />gene_1: Col26a1","globalHigherOrderFunction: 1.9602205<br />higherOrderStatistic: 0.28731276<br />gene_1: Col4a1","globalHigherOrderFunction: 0.6346637<br />higherOrderStatistic: 0.45716024<br />gene_1: Cpm","globalHigherOrderFunction: 0.7235412<br />higherOrderStatistic: 0.37270449<br />gene_1: Cpn1","globalHigherOrderFunction: 0.4810940<br />higherOrderStatistic: 0.14208552<br />gene_1: Cxcl12","globalHigherOrderFunction: 0.2782337<br />higherOrderStatistic: 0.19677325<br />gene_1: Cyp26a1","globalHigherOrderFunction: 0.7152603<br />higherOrderStatistic: 0.40613790<br />gene_1: Dlk1","globalHigherOrderFunction: 0.9229120<br />higherOrderStatistic: 0.25051524<br />gene_1: Dnmt3a","globalHigherOrderFunction: 0.6432469<br />higherOrderStatistic: 0.19710074<br />gene_1: Dnmt3b","globalHigherOrderFunction: 0.2118459<br />higherOrderStatistic: 0.04916625<br />gene_1: Dppa2","globalHigherOrderFunction: 0.8243511<br />higherOrderStatistic: 0.45953594<br />gene_1: Dusp6","globalHigherOrderFunction: 0.4162449<br />higherOrderStatistic: 0.09299000<br />gene_1: Eed","globalHigherOrderFunction: 0.6044004<br />higherOrderStatistic: 0.14184920<br />gene_1: Efna5","globalHigherOrderFunction: 1.4422476<br />higherOrderStatistic: 0.09256354<br />gene_1: Ep300","globalHigherOrderFunction: 1.0759136<br />higherOrderStatistic: 0.22640012<br />gene_1: Epcam","globalHigherOrderFunction: 0.2004224<br />higherOrderStatistic: 0.07180653<br />gene_1: Esam","globalHigherOrderFunction: 0.2702874<br />higherOrderStatistic: 0.07472169<br />gene_1: Ets1","globalHigherOrderFunction: 0.3726968<br />higherOrderStatistic: 0.06703078<br />gene_1: Ezh1","globalHigherOrderFunction: 0.7447732<br />higherOrderStatistic: 0.14743940<br />gene_1: Ezh2","globalHigherOrderFunction: 0.7866581<br />higherOrderStatistic: 0.17347231<br />gene_1: F2r","globalHigherOrderFunction: 0.2210712<br />higherOrderStatistic: 0.10109092<br />gene_1: Fam181b","globalHigherOrderFunction: 1.2763003<br />higherOrderStatistic: 0.18179747<br />gene_1: Fgfr1","globalHigherOrderFunction: 1.1635097<br />higherOrderStatistic: 0.18523525<br />gene_1: Fgfr2","globalHigherOrderFunction: 0.4120982<br />higherOrderStatistic: 0.11757739<br />gene_1: Fgfr3","globalHigherOrderFunction: 0.2573231<br />higherOrderStatistic: 0.16478510<br />gene_1: Fgfr4","globalHigherOrderFunction: 2.0492375<br />higherOrderStatistic: 0.67492003<br />gene_1: Foxa1","globalHigherOrderFunction: 1.3820832<br />higherOrderStatistic: 0.36651634<br />gene_1: Foxa2","globalHigherOrderFunction: 0.2989004<br />higherOrderStatistic: 0.10155143<br />gene_1: Foxc2","globalHigherOrderFunction: 0.3801812<br />higherOrderStatistic: 0.19711678<br />gene_1: Foxf1","globalHigherOrderFunction: 0.2080810<br />higherOrderStatistic: 0.10250083<br />gene_1: Fst","globalHigherOrderFunction: 1.0954941<br />higherOrderStatistic: 0.18832170<br />gene_1: Furin","globalHigherOrderFunction: 1.9601908<br />higherOrderStatistic: 0.18886922<br />gene_1: Fzd2","globalHigherOrderFunction: 0.2699868<br />higherOrderStatistic: 0.19659513<br />gene_1: Gata2","globalHigherOrderFunction: 0.6682402<br />higherOrderStatistic: 0.28064129<br />gene_1: Gata3","globalHigherOrderFunction: 0.3252258<br />higherOrderStatistic: 0.23524278<br />gene_1: Gata4","globalHigherOrderFunction: 0.5892739<br />higherOrderStatistic: 0.48264344<br />gene_1: Gata5","globalHigherOrderFunction: 0.5868233<br />higherOrderStatistic: 0.57536943<br />gene_1: Gata6","globalHigherOrderFunction: 0.2964799<br />higherOrderStatistic: 0.12822926<br />gene_1: Gmpr","globalHigherOrderFunction: 0.1923517<br />higherOrderStatistic: 0.08358814<br />gene_1: Gng3","globalHigherOrderFunction: 0.5088464<br />higherOrderStatistic: 0.45639989<br />gene_1: Gpc4","globalHigherOrderFunction: 0.2472247<br />higherOrderStatistic: 0.05398933<br />gene_1: Hapln1","globalHigherOrderFunction: 0.7676606<br />higherOrderStatistic: 0.55655938<br />gene_1: Hoxa1","globalHigherOrderFunction: 0.5557418<br />higherOrderStatistic: 0.48220467<br />gene_1: Hoxb1","globalHigherOrderFunction: 0.2486453<br />higherOrderStatistic: 0.05808945<br />gene_1: Hoxb3","globalHigherOrderFunction: 0.2538868<br />higherOrderStatistic: 0.13102669<br />gene_1: Hoxb4","globalHigherOrderFunction: 0.6247037<br />higherOrderStatistic: 0.12022547<br />gene_1: Hsf1","globalHigherOrderFunction: 0.2171866<br />higherOrderStatistic: 0.10148234<br />gene_1: Icam2","globalHigherOrderFunction: 0.3540037<br />higherOrderStatistic: 0.27844114<br />gene_1: Igf1","globalHigherOrderFunction: 0.5826600<br />higherOrderStatistic: 0.34495832<br />gene_1: Igfbp3","globalHigherOrderFunction: 0.7884119<br />higherOrderStatistic: 0.28451211<br />gene_1: Irx1","globalHigherOrderFunction: 0.5721672<br />higherOrderStatistic: 0.29033668<br />gene_1: Irx2","globalHigherOrderFunction: 0.5570035<br />higherOrderStatistic: 0.24716737<br />gene_1: Irx3","globalHigherOrderFunction: 0.5822363<br />higherOrderStatistic: 0.25272480<br />gene_1: Isl1","globalHigherOrderFunction: 1.6139119<br />higherOrderStatistic: 0.41528830<br />gene_1: Itga3","globalHigherOrderFunction: 1.0926071<br />higherOrderStatistic: 0.26209008<br />gene_1: Jarid2","globalHigherOrderFunction: 0.2531650<br />higherOrderStatistic: 0.04590774<br />gene_1: Kcng1","globalHigherOrderFunction: 0.2440271<br />higherOrderStatistic: 0.09085192<br />gene_1: Kitl","globalHigherOrderFunction: 1.2309838<br />higherOrderStatistic: 0.19727984<br />gene_1: Kmt2b","globalHigherOrderFunction: 1.4927255<br />higherOrderStatistic: 0.16053794<br />gene_1: Kmt2d","globalHigherOrderFunction: 3.1461736<br />higherOrderStatistic: 0.28946312<br />gene_1: Krt18","globalHigherOrderFunction: 0.5458453<br />higherOrderStatistic: 0.17750890<br />gene_1: Ldhb","globalHigherOrderFunction: 0.5872814<br />higherOrderStatistic: 0.24099352<br />gene_1: Lef1","globalHigherOrderFunction: 2.7900369<br />higherOrderStatistic: 0.36913043<br />gene_1: Lin28a","globalHigherOrderFunction: 1.1214063<br />higherOrderStatistic: 0.33793230<br />gene_1: Marcks","globalHigherOrderFunction: 0.6747006<br />higherOrderStatistic: 0.13422304<br />gene_1: Mcl1","globalHigherOrderFunction: 0.5950189<br />higherOrderStatistic: 0.35986323<br />gene_1: Meis1","globalHigherOrderFunction: 1.2933786<br />higherOrderStatistic: 0.25160005<br />gene_1: Meis2","globalHigherOrderFunction: 0.2137212<br />higherOrderStatistic: 0.07270293<br />gene_1: Meox1","globalHigherOrderFunction: 0.3929140<br />higherOrderStatistic: 0.06636584<br />gene_1: Mkrn1","globalHigherOrderFunction: 0.8098428<br />higherOrderStatistic: 0.16501604<br />gene_1: Mnt","globalHigherOrderFunction: 0.2506248<br />higherOrderStatistic: 0.11689242<br />gene_1: Msx1","globalHigherOrderFunction: 0.2861676<br />higherOrderStatistic: 0.11761443<br />gene_1: Msx2","globalHigherOrderFunction: 3.1714201<br />higherOrderStatistic: 0.31131568<br />gene_1: Myh9","globalHigherOrderFunction: 0.2763773<br />higherOrderStatistic: 0.09324975<br />gene_1: Nid1","globalHigherOrderFunction: 0.5563261<br />higherOrderStatistic: 0.65257377<br />gene_1: Nkx2-3","globalHigherOrderFunction: 0.6010896<br />higherOrderStatistic: 0.14392269<br />gene_1: Nr2f1","globalHigherOrderFunction: 0.3267182<br />higherOrderStatistic: 0.13614020<br />gene_1: Nrp1","globalHigherOrderFunction: 0.9860779<br />higherOrderStatistic: 0.69828736<br />gene_1: Osr1","globalHigherOrderFunction: 0.4211909<br />higherOrderStatistic: 0.49888121<br />gene_1: Otx2","globalHigherOrderFunction: 0.3281526<br />higherOrderStatistic: 0.09367532<br />gene_1: Ovol2","globalHigherOrderFunction: 1.1472874<br />higherOrderStatistic: 0.16645054<br />gene_1: Pcgf2","globalHigherOrderFunction: 0.7118114<br />higherOrderStatistic: 0.10964026<br />gene_1: Pcgf3","globalHigherOrderFunction: 0.2148857<br />higherOrderStatistic: 0.08396984<br />gene_1: Pcgf5","globalHigherOrderFunction: 0.2327335<br />higherOrderStatistic: 0.05364879<br />gene_1: Pcgf6","globalHigherOrderFunction: 0.8570165<br />higherOrderStatistic: 0.27369883<br />gene_1: Pdgfa","globalHigherOrderFunction: 0.2677245<br />higherOrderStatistic: 0.07762314<br />gene_1: Pdgfra","globalHigherOrderFunction: 0.3099086<br />higherOrderStatistic: 0.12466192<br />gene_1: Per2","globalHigherOrderFunction: 0.3357535<br />higherOrderStatistic: 0.29624618<br />gene_1: Perp","globalHigherOrderFunction: 0.2026437<br />higherOrderStatistic: 0.07304853<br />gene_1: Pim2","globalHigherOrderFunction: 0.4542080<br />higherOrderStatistic: 0.56965654<br />gene_1: Pitx1","globalHigherOrderFunction: 0.2397866<br />higherOrderStatistic: 0.05424588<br />gene_1: Plp1","globalHigherOrderFunction: 0.2147951<br />higherOrderStatistic: 0.05840434<br />gene_1: Plvap","globalHigherOrderFunction: 2.4988430<br />higherOrderStatistic: 0.40998662<br />gene_1: Podxl","globalHigherOrderFunction: 0.2460096<br />higherOrderStatistic: 0.16397672<br />gene_1: Popdc2","globalHigherOrderFunction: 0.4197340<br />higherOrderStatistic: 0.22379415<br />gene_1: Prdm1","globalHigherOrderFunction: 0.2330509<br />higherOrderStatistic: 0.08341045<br />gene_1: Prrx1","globalHigherOrderFunction: 0.4160564<br />higherOrderStatistic: 0.17780336<br />gene_1: Prrx2","globalHigherOrderFunction: 0.2792623<br />higherOrderStatistic: 0.10058222<br />gene_1: Ptn","globalHigherOrderFunction: 0.4537272<br />higherOrderStatistic: 0.09915062<br />gene_1: Ramp2","globalHigherOrderFunction: 0.3145716<br />higherOrderStatistic: 0.13081957<br />gene_1: Rgl1","globalHigherOrderFunction: 0.2221189<br />higherOrderStatistic: 0.18269143<br />gene_1: Runx1","globalHigherOrderFunction: 1.1842563<br />higherOrderStatistic: 0.12340876<br />gene_1: Setd1a","globalHigherOrderFunction: 1.0084242<br />higherOrderStatistic: 0.16965359<br />gene_1: Setd1b","globalHigherOrderFunction: 0.7569050<br />higherOrderStatistic: 0.09275774<br />gene_1: Setd2","globalHigherOrderFunction: 0.3426839<br />higherOrderStatistic: 0.14286593<br />gene_1: Sfrp1","globalHigherOrderFunction: 0.2399858<br />higherOrderStatistic: 0.09174292<br />gene_1: Sfrp2","globalHigherOrderFunction: 1.6203130<br />higherOrderStatistic: 0.67710109<br />gene_1: Shh","globalHigherOrderFunction: 1.1630577<br />higherOrderStatistic: 0.75174986<br />gene_1: Shisa2","globalHigherOrderFunction: 0.4316049<br />higherOrderStatistic: 0.34742425<br />gene_1: Six1","globalHigherOrderFunction: 0.4237789<br />higherOrderStatistic: 0.10223663<br />gene_1: Smarcd3","globalHigherOrderFunction: 0.2810399<br />higherOrderStatistic: 0.10097882<br />gene_1: Smim1","globalHigherOrderFunction: 0.7860524<br />higherOrderStatistic: 0.34652477<br />gene_1: Smoc2","globalHigherOrderFunction: 0.3663647<br />higherOrderStatistic: 0.07626656<br />gene_1: Snai1","globalHigherOrderFunction: 0.2388555<br />higherOrderStatistic: 0.11384079<br />gene_1: Socs2","globalHigherOrderFunction: 1.4146298<br />higherOrderStatistic: 0.39927391<br />gene_1: Sox2","globalHigherOrderFunction: 2.1829706<br />higherOrderStatistic: 0.41994711<br />gene_1: Sox4","globalHigherOrderFunction: 0.5467631<br />higherOrderStatistic: 0.42476491<br />gene_1: Sp5","globalHigherOrderFunction: 0.4747255<br />higherOrderStatistic: 0.10639253<br />gene_1: Suz12","globalHigherOrderFunction: 0.4780897<br />higherOrderStatistic: 0.56488052<br />gene_1: T","globalHigherOrderFunction: 0.5859408<br />higherOrderStatistic: 0.34040256<br />gene_1: Tagln","globalHigherOrderFunction: 0.8738609<br />higherOrderStatistic: 0.46242378<br />gene_1: Tbx1","globalHigherOrderFunction: 0.4977646<br />higherOrderStatistic: 0.45279942<br />gene_1: Tbx3","globalHigherOrderFunction: 0.2095726<br />higherOrderStatistic: 0.08365472<br />gene_1: Tbx4","globalHigherOrderFunction: 1.6886508<br />higherOrderStatistic: 0.14457352<br />gene_1: Tcf7l1","globalHigherOrderFunction: 2.3593472<br />higherOrderStatistic: 0.12042512<br />gene_1: Tead2","globalHigherOrderFunction: 0.3203004<br />higherOrderStatistic: 0.05147432<br />gene_1: Tet1","globalHigherOrderFunction: 0.2611588<br />higherOrderStatistic: 0.06210934<br />gene_1: Tet2","globalHigherOrderFunction: 1.2984822<br />higherOrderStatistic: 0.09545655<br />gene_1: Tet3","globalHigherOrderFunction: 0.5071661<br />higherOrderStatistic: 0.26578096<br />gene_1: Thbs1","globalHigherOrderFunction: 0.2961208<br />higherOrderStatistic: 0.11300256<br />gene_1: Tinagl1","globalHigherOrderFunction: 1.3084663<br />higherOrderStatistic: 0.26947466<br />gene_1: Tjp2","globalHigherOrderFunction: 0.2734406<br />higherOrderStatistic: 0.12077107<br />gene_1: Ttn","globalHigherOrderFunction: 0.3190321<br />higherOrderStatistic: 0.06778402<br />gene_1: Usp28","globalHigherOrderFunction: 0.6052358<br />higherOrderStatistic: 0.09406618<br />gene_1: Wdr5","globalHigherOrderFunction: 0.3358199<br />higherOrderStatistic: 0.17229958<br />gene_1: Wnt5a","globalHigherOrderFunction: 0.2203004<br />higherOrderStatistic: 0.10492410<br />gene_1: Wnt5b","globalHigherOrderFunction: 1.1622405<br />higherOrderStatistic: 0.12082808<br />gene_1: Xist","globalHigherOrderFunction: 0.7440820<br />higherOrderStatistic: 0.11892960<br />gene_1: Zfp444","globalHigherOrderFunction: 0.5955189<br />higherOrderStatistic: 0.13012772<br />gene_1: Zfp57"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.6692913385826778,"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.824636418824401,"r":7.3059360730593621,"b":41.778974318367787,"l":43.105022831050235},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.043398312271026723,3.3203735348361243],"tickmode":"array","ticktext":["1","2","3"],"tickvals":[1,2.0000000000000004,3],"categoryorder":"array","categoryarray":["1","2","3"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176002,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"Mean across all cells","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.010615637090506544,0.78704197105575135],"tickmode":"array","ticktext":["0.2","0.4","0.6"],"tickvals":[0.20000000000000001,0.40000000000000002,0.60000000000000009],"categoryorder":"array","categoryarray":["0.2","0.4","0.6"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176002,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"scHOT statistic for local weightedMean","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"fc0c48a92002":{"x":{},"y":{},"label":{},"type":"scatter"}},"cur_data":"fc0c48a92002","visdat":{"fc0c48a92002":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p>Set up the permutation testing schema. For the purposes of this
workshop we set a low number of permutations over a low number of genes
in the testing scaffold, you may want to change this as you work through
the workshop yourself. The testing will take a few minutes to run, here
with the parallel parameters that were set at the beginning of this
document.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_setPermutationScaffold.html" class="external-link">scHOT_setPermutationScaffold</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>,</span>
<span>                                              numberPermutations <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                                              numberScaffold <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_performPermutationTest.html" class="external-link">scHOT_performPermutationTest</a></span><span class="op">(</span></span>
<span>    <span class="va">scHOT_spatial</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Permutation testing combination 40 of 165...</span></span>
<span><span class="co">#&gt; Permutation testing combination 80 of 165...</span></span>
<span><span class="co">#&gt; Permutation testing combination 110 of 165...</span></span>
<span><span class="co">#&gt; Permutation testing combination 150 of 165...</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; DataFrame with 165 rows and 9 columns</span></span>
<span><span class="co">#&gt;              gene_1 globalHigherOrderFunction            higherOrderSequence</span></span>
<span><span class="co">#&gt;         &lt;character&gt;                  &lt;matrix&gt;                  &lt;NumericList&gt;</span></span>
<span><span class="co">#&gt; Acvr1         Acvr1                  0.216666 0.251205,0.275076,0.286668,...</span></span>
<span><span class="co">#&gt; Acvr2a       Acvr2a                  0.375776 0.398236,0.376223,0.361763,...</span></span>
<span><span class="co">#&gt; Ahnak         Ahnak                  0.976418    1.23931,1.22101,1.19278,...</span></span>
<span><span class="co">#&gt; Akr1c19     Akr1c19                  0.744070 0.681732,0.622183,0.625407,...</span></span>
<span><span class="co">#&gt; Aldh1a2     Aldh1a2                  0.245981 0.117491,0.118105,0.121221,...</span></span>
<span><span class="co">#&gt; ...             ...                       ...                            ...</span></span>
<span><span class="co">#&gt; Wnt5a         Wnt5a                  0.335820 0.282418,0.280240,0.268180,...</span></span>
<span><span class="co">#&gt; Wnt5b         Wnt5b                  0.220300 0.262440,0.321449,0.368172,...</span></span>
<span><span class="co">#&gt; Xist           Xist                  1.162241    1.18893,1.17123,1.18238,...</span></span>
<span><span class="co">#&gt; Zfp444       Zfp444                  0.744082 0.529888,0.531771,0.538540,...</span></span>
<span><span class="co">#&gt; Zfp57         Zfp57                  0.595519 0.853046,0.844188,0.838651,...</span></span>
<span><span class="co">#&gt;         higherOrderStatistic numberPermutations storePermutations</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt;          &lt;numeric&gt;         &lt;logical&gt;</span></span>
<span><span class="co">#&gt; Acvr1              0.0750954                  0              TRUE</span></span>
<span><span class="co">#&gt; Acvr2a             0.0665143                  0              TRUE</span></span>
<span><span class="co">#&gt; Ahnak              0.3319897                  0              TRUE</span></span>
<span><span class="co">#&gt; Akr1c19            0.1673342                  0              TRUE</span></span>
<span><span class="co">#&gt; Aldh1a2            0.1827836                 50              TRUE</span></span>
<span><span class="co">#&gt; ...                      ...                ...               ...</span></span>
<span><span class="co">#&gt; Wnt5a               0.172300                  0              TRUE</span></span>
<span><span class="co">#&gt; Wnt5b               0.104924                  0              TRUE</span></span>
<span><span class="co">#&gt; Xist                0.120828                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp444              0.118930                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp57               0.130128                  0              TRUE</span></span>
<span><span class="co">#&gt;                              permutations pvalPermutations FDRPermutations</span></span>
<span><span class="co">#&gt;                             &lt;NumericList&gt;        &lt;numeric&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Acvr2a                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Ahnak                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Akr1c19                                NA               NA              NA</span></span>
<span><span class="co">#&gt; Aldh1a2 0.0448634,0.0877575,0.0508914,...        0.0196078       0.0244444</span></span>
<span><span class="co">#&gt; ...                                   ...              ...             ...</span></span>
<span><span class="co">#&gt; Wnt5a                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Wnt5b                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Xist                                   NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp444                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp57                                  NA               NA              NA</span></span></code></pre>
<p>After the permutation test we can estimate the P-values across all
genes.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_plotPermutationDistributions.html" class="external-link">scHOT_plotPermutationDistributions</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_estimatePvalues.html" class="external-link">scHOT_estimatePvalues</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>,</span>
<span>                                       nperm_estimate <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                                       maxDist <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; DataFrame with 165 rows and 14 columns</span></span>
<span><span class="co">#&gt;              gene_1 globalHigherOrderFunction            higherOrderSequence</span></span>
<span><span class="co">#&gt;         &lt;character&gt;                  &lt;matrix&gt;                  &lt;NumericList&gt;</span></span>
<span><span class="co">#&gt; Acvr1         Acvr1                  0.216666 0.251205,0.275076,0.286668,...</span></span>
<span><span class="co">#&gt; Acvr2a       Acvr2a                  0.375776 0.398236,0.376223,0.361763,...</span></span>
<span><span class="co">#&gt; Ahnak         Ahnak                  0.976418    1.23931,1.22101,1.19278,...</span></span>
<span><span class="co">#&gt; Akr1c19     Akr1c19                  0.744070 0.681732,0.622183,0.625407,...</span></span>
<span><span class="co">#&gt; Aldh1a2     Aldh1a2                  0.245981 0.117491,0.118105,0.121221,...</span></span>
<span><span class="co">#&gt; ...             ...                       ...                            ...</span></span>
<span><span class="co">#&gt; Wnt5a         Wnt5a                  0.335820 0.282418,0.280240,0.268180,...</span></span>
<span><span class="co">#&gt; Wnt5b         Wnt5b                  0.220300 0.262440,0.321449,0.368172,...</span></span>
<span><span class="co">#&gt; Xist           Xist                  1.162241    1.18893,1.17123,1.18238,...</span></span>
<span><span class="co">#&gt; Zfp444       Zfp444                  0.744082 0.529888,0.531771,0.538540,...</span></span>
<span><span class="co">#&gt; Zfp57         Zfp57                  0.595519 0.853046,0.844188,0.838651,...</span></span>
<span><span class="co">#&gt;         higherOrderStatistic numberPermutations storePermutations</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt;          &lt;numeric&gt;         &lt;logical&gt;</span></span>
<span><span class="co">#&gt; Acvr1              0.0750954                  0              TRUE</span></span>
<span><span class="co">#&gt; Acvr2a             0.0665143                  0              TRUE</span></span>
<span><span class="co">#&gt; Ahnak              0.3319897                  0              TRUE</span></span>
<span><span class="co">#&gt; Akr1c19            0.1673342                  0              TRUE</span></span>
<span><span class="co">#&gt; Aldh1a2            0.1827836                 50              TRUE</span></span>
<span><span class="co">#&gt; ...                      ...                ...               ...</span></span>
<span><span class="co">#&gt; Wnt5a               0.172300                  0              TRUE</span></span>
<span><span class="co">#&gt; Wnt5b               0.104924                  0              TRUE</span></span>
<span><span class="co">#&gt; Xist                0.120828                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp444              0.118930                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp57               0.130128                  0              TRUE</span></span>
<span><span class="co">#&gt;                              permutations pvalPermutations FDRPermutations</span></span>
<span><span class="co">#&gt;                             &lt;NumericList&gt;        &lt;numeric&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Acvr2a                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Ahnak                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Akr1c19                                NA               NA              NA</span></span>
<span><span class="co">#&gt; Aldh1a2 0.0448634,0.0877575,0.0508914,...        0.0196078       0.0244444</span></span>
<span><span class="co">#&gt; ...                                   ...              ...             ...</span></span>
<span><span class="co">#&gt; Wnt5a                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Wnt5b                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Xist                                   NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp444                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp57                                  NA               NA              NA</span></span>
<span><span class="co">#&gt;         numberPermutationsEstimated globalLowerRangeEstimated</span></span>
<span><span class="co">#&gt;                           &lt;integer&gt;                 &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                           200                  0.225545</span></span>
<span><span class="co">#&gt; Acvr2a                           50                  0.416245</span></span>
<span><span class="co">#&gt; Ahnak                            50                  1.008424</span></span>
<span><span class="co">#&gt; Akr1c19                         100                  0.788412</span></span>
<span><span class="co">#&gt; Aldh1a2                         200                  0.225545</span></span>
<span><span class="co">#&gt; ...                             ...                       ...</span></span>
<span><span class="co">#&gt; Wnt5a                           150                  0.239986</span></span>
<span><span class="co">#&gt; Wnt5b                           200                  0.225545</span></span>
<span><span class="co">#&gt; Xist                            100                  1.147287</span></span>
<span><span class="co">#&gt; Zfp444                          100                  0.788412</span></span>
<span><span class="co">#&gt; Zfp57                          1100                  0.225545</span></span>
<span><span class="co">#&gt;         globalUpperRangeEstimated pvalEstimated FDREstimated</span></span>
<span><span class="co">#&gt;                         &lt;numeric&gt;     &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                    0.245981    0.07500000    0.0937500</span></span>
<span><span class="co">#&gt; Acvr2a                   0.416245    0.54000000    0.5568750</span></span>
<span><span class="co">#&gt; Ahnak                    1.008424    0.01960784    0.0281330</span></span>
<span><span class="co">#&gt; Akr1c19                  0.809843    0.00990099    0.0194118</span></span>
<span><span class="co">#&gt; Aldh1a2                  0.245981    0.00497512    0.0155660</span></span>
<span><span class="co">#&gt; ...                           ...           ...          ...</span></span>
<span><span class="co">#&gt; Wnt5a                    0.416245    0.00662252    0.0182119</span></span>
<span><span class="co">#&gt; Wnt5b                    0.245981    0.00497512    0.0155660</span></span>
<span><span class="co">#&gt; Xist                     1.163510    0.10000000    0.1170213</span></span>
<span><span class="co">#&gt; Zfp444                   0.809843    0.16000000    0.1846154</span></span>
<span><span class="co">#&gt; Zfp57                    3.171420    0.09000000    0.1091912</span></span></code></pre>
<p>We can now examine the spatial expression of the 5 most significant
genes, both in our scHOT object and over our original spe object.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_sorted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>,</span>
<span>                                                                <span class="st">"scHOT_output"</span><span class="op">)</span><span class="op">$</span><span class="va">pvalEstimated</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">topgenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">output_sorted</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">topgene</span> <span class="kw">in</span> <span class="va">topgenes</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">g_spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">topgene</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">g_scHOT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">topgene</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                           by_exprs_values <span class="op">=</span> <span class="st">"expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">g_all</span> <span class="op">&lt;-</span> <span class="va">g_scHOT</span> <span class="op">+</span> <span class="va">g_spe</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">g_all</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-11-1.png" width="700"><img src="workshop_material_files/figure-html/unnamed-chunk-11-2.png" width="700"><img src="workshop_material_files/figure-html/unnamed-chunk-11-3.png" width="700"><img src="workshop_material_files/figure-html/unnamed-chunk-11-4.png" width="700"><img src="workshop_material_files/figure-html/unnamed-chunk-11-5.png" width="700"></p>
<p>Here we are noting the genes that are found to have the most
statistically significant spatial variation in their local mean
expression. These genes point to specific patterns that govern the
development of individual parts of the gut tube.</p>
<div class="question">
<p><strong>Advanced/Extended Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How would you perform such testing over multiple distinct
samples?</li>
<li>scHOT is developed with all higher order testing in mind, use the
associated <a href="http://www.bioconductor.org/packages/release/bioc/html/scHOT.html" class="external-link">vignette</a>
to get towards assessing changes in variation or correlation structure
in space.</li>
</ol>
</div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## try some code</span></span></code></pre></div>
<p>Now that we have assessed genes varying in expression in space, let’s
look further into the IMC data where we will make use of the multiple
samples to perform clustering and extract some biological
understanding.</p>
</div>
<div class="section level4">
<h4 id="feature-generation">Feature generation<a class="anchor" aria-label="anchor" href="#feature-generation"></a>
</h4>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Farhan Ameen, Ellis Patrick, Shila Ghazanfar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
