<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Unlocking single cell spatial omics analyses with scdney • ScdneySpatial</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney">
<meta property="og:description" content="ScdneySpatial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ScdneySpatial</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/workshop_material.html">Unlocking single cell spatial omics analyses with scdney</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/ScdneySpatial" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="workshop_material_files/htmlwidgets-1.6.3/htmlwidgets.js"></script><script src="workshop_material_files/plotly-binding-4.10.3/plotly.js"></script><script src="workshop_material_files/typedarray-0.1/typedarray.min.js"></script><link href="workshop_material_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="workshop_material_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><link href="workshop_material_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="workshop_material_files/plotly-main-2.11.1/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Unlocking single cell spatial omics analyses
with scdney</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/ScdneySpatial/blob/HEAD/vignettes/workshop_material.Rmd" class="external-link"><code>vignettes/workshop_material.Rmd</code></a></small>
      <div class="hidden name"><code>workshop_material.Rmd</code></div>

    </div>

    
    
<style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>
<p>Farhan Ameen<span class="math inline">\(^{1,2,3}\)</span>, Alex
Qin<span class="math inline">\(^{1,2,3}\)</span>, Nick Robertson<span class="math inline">\(^{1,2,3}\)</span>, Shila Ghazanfar<span class="math inline">\(^{2,3}\)</span>, Ellis Patrick<span class="math inline">\(^{1,2,3}\)</span>.</p>
<p><span class="math inline">\(^1\)</span> Westmead Institute for
Medical Research, University of Sydney, Australia<br><span class="math inline">\(^2\)</span> Sydney Precision Data Science
Centre, University of Sydney, Australia<br><span class="math inline">\(^3\)</span> School of Mathematics and
Statistics, University of Sydney, Australia</p>
<p><br> Contact: ellis.patrick@sydney.edu.au</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>Understanding the interplay between different types of cells and
their immediate environment is critical for understanding the mechanisms
of cells themselves and their function in the context of human diseases.
Recent advances in high dimensional in situ cytometry technologies have
fundamentally revolutionized our ability to observe these complex
cellular relationships providing an unprecedented characterisation of
cellular heterogeneity in a tissue environment.</p>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>In this workshop we will introduce some of the key analytical
concepts needed to analyse data from high dimensional spatial omics
technologies such as, PhenoCycler, IMC, Xenium and MERFISH. We will show
how functionality from our Bioconductor packages simpleSeg, FuseSOM,
scClassify, scHot, spicyR, listClust, statial, scFeatures and ClassifyR
can be used to address various biological hypotheses. By the end of this
workshop attendees will be able to implement and assess some of the key
steps of a spatial analysis pipeline including cell segmentation,
feature normalisation, cell type identification, microenvironment and
cell-state characterisation, spatial hypothesis testing and patient
classification. Understanding these key steps will provide attendees
with the core skills needed to interrogate the comprehensive spatial
information generated by these exciting new technologies.</p>
</div>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<p>It is expected that students will have:</p>
<ul>
<li>basic knowledge of R syntax,</li>
<li>this workshop will not provide an in-depth description of
cell-resolution spatial omics technologies.</li>
</ul>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<p>Several single cell R packages will be used from the scdney package,
for more information visit: <a href="https://sydneybiox.github.io/scdney/" class="external-link uri">https://sydneybiox.github.io/scdney/</a>.</p>
</div>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Data visualisation &amp; cell segmentation</td>
<td>1h 30m</td>
</tr>
<tr class="even">
<td>Cell type clustering &amp; classification</td>
<td>1h 30m</td>
</tr>
<tr class="odd">
<td>Spatial analysis &amp; feature generation</td>
<td>2h</td>
</tr>
<tr class="even">
<td>Patient Classification</td>
<td>1h 30m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>Understand and visualise spatial omics datasets.</li>
<li>Identify key biological questions that can be addressed with these
technologies and spatial analysis.</li>
<li>Understand the key analytical steps involved in spatial omics
analysis, and perform these steps using R.</li>
<li>Evaluate the performance of data normalisation and cell
segmentation.</li>
<li>Understand and generate individual feature representations from
spatial omics data.</li>
<li>Develop appreciation on how to assess performance of classification
models.</li>
<li>Perform disease outcome prediction using the feature representation
and robust classification framework.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"simpleSeg"</span>, <span class="st">"cytomapper"</span>, <span class="st">"scClassify"</span>, <span class="st">"scHOT"</span>, <span class="st">"FuseSOM"</span>,<span class="st">"glmnet"</span>, <span class="st">"spicyR"</span>, <span class="st">"lisaClust"</span>,<span class="st">"Statial"</span>, <span class="st">"scFeatures"</span>, <span class="st">"ClassifyR"</span>, <span class="st">"tidyverse"</span>, <span class="st">"scater"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"STexampleData"</span>, <span class="st">"SpatialDatasets"</span>, <span class="st">"tidySingleCellExperiment"</span>, <span class="st">"scuttle"</span>, <span class="st">"batchelor"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="download-datasets">Download datasets<a class="anchor" aria-label="anchor" href="#download-datasets"></a>
</h3>
<p>Please download these datasets before you start the workshop</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Download data now</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">STexampleData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/STexampleData/man/STexampleData.html" class="external-link">seqFISH_mouseEmbryo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html" class="external-link">spe_Keren_2018</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library(ScdneySpatial)</span></span>
<span></span>
<span><span class="co"># packages from scdney</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/" class="external-link">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scClassify</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scFeatures</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scHOT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/simpleSeg/" class="external-link">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/SpatialDatasets" class="external-link">SpatialDatasets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/" class="external-link">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial" class="external-link">Statial</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Other required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/BumpyMatrix" class="external-link">BumpyMatrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pharmaverse/ggsurvfit" class="external-link">ggsurvfit</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glmnet.stanford.edu" class="external-link">glmnet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/reshape" class="external-link">reshape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lmweber/STexampleData" class="external-link">STexampleData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">8</span>  <span class="co"># Feel free to parallelise things if you have the cores to spare.</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/generateBPParam.html" class="external-link">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"celltype_colours.R"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_SingleCellExperiment_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-data">The data<a class="anchor" aria-label="anchor" href="#the-data"></a>
</h3>
<p>In this workshop, we will be working with two datasets to explore how
biological phenotypes, cellular interactions, and patterns of gene
expression are correlated with disease. Both of these datasets will be
used in different contexts, hopefully these contexts are representative
of scenarios you will encounter in your own datasets.</p>
<p>We will use two motivating datasets:</p>
<ul>
<li>
<a href="https://www.cell.com/fulltext/S0092-8674(18)31100-0" class="external-link">Keren
et al, 2018</a>: A multiplexed ion beam imaging by time-of-flight
(MIBI-TOF) dataset profilining tissue from triple-negative breast cancer
patients. The primary question we will address with this dataset is if
we can predict risk of cancer recurrence and overall survival time based
on imaging data?</li>
<li>
<a href="https://www.nature.com/articles/s41587-021-01006-2" class="external-link">Lohoff
et al, 2022</a>: A seqFISH study of early mouse organogenesis. We will
use a subset of data that is made available from the STExampleData
package. The primary question we will address with this dataset is if we
can identify key transcriptomic drivers of the developing brain?</li>
</ul>
</div>
<div class="section level3">
<h3 id="data-visualisation-and-exploration">Data visualisation and exploration<a class="anchor" aria-label="anchor" href="#data-visualisation-and-exploration"></a>
</h3>
<p>The purpose of the this section is primarily to introduce the
<code>SpatialExperiment</code> class which is used to store information
from the imaging experiments in R. The goal will be to get comfortable
enough manipulating and exploring these objects so that you can progress
through the remainder of the workshop comfortably. Here we will download
a dataset stored in the <code>STexampleData</code> R package , examine
the structure, visualise the data and perform some exploratory
analyses.</p>
<div class="section level4">
<h4 id="seqfish-mouse-embryo">SeqFISH mouse embryo<a class="anchor" aria-label="anchor" href="#seqfish-mouse-embryo"></a>
</h4>
<p>Here we download the seqFISH mouse embryo data. This comes in the
format of a <code>SpatialExperiment</code> object, where summarized
information from an imaging dataset can be compiled and accessed with
relative ease.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#spe &lt;- STexampleData::seqFISH_mouseEmbryo()</span></span>
<span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 351 11026 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts molecules</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(11026): embryo1_Pos0_cell10_z2 embryo1_Pos0_cell100_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos28_cell97_z2 embryo1_Pos28_cell98_z2</span></span>
<span><span class="co">#&gt; colData names(14): cell_id embryo ... segmentation_vertices sample_id</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span></code></pre>
<p>We can use functions designed for <code>SingleCellExperiment</code>
objects in the <code>scater</code> package for plotting via the
<code>reducedDim</code> slot. We multiply the spatial coordinates by a
matrix to flip the y-axis and ensure we fix the aspect ratio.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span><span class="va">coord_transform</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">coord_transform</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sox2"</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/visualiseSeqFISH-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How many cells are in this data?</li>
<li>How many genes?</li>
<li>Plot gene expression mapping point size to the cell area.</li>
</ol>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># try to answer the above question using the spe object. </span></span>
<span><span class="co"># you may want to check the SingleCellExperiment vignette.</span></span>
<span><span class="co"># https://bioconductor.org/packages/3.17/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html</span></span></code></pre></div>
<p>We can perform a typical gene-expression based analysis for this
data. Later in part two we will perform some specific analytical
techniques, but for now let’s explore the dataset and use methods
designed for single cell data.</p>
<p>Typically in single-cell data analysis, we perform dimension
reduction to project the high dimensional cell x gene matrix on to 2D
space. This allows us to visualise various things of interest, such as
distribution of cell types and disease outcomes. Here, we will see how
cells are segregated by their expression of Sox2.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b.out</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/batchCorrect.html" class="external-link">batchCorrect</a></span><span class="op">(</span><span class="va">spe</span>, batch <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">pos</span>, assay.type <span class="op">=</span> <span class="st">"logcounts"</span>, PARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">FastMnnParam</a></span><span class="op">(</span>d<span class="op">=</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"FastMnn"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">b.out</span>, <span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"FastMnn"</span><span class="op">)</span></span>
<span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 351 11026 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts molecules logcounts</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(11026): embryo1_Pos0_cell10_z2 embryo1_Pos0_cell100_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos28_cell97_z2 embryo1_Pos28_cell98_z2</span></span>
<span><span class="co">#&gt; colData names(15): cell_id embryo ... sample_id sizeFactor</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_celltype_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"UMAP"</span>, colour_by <span class="op">=</span> <span class="st">"celltype_mapped_refined"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">celltype_colours</span><span class="op">)</span></span>
<span><span class="va">g_celltype_umap</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/seqFISHEDA-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"UMAP"</span>, colour_by <span class="op">=</span> <span class="st">"Sox2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/seqFISHEDA-2.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_celltype_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"celltype_mapped_refined"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">celltype_colours</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_all</span> <span class="op">&lt;-</span> <span class="va">g_celltype_spatial</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> <span class="va">g_celltype_umap</span></span>
<span><span class="va">g_all</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/seqFISHEDA-3.png" width="700"></p>
<div class="question">
<p><strong>Advanced/Extension Question</strong></p>
<ol style="list-style-type: decimal">
<li>What considerations need to be made for batch correction of spatial
data? What assumptions are being made and/or broken? How could you check
this?</li>
<li>Check out the <a href="https://davidgohel.github.io/ggiraph/index.html" class="external-link"><code>ggiraph</code></a>
package for extending the <code>g_all</code> object to an interactive
plot with a tooltip that links the spatial and UMAP coordinate systems.
(Hint: This may involve generating a new ggplot object outside of the
<code>plotReducedDim</code> function.)</li>
</ol>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># try to examine answer the above questions using the spe object. </span></span>
<span><span class="co"># you may want to set up some small simulation..</span></span></code></pre></div>
<p>At this point we will pause our examination of the seqFISH dataset
that is in the object <code>spe</code>, and turn over to the second
example dataset. In the second part we will revisit this data for
performing <code>scHOT</code> testing.</p>
</div>
<div class="section level4">
<h4 id="mibi-tof-breast-cancer">MIBI-TOF breast cancer<a class="anchor" aria-label="anchor" href="#mibi-tof-breast-cancer"></a>
</h4>
<p>As part of scdney package infrastructure, we provide several
pre-processed datasets which can be loaded in with the
<code>SpatialDatasets</code> package. Below, we’ve loaded in the
MIBI-TOF triple negative breast cancer data from Keren et al. into a
<code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load in Keren et al. data.</span></span>
<span><span class="co">#kerenSPE = SpatialDatasets::spe_Keren_2018() </span></span>
<span></span>
<span><span class="co"># Remove cold tumour types and patients with missing survival data.</span></span>
<span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tumour_type</span> <span class="op">!=</span> <span class="st">"cold"</span>,</span>
<span>         <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">`Survival_days_capped.`</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>event <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">Censored</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">$</span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 48 170171 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): intensities</span></span>
<span><span class="co">#&gt; rownames(48): Na Si ... Ta Au</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(170171): 1 2 ... 197677 197678</span></span>
<span><span class="co">#&gt; colData names(43): CellID imageID ... x y</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre>
<p>At the start of any analysis, it is often good to explore the data to
get a sense of complexity. We can do this by exploring the distribution
of the outcomes and variables in the patients’ meta-data.</p>
<p>Try starting off your exploration by answering the below
questions.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How many cells are in this data?</li>
<li>How many markers? How many images?</li>
</ol>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># try to answer the above question using the imc object. </span></span>
<span><span class="co"># you may want to check the SpatialExperiment vignette.</span></span>
<span><span class="co"># https://www.bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html</span></span></code></pre></div>
<p>Again, we can perform dimension reduction to visualize this dataset
on a 2D plane.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Visualise the UMAP using the <code>plotReducedDim</code> function
and colour the UMAP by <code>cellType</code>. What does this UMAP tell
us?</li>
<li>What are some observations we could make if we coloured by
<code>imageID</code>?</li>
</ol>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">kerenSPE</span>, exprs_values <span class="op">=</span> <span class="st">"intensities"</span>, name <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># try to answer the question here!</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="cell-segmentation-evaluation">Cell segmentation &amp; evaluation<a class="anchor" aria-label="anchor" href="#cell-segmentation-evaluation"></a>
</h3>
<div class="section level4">
<h4 id="reading-images">Reading images<a class="anchor" aria-label="anchor" href="#reading-images"></a>
</h4>
<p>To load in our images we use the <code>loadImages</code> function
from <code>cytomapper</code>, here we use the patient 5 image from Keren
et al. as an example.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imageLocation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kerenPatient5.tiff"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span></span>
<span><span class="va">image5</span> <span class="op">=</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">imageLocation</span>,</span>
<span>  as.is <span class="op">=</span> <span class="cn">TRUE</span> <span class="co">#Needed as 8-bit image</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">image5</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"imageID"</span> <span class="op">=</span> <span class="st">"kerenPatient5"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Setting the channel names according to orginal paper. </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html" class="external-link">channelNames</a></span><span class="op">(</span><span class="va">image5</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Au"</span>, <span class="st">"Background"</span>, <span class="st">"Beta catenin"</span>, <span class="st">"Ca"</span>, <span class="st">"CD11b"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD138"</span>, <span class="st">"CD16"</span>, <span class="st">"CD20"</span>, <span class="st">"CD209"</span>, <span class="st">"CD3"</span>, <span class="st">"CD31"</span>, <span class="st">"CD4"</span>, <span class="st">"CD45"</span>, <span class="st">"CD45RO"</span>, <span class="st">"CD56"</span>, <span class="st">"CD63"</span>, <span class="st">"CD68"</span>, <span class="st">"CD8"</span>, <span class="st">"dsDNA"</span>, <span class="st">"EGFR"</span>, <span class="st">"Fe"</span>, <span class="st">"FoxP3"</span>, <span class="st">"H3K27me3"</span>, <span class="st">"H3K9ac"</span>, <span class="st">"HLA_Class_1"</span>, <span class="st">"HLA_DR"</span>, <span class="st">"IDO"</span>, <span class="st">"Keratin17"</span>, <span class="st">"Keratin6"</span>, <span class="st">"Ki67"</span>, <span class="st">"Lag3"</span>, <span class="st">"MPO"</span>, <span class="st">"Na"</span>, <span class="st">"P"</span>, <span class="st">"p53"</span>, <span class="st">"Pan-Keratin"</span>, <span class="st">"PD-L1"</span>, <span class="st">"PD1"</span>, <span class="st">"phospho-S6"</span>, <span class="st">"Si"</span>, <span class="st">"SMA"</span>, <span class="st">"Ta"</span>, <span class="st">"Vimentin"</span><span class="op">)</span></span></code></pre></div>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>What class is image5? Hint: class()</li>
<li>How many images and markers are in image5?</li>
<li>Challenge: What is the dimension of the image5 image?</li>
</ol>
</div>
</div>
<div class="section level4">
<h4 id="visualise-an-image">Visualise an image<a class="anchor" aria-label="anchor" href="#visualise-an-image"></a>
</h4>
<p>We can visualise this image to see what we have read in. Lets
highlight 4 markers.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance another way.</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image <span class="op">=</span> <span class="va">image5</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45"</span>, <span class="st">"Pan-Keratin"</span>, <span class="st">"SMA"</span>, <span class="st">"dsDNA"</span><span class="op">)</span>,</span>
<span>  colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/plotImage-1.png" width="700"></p>
<p>We can manipulate the brightness, contrast and gamma levels as
follows. See if you can do a better job.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance another way.</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image <span class="op">=</span> <span class="va">image5</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45"</span>, <span class="st">"Pan-Keratin"</span>, <span class="st">"SMA"</span>, <span class="st">"dsDNA"</span><span class="op">)</span>,</span>
<span>  display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span>  ,</span>
<span>  <span class="co"># Adjust the brightness, contrast and gamma of each channel.</span></span>
<span>  bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/plotImage2-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="can-we-identify-the-cells-in-the-image">Can we identify the cells in the image?<a class="anchor" aria-label="anchor" href="#can-we-identify-the-cells-in-the-image"></a>
</h4>
<p>The <code>EBImage</code> package on Bioconductor provides a lot of
useful functions for manipulating imaging data in R. This includes
functionality for finding cells, process called cell segmentation. Lets
work through an example from their vignette. This will use some
functionality that complements that which you’ve already learnt.</p>
<p>We start by loading the images of nuclei and cell bodies. To
visualize the cells we overlay these images as the green and the blue
channel of a false-color image. Notice, that with display you can
zoom!</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nuc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/io.html" class="external-link">readImage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'images'</span>, <span class="st">'nuclei.tif'</span>, package<span class="op">=</span><span class="st">'EBImage'</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">cel</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/io.html" class="external-link">readImage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'images'</span>, <span class="st">'cells.tif'</span>, package<span class="op">=</span><span class="st">'EBImage'</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="va">cells</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/channel.html" class="external-link">rgbImage</a></span><span class="op">(</span>green<span class="op">=</span><span class="fl">1.5</span><span class="op">*</span><span class="va">cel</span>, blue<span class="op">=</span><span class="va">nuc</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">cells</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/readEBImage-1.png" width="700"></p>
<p>We will next segment the nuclei. The <code>nuc</code> channel
contains fluorescent intensities of a protein expressed in the nuclei of
cells. First we create a nuclei mask which will threshold this channel
to separate signal from noise and then clean this with some
morphological operations. Then we identify single nuclei using
<code>bwlabel</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/thresh.html" class="external-link">thresh</a></span><span class="op">(</span><span class="va">nuc</span>, w<span class="op">=</span><span class="fl">10</span>, h<span class="op">=</span><span class="fl">10</span>, offset<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">nmask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">opening</a></span><span class="op">(</span><span class="va">nmask</span>, <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">makeBrush</a></span><span class="op">(</span><span class="fl">5</span>, shape<span class="op">=</span><span class="st">'disc'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nmask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/fillHull.html" class="external-link">fillHull</a></span><span class="op">(</span><span class="va">nmask</span><span class="op">)</span></span>
<span><span class="va">nmask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/bwlabel.html" class="external-link">bwlabel</a></span><span class="op">(</span><span class="va">nmask</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">nmask</span>, all<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Do you understand what <code>nmask</code> is?</li>
<li>What are you seeing when you do <code>table(nmask)</code>
</li>
<li>What happens when you using <code><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html" class="external-link">colorLabels()</a></code> with
<code>display</code>?</li>
</ol>
</div>
<p>Next, we use the segmented nuclei as seeds to perform Voronoi
segmentation of the cytoplasm.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctmask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">opening</a></span><span class="op">(</span><span class="va">cel</span><span class="op">&gt;</span><span class="fl">0.1</span>, <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">makeBrush</a></span><span class="op">(</span><span class="fl">5</span>, shape<span class="op">=</span><span class="st">'disc'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cmask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/propagate.html" class="external-link">propagate</a></span><span class="op">(</span><span class="va">cel</span>, seeds<span class="op">=</span><span class="va">nmask</span>, mask<span class="op">=</span><span class="va">ctmask</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">ctmask</span>, all<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We can then visualise our segmentations with
<code>paintObjects</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">segmented</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/paintObjects.html" class="external-link">paintObjects</a></span><span class="op">(</span><span class="va">cmask</span>, <span class="va">cells</span>, col<span class="op">=</span><span class="st">'#ff00ff'</span><span class="op">)</span></span>
<span><span class="va">segmented</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/paintObjects.html" class="external-link">paintObjects</a></span><span class="op">(</span><span class="va">nmask</span>, <span class="va">segmented</span>, col<span class="op">=</span><span class="st">'#ffff00'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">segmented</span>, all<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="can-we-do-this-in-one-step">Can we do this in one step?<a class="anchor" aria-label="anchor" href="#can-we-do-this-in-one-step"></a>
</h4>
<p>Images stored in a <code>list</code> or <code>CytoImageList</code>
can be segmented using <code>simpleSeg</code>. Below
<code>simpleSeg</code> will identify the nuclei in the image using the
<code>dsDNA</code>, <code>H3K27me3</code> and <code>H3K9ac</code>
channel. To estimate the cell body of the cells we will simply dilate
out from the nuclei by 3 pixels. We also have specified that the
channels be sqrt transformed, and the 99th quantile of values removed to
ensure our segmentation is not affected by outliers.</p>
<p>We can visualise the segmentations using the <code>display</code> and
<code>colorLabels</code> functions in <code>EBImage</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate segmentation masks</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html" class="external-link">simpleSeg</a></span><span class="op">(</span></span>
<span>  <span class="va">image5</span>,</span>
<span>  nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dsDNA"</span>, <span class="st">"H3K27me3"</span>, <span class="st">"H3K9ac"</span><span class="op">)</span>,</span>
<span>  cellBody <span class="op">=</span> <span class="st">"dilate"</span>, </span>
<span>  transform <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sqrt"</span>, <span class="st">"norm99"</span><span class="op">)</span>,</span>
<span>  sizeSelection <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  smooth <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  discSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="is-this-segmentation-appropriate">Is this segmentation appropriate?<a class="anchor" aria-label="anchor" href="#is-this-segmentation-appropriate"></a>
</h4>
<p>One way to assess if segmentation is appropriate is to look. We now
know two ways to do this.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance one way.</span></span>
<span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html" class="external-link">colorLabels</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> makes
it easy to overlay the masks on top of the intensities of markers in the
image. Here we can s</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance another way.</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image <span class="op">=</span> <span class="va">image5</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45"</span>, <span class="st">"Pan-Keratin"</span>, <span class="st">"SMA"</span>, <span class="st">"dsDNA"</span><span class="op">)</span>,</span>
<span>  display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="co"># Adjust the brightness, contrast and gamma of each channel.</span></span>
<span>  bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CD45 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    `Pan-Keratin` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    SMA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>    dsDNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/plotSegmentation-1.png" width="700"></p>
<p>Using the <code>table</code> function in R we can measure the pixel
area of each cell and plot this on a histogram. In this dataset 1 µm is
equivalent to <span class="math inline">\(2.56\)</span> pixels, we can
transform the pixel area of a cell to physical area by dividing the
pixel area by <span class="math inline">\(2.56^2\)</span>. We can see
below the majority of our cells are just below <span class="math inline">\(100µm^2\)</span>, you should decide whether or not
this is an appropriate size based on your knowledge of the cell types
being imaged.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get area of every cell</span></span>
<span><span class="va">cellSize</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fl">2.56</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter out first index, which represents background</span></span>
<span><span class="va">cellSize</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cellSize</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">30</span>,</span>
<span>       xlab <span class="op">=</span> <span class="st">"Pixel area of cell (µm^2)"</span>,</span>
<span>       main <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> </span></code></pre></div>
<p><img src="workshop_material_files/figure-html/pixelArea-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li><p>Do our segmentations look better if other parameters are used?
<em>Hint</em>: use the help function to see what other parameters are
available in <code>simpleSeg</code>.</p></li>
<li><p>We only have one image here, but how would you check the quality
of 100 images?</p></li>
</ol>
</div>
</div>
<div class="section level4">
<h4 id="summarising-images-into-cell-features">Summarising images into cell features?<a class="anchor" aria-label="anchor" href="#summarising-images-into-cell-features"></a>
</h4>
<p>With our segmentation <code>masks</code> we can convert our image to
a <code>SpatialExperiment</code> object using the
<code>measureObjects</code> function in <code>cytomapper</code>.
<code>measureObjects</code> will find the center of each cell and
represent the cell as an x and y coordinate. In addition the average
marker expression within each cell is summarised and stored in the
assays object of the <code>SpatialExperiment</code>. <em>NOTE:</em> If
you are using other segmentation tools or softwares, you can import the
masks as tiffs into R and use <code>measureObjects</code> to create a
<code>SpatialExperiment</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarise the expression of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects</a></span><span class="op">(</span></span>
<span>  mask <span class="op">=</span> <span class="va">masks</span>,</span>
<span>  image <span class="op">=</span> <span class="va">image5</span>,</span>
<span>  feature_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"basic"</span>, <span class="st">"moment"</span><span class="op">)</span>,</span>
<span>  basic_feature <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>  moment_feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cx"</span>, <span class="st">"cy"</span><span class="op">)</span>,</span>
<span>  img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">cells</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 44 3739 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(44): Au Background ... Ta Vimentin</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(5): imageID object_id m.cx m.cy objectNum</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="feature-normalisation">Feature normalisation<a class="anchor" aria-label="anchor" href="#feature-normalisation"></a>
</h3>
<div class="section level4">
<h4 id="are-the-intensities-of-markers-consistent-across-images">Are the intensities of markers consistent across images?<a class="anchor" aria-label="anchor" href="#are-the-intensities-of-markers-consistent-across-images"></a>
</h4>
<p>Now we know where cells are, that is their x-y coordinate, and the
average expression of various markers within each cell.</p>
<p>Before moving onto annotating the cell types, we should check if the
marker intensities of each cell require any transformation or
normalisation. Here we examine the distribution of CD45 in each cell
across all the images, which is a general marker for immune cells. Below
we can see that the intensity of CD45 looks very skewed.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Joining the marker information with the cell information</span></span>
<span><span class="va">proteinIntensities</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"intensities"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the density of CD45 across all images.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">proteinIntensities</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD45</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewingProteinDensity-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="how-can-i-transform-and-normalise-my-data">How can I transform and normalise my data?<a class="anchor" aria-label="anchor" href="#how-can-i-transform-and-normalise-my-data"></a>
</h4>
<p>We can transform and normalise our data using the
<code>normalizeCells</code> function in <code>simpleSeg</code>. Here we
have taken the intensities from the <code>intensities</code> assay,
performed a asinh transform, then for each image trimmed the 99 quantile
and min-max scaled to 0-1. In addition to this we have performed
principle component analysis, and regressed out the first PC. This
modified data is then stored in the <code>normIntensities</code> assay.
We can see that the distribution of CD45 looks much less skewed than
before.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/normalizeCells.html" class="external-link">normalizeCells</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  transformation <span class="op">=</span> <span class="st">"asinh"</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"minMax"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>  assayIn <span class="op">=</span> <span class="st">"intensities"</span>,</span>
<span>  assayOut <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">normProteinIntensities</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"normIntensities"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">normProteinIntensities</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/kerenNormalise-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>CD3 is a marker which characterises T cells, plot the distribution
of CD3. What does it look like? What does this tell us about T cells in
our dataset.</li>
</ol>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Answer question here</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h3>
<div class="section level4">
<h4 id="can-we-use-a-clustering-approach-to-identify-cell-types">Can we use a clustering approach to identify cell types?<a class="anchor" aria-label="anchor" href="#can-we-use-a-clustering-approach-to-identify-cell-types"></a>
</h4>
<div class="section level5">
<h5 id="clustering-cell-types">Clustering cell types<a class="anchor" aria-label="anchor" href="#clustering-cell-types"></a>
</h5>
<p>Here we cluster using the <code>runFuseSOM</code> function from
<code>FuseSOM</code>. We have chosen to specify the same subset of
markers used in the original manuscript for gating cell types. We have
also specified the number of clusters to identify to be
<code>numClusters = 17</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">useMarkers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45"</span>, <span class="st">"FoxP3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"CD3"</span>, <span class="st">"CD20"</span>, <span class="st">"CD16"</span>, <span class="st">"CD68"</span>, <span class="st">"MPO"</span>, <span class="st">"HLA.DR"</span>, <span class="st">"Pan.Keratin"</span>, <span class="st">"Keratin17"</span>, <span class="st">"Keratin6"</span>, <span class="st">"p53"</span>, <span class="st">"Beta.catenin"</span>, <span class="st">"EGFR"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html" class="external-link">runFuseSOM</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  numClusters <span class="op">=</span> <span class="fl">17</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="how-do-i-interpret-my-clusters">How do I interpret my clusters?<a class="anchor" aria-label="anchor" href="#how-do-i-interpret-my-clusters"></a>
</h5>
<p>We can begin the process of understanding what each of these cell
clusters are by using the <code>plotGroupedHeatmap</code> function from
<code>scater</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  features <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/heatmap-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Have we captured distinct cell populations? Can you identify any of
the clusters?</li>
<li>Run the code below to compare the cell types in the dataset to the
clusters we have identified. How do our clusters compare? Are there any
clusters that should be combined?</li>
</ol>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Type your answer here</span></span>
<span></span>
<span></span>
<span><span class="co"># Question 2 code</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">kerenSPE</span>, cellType <span class="op">=</span> <span class="st">"clusters"</span>, region <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell types in data"</span>, y <span class="op">=</span> <span class="st">"Our clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/clusteringQ1-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="what-is-an-appropriate-number-of-clusters">What is an appropriate number of clusters?<a class="anchor" aria-label="anchor" href="#what-is-an-appropriate-number-of-clusters"></a>
</h5>
<p>Great question! There is no right answer here, however there are
several statistics which can be used to estimate the “true” number of
clusters. We can use the <code>estimateNumCluster</code> and
<code>optiPlot</code> functions from <code>FuseSOM</code> to examine if
our choice of 17 clusters is reasonable. Here we use the gap statistic,
other methods such as jump, slope, silhouette and within cluster
distance (wcd) are also available.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html" class="external-link">estimateNumCluster</a></span><span class="op">(</span><span class="va">kerenSPE</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html" class="external-link">optiPlot</a></span><span class="op">(</span><span class="va">kerenSPE</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/optiPlot-1.png" width="700"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">clusterEstimation</span><span class="op">$</span><span class="va">Discriminant</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 10</span></span></code></pre>
<p>Finally we can run a UMAP to examine how distinct our clusters are
from one another.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMP.</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"normIntensities"</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"normUMAP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># UMAP by cell type cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/normUMAP-1.png" width="700"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How does this UMAP compare to the original UMAP.</li>
</ol>
</div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Answer question here.</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="what-if-i-dont-have-an-expert-to-annotate-my-cell-types">What if I don’t have an expert to annotate my cell types?<a class="anchor" aria-label="anchor" href="#what-if-i-dont-have-an-expert-to-annotate-my-cell-types"></a>
</h4>
<p>Cell type clustering may difficult without the aid of domain experts
who know how many cell types to expect and are able to annotate the
clusters. Cell type classification is an alternative approach to
clustering which annotates cell types based on an expert labeled
reference dataset.</p>
<p>To do so we will use <code>scClassify</code>. Below we will use the
intensities an cell type annotations from the <code>kerenSPE</code>
dataset to predict the cell types in image 5.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Splitting data to train and test sets</span></span>
<span><span class="va">kerenSPE_train</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">!=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE_test</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="fl">5</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Converting intensities to dgCMatrix</span></span>
<span><span class="va">train_intensities</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kerenSPE_train</span>, <span class="st">"intensities"</span><span class="op">)</span> <span class="op">|&gt;</span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span> <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">test_intensities</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kerenSPE_test</span>, <span class="st">"intensities"</span><span class="op">)</span> <span class="op">|&gt;</span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span> <span class="st">"dgCMatrix"</span><span class="op">)</span></span></code></pre></div>
<p><code>scClassify</code> first constructs a cell type tree from the
training dataset, where cells are organised in a hierarchy with
increasingly fined tuned annotations, here we use the
<code>HOPACH</code> method to construct this tree. This tree is
constructed using the proteins which are differentially expressed
between one cell type to all the other cells, identified using
<code>limma</code>. To annotate the cells in the unlabeled dataset, the
unlabeled cells are projected into a lower dimensional space and a
weighted KNN (<code>WKNN</code>) is used to classify unlabeled cells to
their closest neighbours using <code>pearson</code> correlation as the
distance metric.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scClassifyResults</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/scClassify.html" class="external-link">scClassify</a></span><span class="op">(</span>exprsMat_train <span class="op">=</span> <span class="va">train_intensities</span>, </span>
<span>                            cellTypes_train <span class="op">=</span> <span class="va">kerenSPE_train</span><span class="op">$</span><span class="va">cellType</span>,</span>
<span>                            exprsMat_test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>keren5 <span class="op">=</span> <span class="va">test_intensities</span><span class="op">)</span>,</span>
<span>                            tree <span class="op">=</span> <span class="st">"HOPACH"</span>,</span>
<span>                            algorithm <span class="op">=</span> <span class="st">"WKNN"</span>,</span>
<span>                            selectFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span>,</span>
<span>                            similarity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>                            returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code></pre></div>
<p>To save time we have preloaded in the <code>scClassify</code>
results, below we can view the hierarchical tree constructed by
<code>scClassify</code>.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"scClassifyResults.rda"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/plotCellTypeTree.html" class="external-link">plotCellTypeTree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/cellTypeTree.html" class="external-link">cellTypeTree</a></span><span class="op">(</span><span class="va">scClassifyResults</span><span class="op">$</span><span class="va">trainRes</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/loadscClassify-1.png" width="700"></p>
<p>Having a look at some of the predictions, we can observe some cells
belonging to multiple cell types. These represent cells which are
classified as parent populations in the hierarchical tree. Here we clean
up the names to make plotting easier. We can compare our original labels
to our classified labels to get a sense of how concordant our
predictions are. As you can see, around 68% of the cell types have
identical labels between the original and classified cell types.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classifiedCellTypes</span> <span class="op">=</span> <span class="va">scClassifyResults</span><span class="op">$</span><span class="va">testRes</span><span class="op">$</span><span class="va">keren5</span><span class="op">$</span><span class="va">pearson_WKNN_limma</span><span class="op">$</span><span class="va">predRes</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">classifiedCellTypes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; classifiedCellTypes</span></span>
<span><span class="co">#&gt;                                                                                           B_cell </span></span>
<span><span class="co">#&gt;                                                                                                2 </span></span>
<span><span class="co">#&gt;                                                                                       CD4_T_cell </span></span>
<span><span class="co">#&gt;                                                                                              209 </span></span>
<span><span class="co">#&gt;                                                                                       CD8_T_cell </span></span>
<span><span class="co">#&gt;                                                                                              440 </span></span>
<span><span class="co">#&gt;                                                                                               DC </span></span>
<span><span class="co">#&gt;                                                                                                1 </span></span>
<span><span class="co">#&gt;                                                                                       DC_or_Mono </span></span>
<span><span class="co">#&gt;                                                                                              217 </span></span>
<span><span class="co">#&gt;                                                               DC_or_Mono_Macrophages_Mono_or_Neu </span></span>
<span><span class="co">#&gt;                                                                                              122 </span></span>
<span><span class="co">#&gt;                                                                                         dn_T_CD3 </span></span>
<span><span class="co">#&gt;                                                                                               36 </span></span>
<span><span class="co">#&gt;                                                            dn_T_CD3_B_cell_CD4_T_cell_CD8_T_cell </span></span>
<span><span class="co">#&gt;                                                                                                1 </span></span>
<span><span class="co">#&gt; dn_T_CD3_B_cell_CD4_T_cell_DC_or_Mono_Macrophages_CD8_T_cell_Mono_or_Neu_Neutrophils_NK_DC_Tregs </span></span>
<span><span class="co">#&gt;                                                                                              194 </span></span>
<span><span class="co">#&gt;                                                                   dn_T_CD3_CD4_T_cell_CD8_T_cell </span></span>
<span><span class="co">#&gt;                                                                                               70 </span></span>
<span><span class="co">#&gt;                                                                                      Endothelial </span></span>
<span><span class="co">#&gt;                                                                                              102 </span></span>
<span><span class="co">#&gt;                                                                          Endothelial_Mesenchymal </span></span>
<span><span class="co">#&gt;                                                                                               11 </span></span>
<span><span class="co">#&gt;                                                                                   Keratin_Tumour </span></span>
<span><span class="co">#&gt;                                                                                             1901 </span></span>
<span><span class="co">#&gt;                          Keratin_Tumour_Unidentified_Other_Immune_Endothelial_Mesenchymal_Tumour </span></span>
<span><span class="co">#&gt;                                                                                              417 </span></span>
<span><span class="co">#&gt;                                                                                      Macrophages </span></span>
<span><span class="co">#&gt;                                                                                              857 </span></span>
<span><span class="co">#&gt;                                                                          Macrophages_Mono_or_Neu </span></span>
<span><span class="co">#&gt;                                                                                              130 </span></span>
<span><span class="co">#&gt;                                                                                      Mesenchymal </span></span>
<span><span class="co">#&gt;                                                                                               76 </span></span>
<span><span class="co">#&gt;                                                                                      Mono_or_Neu </span></span>
<span><span class="co">#&gt;                                                                                               35 </span></span>
<span><span class="co">#&gt;                                                                                      Neutrophils </span></span>
<span><span class="co">#&gt;                                                                                               65 </span></span>
<span><span class="co">#&gt;                                                                                     Other_Immune </span></span>
<span><span class="co">#&gt;                                                                                               75 </span></span>
<span><span class="co">#&gt;                                                             Other_Immune_Endothelial_Mesenchymal </span></span>
<span><span class="co">#&gt;                                                                                               55 </span></span>
<span><span class="co">#&gt;                                                                                            Tregs </span></span>
<span><span class="co">#&gt;                                                                                               31 </span></span>
<span><span class="co">#&gt;                                                                                           Tumour </span></span>
<span><span class="co">#&gt;                                                                                               16 </span></span>
<span><span class="co">#&gt;                                                                                       unassigned </span></span>
<span><span class="co">#&gt;                                                                                              335 </span></span>
<span><span class="co">#&gt;                                                                                     Unidentified </span></span>
<span><span class="co">#&gt;                                                                                                4 </span></span>
<span><span class="co">#&gt;                                                                              Unidentified_Tumour </span></span>
<span><span class="co">#&gt;                                                                                                4</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newCellTypes</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"Keratin_Tumour_Unidentified_Other_Immune_Endothelial_Mesenchymal_Tumour"</span> <span class="op">~</span> <span class="st">"Tumour_parent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"Other_Immune_Endothelial_Mesenchymal"</span> <span class="op">~</span> <span class="st">"Structural_parent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"Macrophages_Mono_or_Neu"</span> <span class="op">~</span> <span class="st">"Monocyte_subparent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"DC_or_Mono_Macrophages_Mono_or_Neu"</span> <span class="op">~</span> <span class="st">"Monocyte_parent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"dn_T_CD3_B_cell_CD4_T_cell_DC_or_Mono_Macrophages_CD8_T_cell_Mono_or_Neu_Neutrophils_NK_DC_Tregs"</span> <span class="op">~</span> <span class="st">"Immune_parent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"dn_T_CD3_CD4_T_cell_CD8_T_cell"</span> <span class="op">~</span> <span class="st">"Tcell_parent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"Endothelial_Mesenchymal"</span> <span class="op">~</span> <span class="st">"Structural_parent"</span>,</span>
<span>  <span class="va">classifiedCellTypes</span> <span class="op">==</span> <span class="st">"dn_T_CD3_B_cell_CD4_T_cell_CD8_T_cell"</span> <span class="op">~</span> <span class="st">"Lymphocytes_parent"</span>,</span>
<span>  <span class="cn">TRUE</span> <span class="op">~</span> <span class="va">classifiedCellTypes</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE_test</span><span class="op">$</span><span class="va">classifiedCellTypes</span> <span class="op">=</span> <span class="va">newCellTypes</span></span>
<span></span>
<span></span>
<span><span class="op">(</span><span class="va">kerenSPE_test</span><span class="op">$</span><span class="va">cellType</span> <span class="op">==</span> <span class="va">kerenSPE_test</span><span class="op">$</span><span class="va">classifiedCellTypes</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 0.6877543</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="viewing-images">Viewing images<a class="anchor" aria-label="anchor" href="#viewing-images"></a>
</h4>
<p>We can look at the distribution of cells in an image. Here we compare
our clusters and classified cell types to the cell types in the
dataset.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Our clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-1.png" width="700"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE_test</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE_test</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE_test</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"classifiedCellTypes"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-2.png" width="700"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Original cell types"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-3.png" width="700"></p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Comparing orignal cell types to clustering and classification cell types. </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">kerenSPE</span>, cellType <span class="op">=</span> <span class="st">"clusters"</span>, region <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Original vs Clustered"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-4.png" width="700"></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">kerenSPE_test</span>, cellType <span class="op">=</span> <span class="st">"classifiedCellTypes"</span>, region <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Original vs Classified"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/viewClustering-5.png" width="700">
::: question <strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Discuss amongst your peers, why might there be differences between
the clusters, the classified cell types and the original cell type
labels in the dataset? :::</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="analytical-techniques">Analytical techniques<a class="anchor" aria-label="anchor" href="#analytical-techniques"></a>
</h3>
<div class="section level4">
<h4 id="schot-analysis-of-the-developing-brain">scHOT analysis of the developing brain<a class="anchor" aria-label="anchor" href="#schot-analysis-of-the-developing-brain"></a>
</h4>
<p>Here we will ask which gene patterns we observe to be changing across
the <code>spe$gutRegion</code> cell type in space. Note that we want to
assess the anatomical region corresponding to the anterior end of the
developing gut developing brain so we will first subset the cells using
the spatial coordinates. We can check what we have selected by
plotting.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter cells which are labeled gut tub, and in the anterior section.</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">gutRegion</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">$</span><span class="va">celltype_mapped_refined</span> <span class="op">==</span> <span class="st">"Gut tube"</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.5</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"gutRegion"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/visualise_spe-1.png" width="700"></p>
<p>Let’s subset the data to only these cells and continue with our scHOT
analysis.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_gut</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>,<span class="va">spe</span><span class="op">$</span><span class="va">gutRegion</span><span class="op">]</span></span>
<span><span class="va">spe_gut</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 351 472 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts molecules logcounts</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(472): embryo1_Pos3_cell377_z2 embryo1_Pos3_cell388_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos27_cell74_z2 embryo1_Pos28_cell373_z2</span></span>
<span><span class="co">#&gt; colData names(16): cell_id embryo ... sizeFactor gutRegion</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre>
<p>We select genes with at least some proportion of expressed cells for
testing, and create the <code>scHOT</code> object.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_gut</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span>, <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/scHOT-init-1.png" width="700"></p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_to_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe_gut</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe_gut</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gene_to_test</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 165</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_to_test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">gene_to_test</span>, <span class="fl">1</span>, <span class="va">paste0</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_to_test</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;         [,1]     </span></span>
<span><span class="co">#&gt; Acvr1   "Acvr1"  </span></span>
<span><span class="co">#&gt; Acvr2a  "Acvr2a" </span></span>
<span><span class="co">#&gt; Ahnak   "Ahnak"  </span></span>
<span><span class="co">#&gt; Akr1c19 "Akr1c19"</span></span>
<span><span class="co">#&gt; Aldh1a2 "Aldh1a2"</span></span>
<span><span class="co">#&gt; Aldh2   "Aldh2"</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_buildFromSCE.html" class="external-link">scHOT_buildFromSCE</a></span><span class="op">(</span><span class="va">spe_gut</span>,</span>
<span>                                    assayName <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                                    positionType <span class="op">=</span> <span class="st">"spatial"</span>,</span>
<span>                                    positionColData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x_global_affine"</span>, <span class="st">"y_global_affine"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scHOT_spatial</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; class: scHOT </span></span>
<span><span class="co">#&gt; dim: 351 472 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): expression</span></span>
<span><span class="co">#&gt; rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(472): embryo1_Pos3_cell377_z2 embryo1_Pos3_cell388_z2 ...</span></span>
<span><span class="co">#&gt;   embryo1_Pos27_cell74_z2 embryo1_Pos28_cell373_z2</span></span>
<span><span class="co">#&gt; colData names(16): cell_id embryo ... sizeFactor gutRegion</span></span>
<span><span class="co">#&gt; testingScaffold dim: 0 0 </span></span>
<span><span class="co">#&gt; weightMatrix dim: 0 0 </span></span>
<span><span class="co">#&gt; scHOT_output colnames (0):</span></span>
<span><span class="co">#&gt; param names (0):</span></span>
<span><span class="co">#&gt; position type: spatial</span></span></code></pre>
<p>We now add the testing scaffold to the <code>scHOT</code> object, and
set the local weight matrix for testing, with a choice of span of 0.1
(the proportion of cells to weight around each cell). We can speed up
computation by not requiring the weight matrix correspond to every
individual cell, but instead a random selection among all the cells
using the <code>thin</code> function.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_addTestingScaffold.html" class="external-link">scHOT_addTestingScaffold</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="va">gene_to_test</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">testingScaffold</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;         gene_1   </span></span>
<span><span class="co">#&gt; Acvr1   "Acvr1"  </span></span>
<span><span class="co">#&gt; Acvr2a  "Acvr2a" </span></span>
<span><span class="co">#&gt; Ahnak   "Ahnak"  </span></span>
<span><span class="co">#&gt; Akr1c19 "Akr1c19"</span></span>
<span><span class="co">#&gt; Aldh1a2 "Aldh1a2"</span></span>
<span><span class="co">#&gt; Aldh2   "Aldh2"</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_setWeightMatrix.html" class="external-link">scHOT_setWeightMatrix</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, span <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">weightMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/thin.html" class="external-link">thin</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">weightMatrix</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"weightMatrix"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1]  53 472</span></span></code></pre>
<p>For a given cell we can visually examine the local weight given by
the span parameter.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellID</span> <span class="op">=</span> <span class="fl">10</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      W <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"weightMatrix"</span><span class="op">)</span><span class="op">[</span><span class="va">cellID</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x_global_affine</span>, y <span class="op">=</span> <span class="op">-</span><span class="va">y_global_affine</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">W</span>, size <span class="op">=</span> <span class="va">W</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"black"</span>, high <span class="op">=</span> <span class="st">"purple"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html" class="external-link">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">2.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spatial Weight"</span><span class="op">)</span>,</span>
<span>         size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spatial Weight"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Central cell: "</span>, <span class="va">cellID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="cn">NULL</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/scHOT-vis-1.png" width="700"></p>
<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>How will the results change if the span is increased/decreased?</li>
</ol>
</div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Make associated changes to the code to test out the question above.</span></span></code></pre></div>
<p>We set the higher order function as the weighted mean function, and
then calculate the observed higher order test statistics. This may take
around 10 seconds.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_calculateGlobalHigherOrderFunction.html" class="external-link">scHOT_calculateGlobalHigherOrderFunction</a></span><span class="op">(</span></span>
<span>    <span class="va">scHOT_spatial</span>,</span>
<span>    higherOrderFunction <span class="op">=</span> <span class="va">weightedMean</span>,</span>
<span>    higherOrderFunctionType <span class="op">=</span> <span class="st">"weighted"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; DataFrame with 165 rows and 2 columns</span></span>
<span><span class="co">#&gt;              gene_1 globalHigherOrderFunction</span></span>
<span><span class="co">#&gt;         &lt;character&gt;                  &lt;matrix&gt;</span></span>
<span><span class="co">#&gt; Acvr1         Acvr1                  0.216666</span></span>
<span><span class="co">#&gt; Acvr2a       Acvr2a                  0.375776</span></span>
<span><span class="co">#&gt; Ahnak         Ahnak                  0.976418</span></span>
<span><span class="co">#&gt; Akr1c19     Akr1c19                  0.744070</span></span>
<span><span class="co">#&gt; Aldh1a2     Aldh1a2                  0.245981</span></span>
<span><span class="co">#&gt; ...             ...                       ...</span></span>
<span><span class="co">#&gt; Wnt5a         Wnt5a                  0.335820</span></span>
<span><span class="co">#&gt; Wnt5b         Wnt5b                  0.220300</span></span>
<span><span class="co">#&gt; Xist           Xist                  1.162241</span></span>
<span><span class="co">#&gt; Zfp444       Zfp444                  0.744082</span></span>
<span><span class="co">#&gt; Zfp57         Zfp57                  0.595519</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_calculateHigherOrderTestStatistics.html" class="external-link">scHOT_calculateHigherOrderTestStatistics</a></span><span class="op">(</span></span>
<span>    <span class="va">scHOT_spatial</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now we can plot the overall mean versus the scHOT statistic to
observe any relationship. Labels can be interactively visualised using
<code>ggplotly</code>. Some genes may have different distributions so we
turn to permutation testing to assess statistical significance.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">@</span><span class="va">scHOT_output</span><span class="op">)</span>, </span>
<span>           <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">globalHigherOrderFunction</span>, y <span class="op">=</span> <span class="va">higherOrderStatistic</span>, label <span class="op">=</span> <span class="va">gene_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mean across all cells"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"scHOT statistic for local weightedMean"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/scHOT-vis2-1.png" width="700"></p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html" class="external-link">ggplotly</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<div class="plotly html-widget html-fill-item" id="htmlwidget-d10f9e6383278f5ad4f8" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-d10f9e6383278f5ad4f8">{"x":{"data":[{"x":[0.21666552735611611,0.37577624073907062,0.97641826897498996,0.74407017030072786,0.24598113331792176,0.57621688385637004,0.23714015727416829,0.21311146138673961,0.57203612263584536,0.21501613520732421,0.51264320790253159,0.78364174184367774,0.61137793088200598,0.23908786096561196,0.57331002310046431,0.79441922973015022,0.29529448321734508,0.22554474469973529,1.3034118346505226,0.65843819340434984,0.36069675428384212,0.50167467078331296,1.6982331364217238,0.6169600807730351,0.31478264601638573,0.44608821671678767,0.34208291118085427,1.9602204845818796,0.63466372607287114,0.72354119106741821,0.48109400065677832,0.2782337364534731,0.71526027492151278,0.92291201508114784,0.6432469482317259,0.21184593402623622,0.82435111277727369,0.41624493839071014,0.60440035919341417,1.4422476150058909,1.075913614761882,0.20042244573595483,0.27028744249370973,0.37269680194193966,0.74477315065749117,0.7866580914135477,0.22107122298289217,1.2763002780546775,1.1635097126550837,0.4120981727395866,0.25732306656694154,2.0492375226753303,1.3820831684358559,0.29890041870737799,0.38018117956576419,0.20808104326229426,1.0954940816795466,1.9601907765655602,0.2699867563461546,0.66824020210897617,0.32522575230510742,0.58927391005327978,0.58682333818527499,0.29647986944199978,0.19235173147853118,0.50884635454788629,0.24722469804609015,0.76766056339513056,0.55574180080135738,0.24864530189996081,0.25388676965681539,0.62470373987410555,0.21718656392817837,0.3540037183500146,0.58266002463937372,0.7884118671369108,0.57216715914481908,0.55700349034140129,0.58223625859697292,1.6139119096047625,1.0926070950474727,0.25316502418075132,0.24402709287273733,1.2309838380385743,1.4927254753415129,3.1461735946697611,0.54584533525953693,0.58728140567460896,2.7900368714686992,1.121406291147095,0.67470063305624151,0.5950188928353165,1.2933786162089944,0.21372117903023019,0.39291396306236454,0.80984277898690182,0.25062482071371628,0.28616762990566519,3.1714201156286199,0.27637731585371667,0.55632613598057212,0.60108958405887536,0.32671818851712797,0.98607793273384281,0.42119090449842173,0.32815258332994046,1.1472873701564217,0.71181135115499838,0.21488565244537938,0.23273347444793882,0.85701652609813805,0.26772453783279287,0.30990857893632007,0.33575352309027001,0.20264373803268534,0.45420800833416286,0.23978661944540236,0.21479506920752414,2.4988430121508967,0.24600963608570298,0.41973395792336066,0.23305094778343419,0.41605637457964789,0.27926233014940444,0.4537272391422531,0.31457161989425569,0.22211894233092785,1.1842563007397477,1.0084242460794188,0.75690504423769356,0.3426839327056237,0.23998577481502317,1.6203129561432765,1.1630577214433306,0.43160490248648653,0.4237788706716783,0.28103994379488711,0.7860524133034128,0.36636472198663073,0.23885545135155606,1.414629844358563,2.1829706132926536,0.54676314117002078,0.47472551591541745,0.47808969273253948,0.58594082534803971,0.87386089471660489,0.49776455063325081,0.20957263415951197,1.6886508256146189,2.3593471668159869,0.32030042961685501,0.2611588156308331,1.2984821828822615,0.50716608578160127,0.29612081131710122,1.30846630033652,0.27344062006270492,0.31903210675918403,0.60523576725855821,0.33581994971876239,0.22030035859149219,1.1622405164885681,0.74408203349418633,0.59551893573281622],"y":[0.075095380214497021,0.066514261705854574,0.33198974116788166,0.16733424713635042,0.18278360718917858,0.047671537659420338,0.07186424187938692,0.072962447674642716,0.073181758790845611,0.075120169081686533,0.30177676403587872,0.1360275037983418,0.49699891953758263,0.056942698676209873,0.49364457413495111,0.41880335773158406,0.20077560847361331,0.11467506362167465,0.40275597064761215,0.35314778827094068,0.19625521212022629,0.29492911074680683,0.45514740322789654,0.4539109990827444,0.15886662220312905,0.31463682633120993,0.096375648802461453,0.28731276477496509,0.45716024335599253,0.37270448722092175,0.14208552431771437,0.19677325129049691,0.40613789875267448,0.25051524280204962,0.19710074006945463,0.049166249684574538,0.45953594375934403,0.092989995065445849,0.14184919566747706,0.092563539791341451,0.22640012377923538,0.071806531265627305,0.074721694907642425,0.067030775412973254,0.14743939925476779,0.1734723099137484,0.10109092167971297,0.18179746784504919,0.18523524547279363,0.1175773883248421,0.16478510133758065,0.67492003058760197,0.366516341294432,0.10155142974125538,0.19711678094904198,0.10250082750830299,0.18832169593096001,0.18886922151675242,0.19659512793646761,0.28064129449400821,0.23524277936887625,0.48264343713798979,0.57536942841077399,0.12822926457478565,0.08358813994269669,0.45639989147226695,0.053989331141883021,0.55655937606082417,0.4822046668653534,0.058089447246819546,0.13102669292250885,0.12022546604755678,0.10148233737483212,0.27844114183633184,0.34495831689211026,0.28451211012487748,0.29033668050777695,0.24716736575919387,0.25272480409698705,0.41528830442036102,0.26209008279250512,0.045907743179835855,0.09085191586359348,0.19727984475908225,0.16053793632899108,0.28946312010240488,0.17750889579707813,0.24099351762276219,0.36913043172168658,0.33793229874715225,0.13422303963379703,0.35986322791147862,0.25160004707783945,0.072702929914132353,0.066365839569215351,0.16501604172984119,0.11689242051874478,0.11761442755781423,0.31131567935608212,0.093249754746791064,0.65257377436243924,0.14392269285478268,0.13614020147214193,0.69828735622624694,0.49888120969141392,0.093675322304645522,0.16645054123515643,0.10964026189916989,0.083969838940518488,0.053648789168821397,0.27369883421291136,0.077623141162213583,0.12466191712904855,0.29624618414932713,0.073048534054927158,0.56965653768002134,0.054245878043803236,0.058404339597059228,0.40998662475795533,0.16397672470283373,0.22379415253303789,0.083410448705932005,0.17780335910001122,0.10058222373924668,0.099150618043280792,0.13081957061787233,0.18269142698974802,0.12340876027492466,0.16965359159267754,0.092757739866083044,0.14286592514491342,0.091742916407448818,0.67710108943687441,0.75174986496642204,0.34742425041077118,0.10223663225145126,0.10097882441763696,0.34652476752963296,0.076266557294053891,0.11384079137320724,0.39927390801112272,0.41994711407553248,0.42476491099519398,0.1063925332329429,0.56488051773448922,0.34040256111296369,0.46242377876376256,0.45279942151473102,0.083654718276944068,0.14457351633888255,0.12042511501795565,0.051474315762060699,0.062109339514151464,0.095456549955225886,0.26578096419295338,0.11300255828680771,0.26947465570342066,0.12077106508792358,0.067784020669587802,0.094066179765152511,0.17229957637974883,0.10492409670964814,0.12082808057197614,0.1189296002916481,0.13012771733103512],"text":["globalHigherOrderFunction: 0.2166655<br />higherOrderStatistic: 0.07509538<br />gene_1: Acvr1","globalHigherOrderFunction: 0.3757762<br />higherOrderStatistic: 0.06651426<br />gene_1: Acvr2a","globalHigherOrderFunction: 0.9764183<br />higherOrderStatistic: 0.33198974<br />gene_1: Ahnak","globalHigherOrderFunction: 0.7440702<br />higherOrderStatistic: 0.16733425<br />gene_1: Akr1c19","globalHigherOrderFunction: 0.2459811<br />higherOrderStatistic: 0.18278361<br />gene_1: Aldh1a2","globalHigherOrderFunction: 0.5762169<br />higherOrderStatistic: 0.04767154<br />gene_1: Aldh2","globalHigherOrderFunction: 0.2371402<br />higherOrderStatistic: 0.07186424<br />gene_1: Apln","globalHigherOrderFunction: 0.2131115<br />higherOrderStatistic: 0.07296245<br />gene_1: Apoa4","globalHigherOrderFunction: 0.5720361<br />higherOrderStatistic: 0.07318176<br />gene_1: Ash2l","globalHigherOrderFunction: 0.2150161<br />higherOrderStatistic: 0.07512017<br />gene_1: Atp1b1","globalHigherOrderFunction: 0.5126432<br />higherOrderStatistic: 0.30177676<br />gene_1: Axin2","globalHigherOrderFunction: 0.7836417<br />higherOrderStatistic: 0.13602750<br />gene_1: Bak1","globalHigherOrderFunction: 0.6113779<br />higherOrderStatistic: 0.49699892<br />gene_1: Bambi","globalHigherOrderFunction: 0.2390879<br />higherOrderStatistic: 0.05694270<br />gene_1: Bid","globalHigherOrderFunction: 0.5733100<br />higherOrderStatistic: 0.49364457<br />gene_1: Bmp4","globalHigherOrderFunction: 0.7944192<br />higherOrderStatistic: 0.41880336<br />gene_1: Bmp7","globalHigherOrderFunction: 0.2952945<br />higherOrderStatistic: 0.20077561<br />gene_1: Car7","globalHigherOrderFunction: 0.2255447<br />higherOrderStatistic: 0.11467506<br />gene_1: Cavin3","globalHigherOrderFunction: 1.3034118<br />higherOrderStatistic: 0.40275597<br />gene_1: Cdh1","globalHigherOrderFunction: 0.6584382<br />higherOrderStatistic: 0.35314779<br />gene_1: Cdh2","globalHigherOrderFunction: 0.3606968<br />higherOrderStatistic: 0.19625521<br />gene_1: Cers4","globalHigherOrderFunction: 0.5016747<br />higherOrderStatistic: 0.29492911<br />gene_1: Chrd","globalHigherOrderFunction: 1.6982331<br />higherOrderStatistic: 0.45514740<br />gene_1: Cldn4","globalHigherOrderFunction: 0.6169601<br />higherOrderStatistic: 0.45391100<br />gene_1: Clic6","globalHigherOrderFunction: 0.3147826<br />higherOrderStatistic: 0.15886662<br />gene_1: Cntfr","globalHigherOrderFunction: 0.4460882<br />higherOrderStatistic: 0.31463683<br />gene_1: Col1a2","globalHigherOrderFunction: 0.3420829<br />higherOrderStatistic: 0.09637565<br />gene_1: Col26a1","globalHigherOrderFunction: 1.9602205<br />higherOrderStatistic: 0.28731276<br />gene_1: Col4a1","globalHigherOrderFunction: 0.6346637<br />higherOrderStatistic: 0.45716024<br />gene_1: Cpm","globalHigherOrderFunction: 0.7235412<br />higherOrderStatistic: 0.37270449<br />gene_1: Cpn1","globalHigherOrderFunction: 0.4810940<br />higherOrderStatistic: 0.14208552<br />gene_1: Cxcl12","globalHigherOrderFunction: 0.2782337<br />higherOrderStatistic: 0.19677325<br />gene_1: Cyp26a1","globalHigherOrderFunction: 0.7152603<br />higherOrderStatistic: 0.40613790<br />gene_1: Dlk1","globalHigherOrderFunction: 0.9229120<br />higherOrderStatistic: 0.25051524<br />gene_1: Dnmt3a","globalHigherOrderFunction: 0.6432469<br />higherOrderStatistic: 0.19710074<br />gene_1: Dnmt3b","globalHigherOrderFunction: 0.2118459<br />higherOrderStatistic: 0.04916625<br />gene_1: Dppa2","globalHigherOrderFunction: 0.8243511<br />higherOrderStatistic: 0.45953594<br />gene_1: Dusp6","globalHigherOrderFunction: 0.4162449<br />higherOrderStatistic: 0.09299000<br />gene_1: Eed","globalHigherOrderFunction: 0.6044004<br />higherOrderStatistic: 0.14184920<br />gene_1: Efna5","globalHigherOrderFunction: 1.4422476<br />higherOrderStatistic: 0.09256354<br />gene_1: Ep300","globalHigherOrderFunction: 1.0759136<br />higherOrderStatistic: 0.22640012<br />gene_1: Epcam","globalHigherOrderFunction: 0.2004224<br />higherOrderStatistic: 0.07180653<br />gene_1: Esam","globalHigherOrderFunction: 0.2702874<br />higherOrderStatistic: 0.07472169<br />gene_1: Ets1","globalHigherOrderFunction: 0.3726968<br />higherOrderStatistic: 0.06703078<br />gene_1: Ezh1","globalHigherOrderFunction: 0.7447732<br />higherOrderStatistic: 0.14743940<br />gene_1: Ezh2","globalHigherOrderFunction: 0.7866581<br />higherOrderStatistic: 0.17347231<br />gene_1: F2r","globalHigherOrderFunction: 0.2210712<br />higherOrderStatistic: 0.10109092<br />gene_1: Fam181b","globalHigherOrderFunction: 1.2763003<br />higherOrderStatistic: 0.18179747<br />gene_1: Fgfr1","globalHigherOrderFunction: 1.1635097<br />higherOrderStatistic: 0.18523525<br />gene_1: Fgfr2","globalHigherOrderFunction: 0.4120982<br />higherOrderStatistic: 0.11757739<br />gene_1: Fgfr3","globalHigherOrderFunction: 0.2573231<br />higherOrderStatistic: 0.16478510<br />gene_1: Fgfr4","globalHigherOrderFunction: 2.0492375<br />higherOrderStatistic: 0.67492003<br />gene_1: Foxa1","globalHigherOrderFunction: 1.3820832<br />higherOrderStatistic: 0.36651634<br />gene_1: Foxa2","globalHigherOrderFunction: 0.2989004<br />higherOrderStatistic: 0.10155143<br />gene_1: Foxc2","globalHigherOrderFunction: 0.3801812<br />higherOrderStatistic: 0.19711678<br />gene_1: Foxf1","globalHigherOrderFunction: 0.2080810<br />higherOrderStatistic: 0.10250083<br />gene_1: Fst","globalHigherOrderFunction: 1.0954941<br />higherOrderStatistic: 0.18832170<br />gene_1: Furin","globalHigherOrderFunction: 1.9601908<br />higherOrderStatistic: 0.18886922<br />gene_1: Fzd2","globalHigherOrderFunction: 0.2699868<br />higherOrderStatistic: 0.19659513<br />gene_1: Gata2","globalHigherOrderFunction: 0.6682402<br />higherOrderStatistic: 0.28064129<br />gene_1: Gata3","globalHigherOrderFunction: 0.3252258<br />higherOrderStatistic: 0.23524278<br />gene_1: Gata4","globalHigherOrderFunction: 0.5892739<br />higherOrderStatistic: 0.48264344<br />gene_1: Gata5","globalHigherOrderFunction: 0.5868233<br />higherOrderStatistic: 0.57536943<br />gene_1: Gata6","globalHigherOrderFunction: 0.2964799<br />higherOrderStatistic: 0.12822926<br />gene_1: Gmpr","globalHigherOrderFunction: 0.1923517<br />higherOrderStatistic: 0.08358814<br />gene_1: Gng3","globalHigherOrderFunction: 0.5088464<br />higherOrderStatistic: 0.45639989<br />gene_1: Gpc4","globalHigherOrderFunction: 0.2472247<br />higherOrderStatistic: 0.05398933<br />gene_1: Hapln1","globalHigherOrderFunction: 0.7676606<br />higherOrderStatistic: 0.55655938<br />gene_1: Hoxa1","globalHigherOrderFunction: 0.5557418<br />higherOrderStatistic: 0.48220467<br />gene_1: Hoxb1","globalHigherOrderFunction: 0.2486453<br />higherOrderStatistic: 0.05808945<br />gene_1: Hoxb3","globalHigherOrderFunction: 0.2538868<br />higherOrderStatistic: 0.13102669<br />gene_1: Hoxb4","globalHigherOrderFunction: 0.6247037<br />higherOrderStatistic: 0.12022547<br />gene_1: Hsf1","globalHigherOrderFunction: 0.2171866<br />higherOrderStatistic: 0.10148234<br />gene_1: Icam2","globalHigherOrderFunction: 0.3540037<br />higherOrderStatistic: 0.27844114<br />gene_1: Igf1","globalHigherOrderFunction: 0.5826600<br />higherOrderStatistic: 0.34495832<br />gene_1: Igfbp3","globalHigherOrderFunction: 0.7884119<br />higherOrderStatistic: 0.28451211<br />gene_1: Irx1","globalHigherOrderFunction: 0.5721672<br />higherOrderStatistic: 0.29033668<br />gene_1: Irx2","globalHigherOrderFunction: 0.5570035<br />higherOrderStatistic: 0.24716737<br />gene_1: Irx3","globalHigherOrderFunction: 0.5822363<br />higherOrderStatistic: 0.25272480<br />gene_1: Isl1","globalHigherOrderFunction: 1.6139119<br />higherOrderStatistic: 0.41528830<br />gene_1: Itga3","globalHigherOrderFunction: 1.0926071<br />higherOrderStatistic: 0.26209008<br />gene_1: Jarid2","globalHigherOrderFunction: 0.2531650<br />higherOrderStatistic: 0.04590774<br />gene_1: Kcng1","globalHigherOrderFunction: 0.2440271<br />higherOrderStatistic: 0.09085192<br />gene_1: Kitl","globalHigherOrderFunction: 1.2309838<br />higherOrderStatistic: 0.19727984<br />gene_1: Kmt2b","globalHigherOrderFunction: 1.4927255<br />higherOrderStatistic: 0.16053794<br />gene_1: Kmt2d","globalHigherOrderFunction: 3.1461736<br />higherOrderStatistic: 0.28946312<br />gene_1: Krt18","globalHigherOrderFunction: 0.5458453<br />higherOrderStatistic: 0.17750890<br />gene_1: Ldhb","globalHigherOrderFunction: 0.5872814<br />higherOrderStatistic: 0.24099352<br />gene_1: Lef1","globalHigherOrderFunction: 2.7900369<br />higherOrderStatistic: 0.36913043<br />gene_1: Lin28a","globalHigherOrderFunction: 1.1214063<br />higherOrderStatistic: 0.33793230<br />gene_1: Marcks","globalHigherOrderFunction: 0.6747006<br />higherOrderStatistic: 0.13422304<br />gene_1: Mcl1","globalHigherOrderFunction: 0.5950189<br />higherOrderStatistic: 0.35986323<br />gene_1: Meis1","globalHigherOrderFunction: 1.2933786<br />higherOrderStatistic: 0.25160005<br />gene_1: Meis2","globalHigherOrderFunction: 0.2137212<br />higherOrderStatistic: 0.07270293<br />gene_1: Meox1","globalHigherOrderFunction: 0.3929140<br />higherOrderStatistic: 0.06636584<br />gene_1: Mkrn1","globalHigherOrderFunction: 0.8098428<br />higherOrderStatistic: 0.16501604<br />gene_1: Mnt","globalHigherOrderFunction: 0.2506248<br />higherOrderStatistic: 0.11689242<br />gene_1: Msx1","globalHigherOrderFunction: 0.2861676<br />higherOrderStatistic: 0.11761443<br />gene_1: Msx2","globalHigherOrderFunction: 3.1714201<br />higherOrderStatistic: 0.31131568<br />gene_1: Myh9","globalHigherOrderFunction: 0.2763773<br />higherOrderStatistic: 0.09324975<br />gene_1: Nid1","globalHigherOrderFunction: 0.5563261<br />higherOrderStatistic: 0.65257377<br />gene_1: Nkx2-3","globalHigherOrderFunction: 0.6010896<br />higherOrderStatistic: 0.14392269<br />gene_1: Nr2f1","globalHigherOrderFunction: 0.3267182<br />higherOrderStatistic: 0.13614020<br />gene_1: Nrp1","globalHigherOrderFunction: 0.9860779<br />higherOrderStatistic: 0.69828736<br />gene_1: Osr1","globalHigherOrderFunction: 0.4211909<br />higherOrderStatistic: 0.49888121<br />gene_1: Otx2","globalHigherOrderFunction: 0.3281526<br />higherOrderStatistic: 0.09367532<br />gene_1: Ovol2","globalHigherOrderFunction: 1.1472874<br />higherOrderStatistic: 0.16645054<br />gene_1: Pcgf2","globalHigherOrderFunction: 0.7118114<br />higherOrderStatistic: 0.10964026<br />gene_1: Pcgf3","globalHigherOrderFunction: 0.2148857<br />higherOrderStatistic: 0.08396984<br />gene_1: Pcgf5","globalHigherOrderFunction: 0.2327335<br />higherOrderStatistic: 0.05364879<br />gene_1: Pcgf6","globalHigherOrderFunction: 0.8570165<br />higherOrderStatistic: 0.27369883<br />gene_1: Pdgfa","globalHigherOrderFunction: 0.2677245<br />higherOrderStatistic: 0.07762314<br />gene_1: Pdgfra","globalHigherOrderFunction: 0.3099086<br />higherOrderStatistic: 0.12466192<br />gene_1: Per2","globalHigherOrderFunction: 0.3357535<br />higherOrderStatistic: 0.29624618<br />gene_1: Perp","globalHigherOrderFunction: 0.2026437<br />higherOrderStatistic: 0.07304853<br />gene_1: Pim2","globalHigherOrderFunction: 0.4542080<br />higherOrderStatistic: 0.56965654<br />gene_1: Pitx1","globalHigherOrderFunction: 0.2397866<br />higherOrderStatistic: 0.05424588<br />gene_1: Plp1","globalHigherOrderFunction: 0.2147951<br />higherOrderStatistic: 0.05840434<br />gene_1: Plvap","globalHigherOrderFunction: 2.4988430<br />higherOrderStatistic: 0.40998662<br />gene_1: Podxl","globalHigherOrderFunction: 0.2460096<br />higherOrderStatistic: 0.16397672<br />gene_1: Popdc2","globalHigherOrderFunction: 0.4197340<br />higherOrderStatistic: 0.22379415<br />gene_1: Prdm1","globalHigherOrderFunction: 0.2330509<br />higherOrderStatistic: 0.08341045<br />gene_1: Prrx1","globalHigherOrderFunction: 0.4160564<br />higherOrderStatistic: 0.17780336<br />gene_1: Prrx2","globalHigherOrderFunction: 0.2792623<br />higherOrderStatistic: 0.10058222<br />gene_1: Ptn","globalHigherOrderFunction: 0.4537272<br />higherOrderStatistic: 0.09915062<br />gene_1: Ramp2","globalHigherOrderFunction: 0.3145716<br />higherOrderStatistic: 0.13081957<br />gene_1: Rgl1","globalHigherOrderFunction: 0.2221189<br />higherOrderStatistic: 0.18269143<br />gene_1: Runx1","globalHigherOrderFunction: 1.1842563<br />higherOrderStatistic: 0.12340876<br />gene_1: Setd1a","globalHigherOrderFunction: 1.0084242<br />higherOrderStatistic: 0.16965359<br />gene_1: Setd1b","globalHigherOrderFunction: 0.7569050<br />higherOrderStatistic: 0.09275774<br />gene_1: Setd2","globalHigherOrderFunction: 0.3426839<br />higherOrderStatistic: 0.14286593<br />gene_1: Sfrp1","globalHigherOrderFunction: 0.2399858<br />higherOrderStatistic: 0.09174292<br />gene_1: Sfrp2","globalHigherOrderFunction: 1.6203130<br />higherOrderStatistic: 0.67710109<br />gene_1: Shh","globalHigherOrderFunction: 1.1630577<br />higherOrderStatistic: 0.75174986<br />gene_1: Shisa2","globalHigherOrderFunction: 0.4316049<br />higherOrderStatistic: 0.34742425<br />gene_1: Six1","globalHigherOrderFunction: 0.4237789<br />higherOrderStatistic: 0.10223663<br />gene_1: Smarcd3","globalHigherOrderFunction: 0.2810399<br />higherOrderStatistic: 0.10097882<br />gene_1: Smim1","globalHigherOrderFunction: 0.7860524<br />higherOrderStatistic: 0.34652477<br />gene_1: Smoc2","globalHigherOrderFunction: 0.3663647<br />higherOrderStatistic: 0.07626656<br />gene_1: Snai1","globalHigherOrderFunction: 0.2388555<br />higherOrderStatistic: 0.11384079<br />gene_1: Socs2","globalHigherOrderFunction: 1.4146298<br />higherOrderStatistic: 0.39927391<br />gene_1: Sox2","globalHigherOrderFunction: 2.1829706<br />higherOrderStatistic: 0.41994711<br />gene_1: Sox4","globalHigherOrderFunction: 0.5467631<br />higherOrderStatistic: 0.42476491<br />gene_1: Sp5","globalHigherOrderFunction: 0.4747255<br />higherOrderStatistic: 0.10639253<br />gene_1: Suz12","globalHigherOrderFunction: 0.4780897<br />higherOrderStatistic: 0.56488052<br />gene_1: T","globalHigherOrderFunction: 0.5859408<br />higherOrderStatistic: 0.34040256<br />gene_1: Tagln","globalHigherOrderFunction: 0.8738609<br />higherOrderStatistic: 0.46242378<br />gene_1: Tbx1","globalHigherOrderFunction: 0.4977646<br />higherOrderStatistic: 0.45279942<br />gene_1: Tbx3","globalHigherOrderFunction: 0.2095726<br />higherOrderStatistic: 0.08365472<br />gene_1: Tbx4","globalHigherOrderFunction: 1.6886508<br />higherOrderStatistic: 0.14457352<br />gene_1: Tcf7l1","globalHigherOrderFunction: 2.3593472<br />higherOrderStatistic: 0.12042512<br />gene_1: Tead2","globalHigherOrderFunction: 0.3203004<br />higherOrderStatistic: 0.05147432<br />gene_1: Tet1","globalHigherOrderFunction: 0.2611588<br />higherOrderStatistic: 0.06210934<br />gene_1: Tet2","globalHigherOrderFunction: 1.2984822<br />higherOrderStatistic: 0.09545655<br />gene_1: Tet3","globalHigherOrderFunction: 0.5071661<br />higherOrderStatistic: 0.26578096<br />gene_1: Thbs1","globalHigherOrderFunction: 0.2961208<br />higherOrderStatistic: 0.11300256<br />gene_1: Tinagl1","globalHigherOrderFunction: 1.3084663<br />higherOrderStatistic: 0.26947466<br />gene_1: Tjp2","globalHigherOrderFunction: 0.2734406<br />higherOrderStatistic: 0.12077107<br />gene_1: Ttn","globalHigherOrderFunction: 0.3190321<br />higherOrderStatistic: 0.06778402<br />gene_1: Usp28","globalHigherOrderFunction: 0.6052358<br />higherOrderStatistic: 0.09406618<br />gene_1: Wdr5","globalHigherOrderFunction: 0.3358199<br />higherOrderStatistic: 0.17229958<br />gene_1: Wnt5a","globalHigherOrderFunction: 0.2203004<br />higherOrderStatistic: 0.10492410<br />gene_1: Wnt5b","globalHigherOrderFunction: 1.1622405<br />higherOrderStatistic: 0.12082808<br />gene_1: Xist","globalHigherOrderFunction: 0.7440820<br />higherOrderStatistic: 0.11892960<br />gene_1: Zfp444","globalHigherOrderFunction: 0.5955189<br />higherOrderStatistic: 0.13012772<br />gene_1: Zfp57"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.6692913385826778,"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.824636418824401,"r":7.3059360730593621,"b":41.778974318367787,"l":43.105022831050235},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.043398312271026723,3.3203735348361243],"tickmode":"array","ticktext":["1","2","3"],"tickvals":[1,2.0000000000000004,3],"categoryorder":"array","categoryarray":["1","2","3"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176002,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"Mean across all cells","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.010615637090506544,0.78704197105575135],"tickmode":"array","ticktext":["0.2","0.4","0.6"],"tickvals":[0.20000000000000001,0.40000000000000002,0.60000000000000009],"categoryorder":"array","categoryarray":["0.2","0.4","0.6"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176002,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"scHOT statistic for local weightedMean","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"296767065833":{"x":{},"y":{},"label":{},"type":"scatter"}},"cur_data":"296767065833","visdat":{"296767065833":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p>Set up the permutation testing schema. For the purposes of this
workshop we set a low number of permutations over a low number of genes
in the testing scaffold, you may want to change this as you work through
the workshop yourself. The testing will take a few minutes to run, here
with the parallel parameters that were set at the beginning of this
document.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_setPermutationScaffold.html" class="external-link">scHOT_setPermutationScaffold</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>,</span>
<span>                                              numberPermutations <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                                              numberScaffold <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_performPermutationTest.html" class="external-link">scHOT_performPermutationTest</a></span><span class="op">(</span></span>
<span>    <span class="va">scHOT_spatial</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Permutation testing combination 40 of 165...</span></span>
<span><span class="co">#&gt; Permutation testing combination 80 of 165...</span></span>
<span><span class="co">#&gt; Permutation testing combination 110 of 165...</span></span>
<span><span class="co">#&gt; Permutation testing combination 150 of 165...</span></span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; DataFrame with 165 rows and 9 columns</span></span>
<span><span class="co">#&gt;              gene_1 globalHigherOrderFunction            higherOrderSequence</span></span>
<span><span class="co">#&gt;         &lt;character&gt;                  &lt;matrix&gt;                  &lt;NumericList&gt;</span></span>
<span><span class="co">#&gt; Acvr1         Acvr1                  0.216666 0.251205,0.275076,0.286668,...</span></span>
<span><span class="co">#&gt; Acvr2a       Acvr2a                  0.375776 0.398236,0.376223,0.361763,...</span></span>
<span><span class="co">#&gt; Ahnak         Ahnak                  0.976418    1.23931,1.22101,1.19278,...</span></span>
<span><span class="co">#&gt; Akr1c19     Akr1c19                  0.744070 0.681732,0.622183,0.625407,...</span></span>
<span><span class="co">#&gt; Aldh1a2     Aldh1a2                  0.245981 0.117491,0.118105,0.121221,...</span></span>
<span><span class="co">#&gt; ...             ...                       ...                            ...</span></span>
<span><span class="co">#&gt; Wnt5a         Wnt5a                  0.335820 0.282418,0.280240,0.268180,...</span></span>
<span><span class="co">#&gt; Wnt5b         Wnt5b                  0.220300 0.262440,0.321449,0.368172,...</span></span>
<span><span class="co">#&gt; Xist           Xist                  1.162241    1.18893,1.17123,1.18238,...</span></span>
<span><span class="co">#&gt; Zfp444       Zfp444                  0.744082 0.529888,0.531771,0.538540,...</span></span>
<span><span class="co">#&gt; Zfp57         Zfp57                  0.595519 0.853046,0.844188,0.838651,...</span></span>
<span><span class="co">#&gt;         higherOrderStatistic numberPermutations storePermutations</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt;          &lt;numeric&gt;         &lt;logical&gt;</span></span>
<span><span class="co">#&gt; Acvr1              0.0750954                  0              TRUE</span></span>
<span><span class="co">#&gt; Acvr2a             0.0665143                  0              TRUE</span></span>
<span><span class="co">#&gt; Ahnak              0.3319897                  0              TRUE</span></span>
<span><span class="co">#&gt; Akr1c19            0.1673342                  0              TRUE</span></span>
<span><span class="co">#&gt; Aldh1a2            0.1827836                 50              TRUE</span></span>
<span><span class="co">#&gt; ...                      ...                ...               ...</span></span>
<span><span class="co">#&gt; Wnt5a               0.172300                  0              TRUE</span></span>
<span><span class="co">#&gt; Wnt5b               0.104924                  0              TRUE</span></span>
<span><span class="co">#&gt; Xist                0.120828                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp444              0.118930                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp57               0.130128                  0              TRUE</span></span>
<span><span class="co">#&gt;                              permutations pvalPermutations FDRPermutations</span></span>
<span><span class="co">#&gt;                             &lt;NumericList&gt;        &lt;numeric&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Acvr2a                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Ahnak                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Akr1c19                                NA               NA              NA</span></span>
<span><span class="co">#&gt; Aldh1a2 0.0448634,0.0877575,0.0508914,...        0.0196078       0.0244444</span></span>
<span><span class="co">#&gt; ...                                   ...              ...             ...</span></span>
<span><span class="co">#&gt; Wnt5a                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Wnt5b                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Xist                                   NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp444                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp57                                  NA               NA              NA</span></span></code></pre>
<p>After the permutation test we can estimate the P-values across all
genes.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_plotPermutationDistributions.html" class="external-link">scHOT_plotPermutationDistributions</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/scHOT-pval-1.png" width="700"></p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scHOT_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scHOT/man/scHOT_estimatePvalues.html" class="external-link">scHOT_estimatePvalues</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>,</span>
<span>                                       nperm_estimate <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                                       maxDist <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; DataFrame with 165 rows and 14 columns</span></span>
<span><span class="co">#&gt;              gene_1 globalHigherOrderFunction            higherOrderSequence</span></span>
<span><span class="co">#&gt;         &lt;character&gt;                  &lt;matrix&gt;                  &lt;NumericList&gt;</span></span>
<span><span class="co">#&gt; Acvr1         Acvr1                  0.216666 0.251205,0.275076,0.286668,...</span></span>
<span><span class="co">#&gt; Acvr2a       Acvr2a                  0.375776 0.398236,0.376223,0.361763,...</span></span>
<span><span class="co">#&gt; Ahnak         Ahnak                  0.976418    1.23931,1.22101,1.19278,...</span></span>
<span><span class="co">#&gt; Akr1c19     Akr1c19                  0.744070 0.681732,0.622183,0.625407,...</span></span>
<span><span class="co">#&gt; Aldh1a2     Aldh1a2                  0.245981 0.117491,0.118105,0.121221,...</span></span>
<span><span class="co">#&gt; ...             ...                       ...                            ...</span></span>
<span><span class="co">#&gt; Wnt5a         Wnt5a                  0.335820 0.282418,0.280240,0.268180,...</span></span>
<span><span class="co">#&gt; Wnt5b         Wnt5b                  0.220300 0.262440,0.321449,0.368172,...</span></span>
<span><span class="co">#&gt; Xist           Xist                  1.162241    1.18893,1.17123,1.18238,...</span></span>
<span><span class="co">#&gt; Zfp444       Zfp444                  0.744082 0.529888,0.531771,0.538540,...</span></span>
<span><span class="co">#&gt; Zfp57         Zfp57                  0.595519 0.853046,0.844188,0.838651,...</span></span>
<span><span class="co">#&gt;         higherOrderStatistic numberPermutations storePermutations</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt;          &lt;numeric&gt;         &lt;logical&gt;</span></span>
<span><span class="co">#&gt; Acvr1              0.0750954                  0              TRUE</span></span>
<span><span class="co">#&gt; Acvr2a             0.0665143                  0              TRUE</span></span>
<span><span class="co">#&gt; Ahnak              0.3319897                  0              TRUE</span></span>
<span><span class="co">#&gt; Akr1c19            0.1673342                  0              TRUE</span></span>
<span><span class="co">#&gt; Aldh1a2            0.1827836                 50              TRUE</span></span>
<span><span class="co">#&gt; ...                      ...                ...               ...</span></span>
<span><span class="co">#&gt; Wnt5a               0.172300                  0              TRUE</span></span>
<span><span class="co">#&gt; Wnt5b               0.104924                  0              TRUE</span></span>
<span><span class="co">#&gt; Xist                0.120828                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp444              0.118930                  0              TRUE</span></span>
<span><span class="co">#&gt; Zfp57               0.130128                  0              TRUE</span></span>
<span><span class="co">#&gt;                              permutations pvalPermutations FDRPermutations</span></span>
<span><span class="co">#&gt;                             &lt;NumericList&gt;        &lt;numeric&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Acvr2a                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Ahnak                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Akr1c19                                NA               NA              NA</span></span>
<span><span class="co">#&gt; Aldh1a2 0.0448634,0.0877575,0.0508914,...        0.0196078       0.0244444</span></span>
<span><span class="co">#&gt; ...                                   ...              ...             ...</span></span>
<span><span class="co">#&gt; Wnt5a                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Wnt5b                                  NA               NA              NA</span></span>
<span><span class="co">#&gt; Xist                                   NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp444                                 NA               NA              NA</span></span>
<span><span class="co">#&gt; Zfp57                                  NA               NA              NA</span></span>
<span><span class="co">#&gt;         numberPermutationsEstimated globalLowerRangeEstimated</span></span>
<span><span class="co">#&gt;                           &lt;integer&gt;                 &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                           200                  0.225545</span></span>
<span><span class="co">#&gt; Acvr2a                           50                  0.416245</span></span>
<span><span class="co">#&gt; Ahnak                            50                  1.008424</span></span>
<span><span class="co">#&gt; Akr1c19                         100                  0.788412</span></span>
<span><span class="co">#&gt; Aldh1a2                         200                  0.225545</span></span>
<span><span class="co">#&gt; ...                             ...                       ...</span></span>
<span><span class="co">#&gt; Wnt5a                           150                  0.239986</span></span>
<span><span class="co">#&gt; Wnt5b                           200                  0.225545</span></span>
<span><span class="co">#&gt; Xist                            100                  1.147287</span></span>
<span><span class="co">#&gt; Zfp444                          100                  0.788412</span></span>
<span><span class="co">#&gt; Zfp57                          1100                  0.225545</span></span>
<span><span class="co">#&gt;         globalUpperRangeEstimated pvalEstimated FDREstimated</span></span>
<span><span class="co">#&gt;                         &lt;numeric&gt;     &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; Acvr1                    0.245981    0.07500000    0.0937500</span></span>
<span><span class="co">#&gt; Acvr2a                   0.416245    0.54000000    0.5568750</span></span>
<span><span class="co">#&gt; Ahnak                    1.008424    0.01960784    0.0281330</span></span>
<span><span class="co">#&gt; Akr1c19                  0.809843    0.00990099    0.0194118</span></span>
<span><span class="co">#&gt; Aldh1a2                  0.245981    0.00497512    0.0155660</span></span>
<span><span class="co">#&gt; ...                           ...           ...          ...</span></span>
<span><span class="co">#&gt; Wnt5a                    0.416245    0.00662252    0.0182119</span></span>
<span><span class="co">#&gt; Wnt5b                    0.245981    0.00497512    0.0155660</span></span>
<span><span class="co">#&gt; Xist                     1.163510    0.10000000    0.1170213</span></span>
<span><span class="co">#&gt; Zfp444                   0.809843    0.16000000    0.1846154</span></span>
<span><span class="co">#&gt; Zfp57                    3.171420    0.09000000    0.1091912</span></span></code></pre>
<p>We can now examine the spatial expression of the 5 most significant
genes, both in our scHOT object and over our original spe object.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_sorted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"scHOT_output"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>,</span>
<span>                                                                <span class="st">"scHOT_output"</span><span class="op">)</span><span class="op">$</span><span class="va">pvalEstimated</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">topgenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">output_sorted</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scHOT_spatial</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">topgene</span> <span class="kw">in</span> <span class="va">topgenes</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">g_spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">topgene</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">g_scHOT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">scHOT_spatial</span>, <span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">topgene</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                           by_exprs_values <span class="op">=</span> <span class="st">"expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">g_all</span> <span class="op">&lt;-</span> <span class="va">g_scHOT</span> <span class="op">+</span> <span class="va">g_spe</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">g_all</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/scHOT-final-1.png" width="700"><img src="workshop_material_files/figure-html/scHOT-final-2.png" width="700"><img src="workshop_material_files/figure-html/scHOT-final-3.png" width="700"><img src="workshop_material_files/figure-html/scHOT-final-4.png" width="700"><img src="workshop_material_files/figure-html/scHOT-final-5.png" width="700"></p>
<p>Here we are noting the genes that are found to have the most
statistically significant spatial variation in their local mean
expression. These genes point to specific patterns that govern the
development of individual parts of the gut tube.</p>
<div class="question">
<p><strong>Advanced/Extended Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How would you perform such testing over multiple distinct
samples?</li>
<li>scHOT is developed with all higher order testing in mind, use the
associated <a href="http://www.bioconductor.org/packages/release/bioc/html/scHOT.html" class="external-link">vignette</a>
to get towards assessing changes in variation or correlation structure
in space.</li>
</ol>
</div>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## try some code</span></span></code></pre></div>
<p>Now that we have assessed genes varying in expression in space, let’s
look further into the IMC data where we will make use of the multiple
samples to perform clustering and extract some biological
understanding.</p>
</div>
</div>
<div class="section level3">
<h3 id="how-do-i-make-sense-of-and-summarise-spatially-defined-cellular-interactions-that-occur-in-my-data">How do I make sense of and summarise spatially defined cellular
interactions that occur in my data?<a class="anchor" aria-label="anchor" href="#how-do-i-make-sense-of-and-summarise-spatially-defined-cellular-interactions-that-occur-in-my-data"></a>
</h3>
<p>Having now clustered and annotated our data, we now want to see what
biological conclusions and hypotheses we can draw from it.</p>
<p>For this workshop, we will use 2 packages, <code>Statial</code> and
<code>scFeatures</code>. <code>Statial</code> is the latest of scdney
bioconductor packages aimed at addressing biological questions relating
to cellular interactions <em>in-situ</em>. For example, we can answer
questions such as:</p>
<p>- Which cell types localise with each other?</p>
<p>- How does the functionality of a cell affect its localisation with
other cell types?</p>
<p>- How do the phenotypes of cells change during localisation?</p>
<div class="section level4">
<h4 id="how-do-i-determine-which-cell-types-are-interacting-with-each-other-in-my-data">How do I determine which cell types are interacting with each other
in my data?<a class="anchor" aria-label="anchor" href="#how-do-i-determine-which-cell-types-are-interacting-with-each-other-in-my-data"></a>
</h4>
<p><code>Kontextual</code> is a part of the <code>Statial</code> package
which models spatial relationships between cells in the context of
hierarchical cell lineage structures. By assessing spatial relationships
between pairs of cells in the context of other related cell types,
Kontextual provides robust quantification of cell type relationships
which are invariant to changes in tissue structure.</p>
<p>For the purposes of using <code>Kontextual</code> we define cell
functions as identified clusters of cells, where larger clusters
represent a “parent” cell population, and finer sub-clusters
representing a “child” cell population with a particular function. For
example, CD4+ T cells have a highly specified function, and may be
considered a child to a larger parent population of T cells.
<code>Kontextual</code> thus aims to quantify how the localisation
patterns of a child population of cells deviate from the spatial
behaviour of their parent population, and how their cellular function
defines their spatial localisation.</p>
<div class="section level5">
<h5 id="cell-type-hierarchy">Cell type hierarchy<a class="anchor" aria-label="anchor" href="#cell-type-hierarchy"></a>
</h5>
<p>A key input for Kontextual is an annotation of cell type hierarchies.
We will need these to organise all the cells present into cell state
populations or clusters, e.g. all the different B cell types are put in
a vector called bcells.</p>
<p>To make our lives easier, we will start by defining these here. I’m
happy to talk about how we use our bioconductor package <a href="http://www.bioconductor.org/packages/release/bioc/html/treekoR.html" class="external-link">treekoR</a>
to define these hierarchies in a data driven way.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up cell populations</span></span>
<span></span>
<span><span class="va">tumour</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Keratin_Tumour"</span>, <span class="st">"Tumour"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B_cell"</span><span class="op">)</span></span>
<span><span class="va">tcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dn_T_cell"</span>, <span class="st">"CD4_T_cell"</span>, <span class="st">"CD8_T_cell"</span>, <span class="st">"Tregs"</span><span class="op">)</span></span>
<span><span class="va">myeloid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Dc_or_Mono"</span>, <span class="st">"DC"</span>, <span class="st">"Mono_or_Neu"</span>, <span class="st">"Macrophages"</span>, <span class="st">"Other_Immune"</span>, <span class="st">"Neutrophils"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">endothelial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Endothelial"</span><span class="op">)</span></span>
<span><span class="va">mesenchymal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mesenchymal"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">endothelial</span>, <span class="va">mesenchymal</span><span class="op">)</span></span>
<span><span class="va">immune</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bcells</span>, <span class="va">tcells</span>, <span class="va">myeloid</span>, <span class="st">"NK"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">tumour</span>, <span class="va">tissue</span>, <span class="va">immune</span>, <span class="st">"Unidentified"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="do-p53-tumour-cells-exhibit-a-different-spatial-behaviour-compared-to-all-tumour-cells">Do p53+ tumour cells exhibit a different spatial behaviour compared
to all tumour cells?<a class="anchor" aria-label="anchor" href="#do-p53-tumour-cells-exhibit-a-different-spatial-behaviour-compared-to-all-tumour-cells"></a>
</h5>
<p>Here we examine an image highlighted in the Keren et al. 2018
manuscript where the relationship between two cell types depends on a
parent cell population. In image 6 of the Keren et al. dataset, we can
see that <em>p53+ tumour cells</em> and <em>immune cells</em> are
dispersed. However when the behaviour of <em>p53+ tumour cells</em> are
placed in the context of the spatial behaviour of its broader parent
population <em>tumour cells</em>, <em>p53+ tumour cells</em> and
<em>immune</em> would appear localised.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Lets define a new cell type vector</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellTypeNew</span> <span class="op">&lt;-</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellType</span></span>
<span></span>
<span><span class="co"># Select for all cells that express higher than baseline level of p53</span></span>
<span><span class="va">p53Pos</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span><span class="st">"p53"</span>,<span class="op">]</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">0.300460</span></span>
<span></span>
<span><span class="co"># Find p53+ tumour cells</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellTypeNew</span><span class="op">[</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tumour</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Tumour"</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellTypeNew</span><span class="op">[</span><span class="va">p53Pos</span> <span class="op">&amp;</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tumour</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"p53_Tumour"</span></span>
<span></span>
<span><span class="co">#Group all immune cells under the name "Immune"</span></span>
<span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellTypeNew</span><span class="op">[</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">immune</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Immune"</span></span>
<span></span>
<span></span>
<span><span class="co"># Plot image 6</span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"6"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cellTypeNew</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Immune"</span>, <span class="st">"Tumour"</span>, <span class="st">"p53_Tumour"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">cellTypeNew</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">cellTypeNew</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#505050"</span>, <span class="st">"#64BC46"</span>,<span class="st">"#D6D6D6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Cell types"</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/kontext-example-1.png" width="528"></p>
</div>
<div class="section level5">
<h5 id="how-do-we-quantify-this-relationship">How do we quantify this relationship?<a class="anchor" aria-label="anchor" href="#how-do-we-quantify-this-relationship"></a>
</h5>
<p>The <code>Kontextual</code> function accepts a
<code>SingleCellExperiment</code> object, or a single image, or list of
images from a <code>SingleCellExperiment</code> object, this gets passed
into the <code>cells</code> argument. The two cell types which will be
evaluated are specified in the <code>to</code> and <code>from</code>
arguments. A parent population must also be specified in the
<code>parent</code> argument, note the parent cell population must
include the <code>to</code> cell type. The argument <code>r</code> will
specify the radius which the cell relationship will be evaluated on.
<code>Kontextual</code> supports parallel processing, the number of
cores can be specified using the <code>cores</code> argument.
<code>Kontextual</code> can take a single value or multiple values for
each argument and will test all combinations of the arguments
specified.</p>
<p>We can calculate these relationships for a single radius.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p53_Kontextual</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/Kontextual.html" class="external-link">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  image <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"p53_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Immune"</span>,</span>
<span>  parent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"Tumour"</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellTypeNew"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p53_Kontextual</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   imageID               test  original kontextual  r weightQuantile inhom edge</span></span>
<span><span class="co">#&gt; 1       6 p53_Tumour__Immune -20.41372        NaN 50            0.8 FALSE TRUE</span></span>
<span><span class="co">#&gt;   includeZeroCells window window.length</span></span>
<span><span class="co">#&gt; 1             TRUE convex            NA</span></span></code></pre>
</div>
<div class="section level5">
<h5 id="how-do-i-choose-a-radius">How do I choose a radius?<a class="anchor" aria-label="anchor" href="#how-do-i-choose-a-radius"></a>
</h5>
<p>The <code>kontextCurve</code> calculates the L-function value and
Kontextual values over a range of radii. While <code>kontextPlot</code>
plots these values. If the points lie above the red line (expected
pattern) then localisation is indicated for that radius, if the points
lie below the red line then dispersion is indicated. As seen in the
following plot Kontextual is able to correctly identify localisation
between p53+ tumour cells and immune cells in the example image for a
certain range of radii. The original L-function is not able to identify
localisation at any value of radii.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">curves</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/kontextCurve.html" class="external-link">kontextCurve</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"p53_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Immune"</span>,</span>
<span>  parent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53+Tumour"</span>, <span class="st">"Tumour"</span><span class="op">)</span>,</span>
<span>  rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">510</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellTypeNew"</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/kontextPlot.html" class="external-link">kontextPlot</a></span><span class="op">(</span><span class="va">curves</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/kontextCurve-1.png" width="576"></p>
</div>
<div class="section level5">
<h5 id="how-do-i-quantitatively-identify-these-function-dictated-spatial-patterns-across-all-my-images">How do I quantitatively identify these function-dictated spatial
patterns across all my images?<a class="anchor" aria-label="anchor" href="#how-do-i-quantitatively-identify-these-function-dictated-spatial-patterns-across-all-my-images"></a>
</h5>
<p>Alternatively all pairwise cell relationships and their corresponding
parents in the dataset can be tested. A data frame with all pairwise
combinations can be creating using the <code>parentCombinations</code>
function. This function takes in a vector of all the cells, as well as
all the parent vectors set up earlier. As shown below the output is a
data frame specifying the <code>to</code>, <code>from</code>, and
<code>parent</code> arguments for <code>Kontextual</code>.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get all relationships between cell types and their parents</span></span>
<span><span class="va">parentDf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/parentCombinations.html" class="external-link">parentCombinations</a></span><span class="op">(</span></span>
<span>  all <span class="op">=</span> <span class="va">all</span>,</span>
<span>  <span class="va">tumour</span>,</span>
<span>  <span class="va">bcells</span>,</span>
<span>  <span class="va">tcells</span>,</span>
<span>  <span class="va">myeloid</span>,</span>
<span>  <span class="va">endothelial</span>,</span>
<span>  <span class="va">mesenchymal</span>,</span>
<span>  <span class="va">tissue</span>,</span>
<span>  <span class="va">immune</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Rather than specifying <code>to</code>, <code>from</code>, and
<code>parent</code> in Kontextual, the output from
<code>parentCombinations</code> can be input into
<code>Kontextual</code> using the <code>parentDf</code> argument, to
examine all pairwise relationships in the dataset. This chunk will take
a significant amount of time to run (~20 minutes), for demonstration the
results have been saved and are loaded in.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Running Kontextual on all relationships across all images.</span></span>
<span><span class="va">kerenKontextual</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/Kontextual.html" class="external-link">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  parentDf <span class="op">=</span> <span class="va">parentDf</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kerenKontextual.rda"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="are-these-spatial-patterns-associated-with-survival-outcomes">Are these spatial patterns associated with survival outcomes?<a class="anchor" aria-label="anchor" href="#are-these-spatial-patterns-associated-with-survival-outcomes"></a>
</h5>
<p>To examine whether the features obtained from <code>Statial</code>
are associated with patient outcomes or groupings, we can use the
<code>colTest</code> function from <code>SpicyR</code>. To understand if
survival outcomes differ significantly between 2 patient groups, specify
<code>type = "survival"</code> in <code>colTest</code>. Here we examine
which features are most associated with patient survival using the
Kontextual values as an example. To do so, survival data is extracted
from <code>kerenSPE</code> and converted into the survival object
<code>kerenSurv</code>.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extracting survival data</span></span>
<span><span class="va">survData</span> <span class="op">=</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">imageID</span>, <span class="va">Survival_days_capped.</span>, <span class="va">event</span>, <span class="va">RECURRENCE_LABEL</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Creating survival vector</span></span>
<span><span class="va">kerenSurv</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">survData</span><span class="op">$</span><span class="va">Survival_days_capped</span>, <span class="va">survData</span><span class="op">$</span><span class="va">event</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span> <span class="op">=</span> <span class="va">survData</span><span class="op">$</span><span class="va">imageID</span></span></code></pre></div>
<p>In addition to this, the Kontextual results must be converted from a
<code>data.frame</code> to a wide <code>matrix</code>, this can be done
using <code>prepMatrix</code>. Note, to extract the original L-function
values, specify <code>column = "original"</code> in
<code>prepMatrix</code>.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Kontextual result into data matrix</span></span>
<span><span class="va">kontextMat</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">kerenKontextual</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensuring rownames of kontextMat match up with rownames of the survival vector </span></span>
<span><span class="va">kontextMat</span> <span class="op">=</span> <span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Replace NAs with 0</span></span>
<span><span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">kontextMat</span> <span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Finally, both the Kontextual matrix and survival object are passed
into <code>colTest</code>, with <code>type = "survival"</code> to obtain
the survival results.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Running survival analysis</span></span>
<span><span class="va">survivalResults</span> <span class="op">=</span> <span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">kontextMat</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;                                      coef se.coef   pval adjPval</span></span>
<span><span class="co">#&gt; CD8_T_cell__Keratin_Tumour__immune -0.150   0.043 0.0006    0.21</span></span>
<span><span class="co">#&gt; CD4_T_cell__CD8_T_cell__tcells     -0.054   0.019 0.0039    0.51</span></span>
<span><span class="co">#&gt; CD4_T_cell__CD8_T_cell__immune     -0.040   0.015 0.0094    0.51</span></span>
<span><span class="co">#&gt; Endothelial__Mono_or_Neu__tissue   -0.031   0.012 0.0110    0.51</span></span>
<span><span class="co">#&gt; Other_Immune__NK__myeloid           0.045   0.018 0.0120    0.51</span></span>
<span><span class="co">#&gt; Endothelial__CD8_T_cell__tissue    -0.033   0.013 0.0120    0.51</span></span>
<span><span class="co">#&gt;                                                               cluster</span></span>
<span><span class="co">#&gt; CD8_T_cell__Keratin_Tumour__immune CD8_T_cell__Keratin_Tumour__immune</span></span>
<span><span class="co">#&gt; CD4_T_cell__CD8_T_cell__tcells         CD4_T_cell__CD8_T_cell__tcells</span></span>
<span><span class="co">#&gt; CD4_T_cell__CD8_T_cell__immune         CD4_T_cell__CD8_T_cell__immune</span></span>
<span><span class="co">#&gt; Endothelial__Mono_or_Neu__tissue     Endothelial__Mono_or_Neu__tissue</span></span>
<span><span class="co">#&gt; Other_Immune__NK__myeloid                   Other_Immune__NK__myeloid</span></span>
<span><span class="co">#&gt; Endothelial__CD8_T_cell__tissue       Endothelial__CD8_T_cell__tissue</span></span></code></pre>
<p>As we can see from the results
<code>Mesenchymal__Macrophages__tissue</code> is the most significant
pairwise relationship which contributes to patient survival. That is the
relationship between Mesenchymal cells and macrophage cells, relative to
the parent population of all tissue cells. We can see that there is a
negative coefficient associated with this relationship, which tells us a
decrease in localisation of Mesenchymal and Macrophages leads to poorer
survival outcomes for patients.</p>
<p>The association between <code>Mesenchymal__Macrophages__tissue</code>
and survival can also be visualised on a Kaplan-Meier curve. We must
first extract the Kontextual values of this relationship across all
images. Next we determine if Mesenchymal and Macrophages are relatively
attracted or avoiding in each image, by comparing the Kontextual value
in each image to the median Kontextual value. Finally we plot the
Kaplan-Meier curve using the <code>ggsurvfit</code> package.</p>
<p>As shown below, when Mesenchymal and Macrophages are relatively more
dispersed to one another, patients tend to have worse survival
outcomes.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="va">kontextMat</span><span class="op">[[</span><span class="st">"Mesenchymal__Macrophages__tissue"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Localised"</span>, <span class="st">"Dispersed"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html" class="external-link">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html" class="external-link">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html" class="external-link">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mesenchymal__Macrophages__tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/kontextSurvPlot-1.png" width="480"></p>
</div>
</div>
<div class="section level4">
<h4 id="can-i-visualise-changes-in-specific-markers-as-spatial-proximity-between-cell-types-change">Can I visualise changes in specific markers as spatial proximity
between cell types change?<a class="anchor" aria-label="anchor" href="#can-i-visualise-changes-in-specific-markers-as-spatial-proximity-between-cell-types-change"></a>
</h4>
<p>Changes in cell phenotype can be analytically framed as the change in
abundance of a gene or protein within a particular cell type. We can
analytically determine whether continuous changes occur to a cell’s
proteomic or transcriptomic signature as changes occur in its spatial
proximity to another cell type. In the figures below we see the
expression of a marker increased in cell type A as it grows closer in
spatial proximity to cell type B. This can then be quantified with a
scatterplot to determine statistical significance. In the next section
of this workshop, we will be exploring the analytical functionalities of
Statial which can uncover these continuous changes in cell state.</p>
<p><img src="images/ctsCombined.jpg" alt="ctsCombined" style="border: 0px"></p>
<div class="section level5">
<h5 id="how-do-we-identify-specific-marker-changes-when-distinct-cell-types-increase-or-decrease-in-spatial-proximity">How do we identify specific marker changes when distinct cell types
increase or decrease in spatial proximity?<a class="anchor" aria-label="anchor" href="#how-do-we-identify-specific-marker-changes-when-distinct-cell-types-increase-or-decrease-in-spatial-proximity"></a>
</h5>
<p>The first step in analysing these changes is to calculate the spatial
proximity (<code>getDistances</code>) and abundance
(<code>getAbundances</code>) of each cell to every cell type. These
values will then be stored in the <code>reducedDims</code> slot of the
<code>SpatialExperiment</code> object under the names
<code>distances</code> and <code>abundances</code> respectively.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getDistances.html" class="external-link">getDistances</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>                    maxDist <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getAbundances.html" class="external-link">getAbundances</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>                     r <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"abundances"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"abundances"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>First, let’s examine the same effect observed earlier with
Kontextual. To avoid redefining cell types we’ll examine the distance
between p53-positive tumour cells and macrophages in the context of
total keratin/tumour cells for image 6.</p>
<p>Statial provides two main functions to assess this relationship -
<code>calcStateChanges</code> and <code>plotStateChanges</code>. We can
use <code>calcStateChanges</code> to examine the relationship between 2
cell types for 1 marker in a specific image. Similar to
<code>Kontextual</code>, we can specify the two cell types with the
<code>to</code> and <code>from</code> arguments, and the marker of
interest with the <code>marker</code> argument. We can appreciate that
the <code>fdr</code> statistic for this relationship is significant, and
with a negative <code>coef</code>, or coefficient value, indicating that
the expression of p53 in keratin/tumour cells decreases as distance from
macrophages increases.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"p53"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChanges</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   imageID primaryCellType otherCellType marker         coef      tval</span></span>
<span><span class="co">#&gt; 1       6  Keratin_Tumour   Macrophages    p53 -0.001402178 -7.010113</span></span>
<span><span class="co">#&gt;           pval          fdr</span></span>
<span><span class="co">#&gt; 1 2.868257e-12 2.868257e-12</span></span></code></pre>
<p>Statial provides a convenient function for visualising this
relationship - <code>plotStateChanges</code>. Similar to
<code>Kontextual</code> and <code>calcStateChanges</code>, we can
specify the cell types to be evaluated with the <code>to</code> and
<code>from</code> arguments and the marker of interest with
<code>marker</code>.</p>
<p>Through this analysis, we can observe that keratin/tumour cells
closer to a group of macrophages tend to have higher expression of p53,
as observed in the first graph. This relationship is quantified with the
second graph, showing an overall decrease of p53 expression in
keratin/tumour cells as distance to macrophages increase.</p>
<p>These results allow us to essentially arrive at the same result as
Kontextual, which calculated a localisation between p53+ keratin/tumour
cells and macrophages in the wider context of keratin/tumour cells.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html" class="external-link">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"p53"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-example2-1.png" width="576"></p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"p53 expression in Keratin_Tumour"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Distance of Keratin_Tumour to Macrophages"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-example2-2.png" width="576"></p>
<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>What information does this form of analysis provide that Kontextual
does not?</li>
<li>Is this observation of localisation consistent across images?</li>
<li>Can you find an interaction where the coefficient is positive? i.e.
marker expression in the <code>to</code> cell type rises as distances
increases from the <code>from</code> cell type.</li>
</ol>
</div>
</div>
<div class="section level5">
<h5 id="can-we-repeat-this-analysis-across-all-images-to-identify-global-relationships">Can we repeat this analysis across all images to identify global
relationships?<a class="anchor" aria-label="anchor" href="#can-we-repeat-this-analysis-across-all-images-to-identify-global-relationships"></a>
</h5>
<p>Beyond looking at single cell-to-cell interactions for a single
image, we can also look at all interactions across all images. The
<code>calcStateChanges</code> function provided by Statial can be
expanded for this exact purpose - by not specifying cell types, a
marker, or an image, <code>calcStateChanges</code> will examine the most
significant correlations between distance and marker expression across
the entire dataset. Here, we’ve calculated all state changes across all
images in case you would like to have a play, but first we’ll be taking
a closer examination at the most significant interactions found within
image 6 of the Keren et al. dataset.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChanges</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="fl">6</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    imageID primaryCellType otherCellType       marker         coef      tval</span></span>
<span><span class="co">#&gt; 1        6  Keratin_Tumour  Unidentified           Na  0.004218419  25.03039</span></span>
<span><span class="co">#&gt; 2        6  Keratin_Tumour   Macrophages  HLA_Class_1 -0.003823497 -24.69629</span></span>
<span><span class="co">#&gt; 3        6  Keratin_Tumour    CD4_T_cell  HLA_Class_1 -0.003582774 -23.87797</span></span>
<span><span class="co">#&gt; 4        6  Keratin_Tumour  Unidentified Beta.catenin  0.005893120  23.41953</span></span>
<span><span class="co">#&gt; 5        6  Keratin_Tumour    CD8_T_cell  HLA_Class_1 -0.003154544 -23.13804</span></span>
<span><span class="co">#&gt; 6        6  Keratin_Tumour    DC_or_Mono  HLA_Class_1 -0.003353834 -22.98944</span></span>
<span><span class="co">#&gt; 7        6  Keratin_Tumour      dn_T_CD3  HLA_Class_1 -0.003123446 -22.63197</span></span>
<span><span class="co">#&gt; 8        6  Keratin_Tumour        Tumour  HLA_Class_1  0.003684079  21.94265</span></span>
<span><span class="co">#&gt; 9        6  Keratin_Tumour    CD4_T_cell           Fe -0.003457338 -21.43550</span></span>
<span><span class="co">#&gt; 10       6  Keratin_Tumour    CD4_T_cell   phospho.S6 -0.002892457 -20.50767</span></span>
<span><span class="co">#&gt;             pval           fdr</span></span>
<span><span class="co">#&gt; 1  6.971648e-127 3.397517e-123</span></span>
<span><span class="co">#&gt; 2  7.814253e-124 3.427331e-120</span></span>
<span><span class="co">#&gt; 3  1.745242e-116 5.670098e-113</span></span>
<span><span class="co">#&gt; 4  1.917245e-112 5.799337e-109</span></span>
<span><span class="co">#&gt; 5  5.444541e-110 1.516175e-106</span></span>
<span><span class="co">#&gt; 6  1.053130e-108 2.842478e-105</span></span>
<span><span class="co">#&gt; 7  1.237988e-105 3.016563e-102</span></span>
<span><span class="co">#&gt; 8  8.188258e-100  1.710176e-96</span></span>
<span><span class="co">#&gt; 9   1.287478e-95  2.352866e-92</span></span>
<span><span class="co">#&gt; 10  3.928912e-88  5.891353e-85</span></span></code></pre>
<p>In image 6, the majority of the top 10 most significant interactions
occur between keratin/tumour cells and an immune population, and many of
these interactions appear to involve the HLA class I ligand.</p>
<p>We can examine some of these interactions further with the
<code>plotStateChanges</code> function. Taking a closer examination of
the relationship between macrophages and keratin/tumour HLA class I
expression, the plot below shows us a clear visual correlation - as
macrophage density increases, keratin/tumour cells increase their
expression HLA class I.</p>
<p>Biologically, HLA Class I is a ligand which exists on all nucleated
cells, tasked with presenting internal cell antigens for recognition by
the immune system, marking aberrant cells for destruction by either CD8+
T cells or NK cells.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html" class="external-link">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"HLA_Class_1"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-example3-1.png" width="576"></p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"HLA_Class_1 expression in Keratin_Tumour"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Distance of Keratin_Tumour to Macrophages"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-example3-2.png" width="576"></p>
<p>Now, we can take a look at the top 10 most significant results across
all images.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;        imageID primaryCellType otherCellType     marker         coef</span></span>
<span><span class="co">#&gt; 69468       37     Endothelial        Tumour       Lag3 -0.001621517</span></span>
<span><span class="co">#&gt; 144255      11     Neutrophils            NK       CD56 -0.059936866</span></span>
<span><span class="co">#&gt; 16402       35      CD4_T_cell        B_cell       CD20 -0.029185750</span></span>
<span><span class="co">#&gt; 16498       35      CD4_T_cell    DC_or_Mono       CD20  0.019125946</span></span>
<span><span class="co">#&gt; 4891        35          B_cell    DC_or_Mono phospho.S6  0.005282065</span></span>
<span><span class="co">#&gt; 16507       35      CD4_T_cell    DC_or_Mono phospho.S6  0.004033218</span></span>
<span><span class="co">#&gt; 4885        35          B_cell    DC_or_Mono     HLA.DR  0.011120703</span></span>
<span><span class="co">#&gt; 5043        35          B_cell  Other_Immune          P  0.011182182</span></span>
<span><span class="co">#&gt; 16354       35      CD4_T_cell      dn_T_CD3       CD20  0.016349492</span></span>
<span><span class="co">#&gt; 4888        35          B_cell    DC_or_Mono     H3K9ac  0.005096632</span></span>
<span><span class="co">#&gt;                 tval          pval           fdr</span></span>
<span><span class="co">#&gt; 69468  -1.601116e+15  0.000000e+00  0.000000e+00</span></span>
<span><span class="co">#&gt; 144255 -2.435997e+15  0.000000e+00  0.000000e+00</span></span>
<span><span class="co">#&gt; 16402  -4.057355e+01 7.019343e-282 4.104912e-277</span></span>
<span><span class="co">#&gt; 16498   4.053436e+01 1.891267e-281 8.295098e-277</span></span>
<span><span class="co">#&gt; 4891    4.041385e+01 5.306590e-278 1.861976e-273</span></span>
<span><span class="co">#&gt; 16507   3.472882e+01 4.519947e-219 1.321633e-214</span></span>
<span><span class="co">#&gt; 4885    3.415344e+01 8.401034e-212 2.105539e-207</span></span>
<span><span class="co">#&gt; 5043    3.414375e+01 1.056403e-211 2.316692e-207</span></span>
<span><span class="co">#&gt; 16354   3.391901e+01 1.219488e-210 2.377188e-206</span></span>
<span><span class="co">#&gt; 4888    3.399856e+01 3.266533e-210 5.730805e-206</span></span></code></pre>
<p>Immediately, we can appreciate that a couple of interactions appear a
bit strange. One of the most significant interactions occurs between B
cells and CD4 T cells, where CD4 T cells are found to increase in CD20
expression when in close proximity to B cells. Biologically, CD20 is a
highly specific ligand for B cells, and under healthy circumstances are
usually not expressed in T cells.</p>
<p>Could this potentially be an artefact of
<code>calcStateChanges</code>? We can examine the image through the
<code>plotStateChanges</code> function, where we indeed observe an
apparent localisation between B cells and T cells.</p>
<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li><p>Are there any other interactions here that you think might not
make biological sense?<br></p></li>
<li><p>Does the relationship between T cell CD20 expression and B cell
proximity occur across images?<br></p></li>
<li>
<p>Why are the majority of most significant interactions occurring
in image 35?</p>
<p>HINT: Configure the parameters of <code>plotStateChanges</code> to
examine some these other significant interactions. Do they look like
artefacts?</p>
</li>
</ol>
</div>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html" class="external-link">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"35"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"CD4_T_cell"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"B_cell"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"CD20"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"CD20 expression in CD4_T_cell"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Distance of B_cell to CD4_T_cell"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-contamination-1.png" width="576"></p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-contamination-2.png" width="576"></p>
<p>So why are T cells expressing CD20? This brings us to a key
limitation of cell segmentation.</p>
</div>
<div class="section level5">
<h5 id="how-do-we-deal-with-lateral-cell-spill-over">How do we deal with lateral cell spill over?<a class="anchor" aria-label="anchor" href="#how-do-we-deal-with-lateral-cell-spill-over"></a>
</h5>
<p>Contamination, or more specifically known as lateral marker spill
over, is an issue that results in a cell’s marker expressions being
wrongly attributed to another adjacent cell. This issue arises from
incorrect segmentation where components of one cell are wrongly
determined as belonging to another cell. Alternatively, this issue can
arise when antibodies used to tag and measure marker expressions do not
latch on properly to a cell of interest, thereby resulting in residual
markers being wrongly assigned as belonging to a cell near the intended
target cell. It is important that we either correct or account for this
incorrect attribution of markers in our modelling process. This is
critical in understanding whether significant cell-cell interactions
detected are an artifact of technical measurement errors driven by spill
over or are real biological changes that represent a shift in a cell’s
state.</p>
<p><img src="images/Contamination.png" alt="Contamination" style="width: 4in; border: 0px"></p>
<p>To circumvent this problem, <code>Statial</code> provides a function
that predicts the probability that a cell is any particular cell type -
<code>calcContamination</code>. <code>calcContamination</code> returns a
dataframe of probabilities demarcating the chance of a cell being any
particular cell type. This dataframe is stored under
<code>contaminations</code> in the <code>reducedDim</code> slot of the
<code>SpatialExperiment</code> object. It also provides the
<code>rfMainCellProb</code> column, which provides the probability that
a cell is indeed the cell type it has been designated. E.g. For a cell
designated as a CD8+ T cell, rfMainCellProb could give a 80% chance that
the cell is indeed CD8+ T cell, due to contamination.</p>
<p>We can then introduce these probabilities as covariates into our
linear model by setting <code>contamination = TRUE</code> as a parameter
in our <code>calcStateChanges</code> function. However, this is not a
perfect solution for the issue of contamination. As we can see, despite
factoring in contamination into our linear model, the correlation
between B cell density and CD20 expression in CD4+ T cells remains one
of the most significant interactions in our model.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcContamination.html" class="external-link">calcContamination</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Growing trees.. Progress: 53%. Estimated remaining time: 27 seconds.</span></span></code></pre>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stateChangesCorrected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  contamination <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChangesCorrected</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;        imageID primaryCellType otherCellType      marker         coef</span></span>
<span><span class="co">#&gt; 69468       37     Endothelial        Tumour        Lag3 -0.001621517</span></span>
<span><span class="co">#&gt; 144255      11     Neutrophils            NK        CD56 -0.059936866</span></span>
<span><span class="co">#&gt; 16402       35      CD4_T_cell        B_cell        CD20 -0.025417077</span></span>
<span><span class="co">#&gt; 16498       35      CD4_T_cell    DC_or_Mono        CD20  0.016446000</span></span>
<span><span class="co">#&gt; 4891        35          B_cell    DC_or_Mono  phospho.S6  0.004436754</span></span>
<span><span class="co">#&gt; 16354       35      CD4_T_cell      dn_T_CD3        CD20  0.014094253</span></span>
<span><span class="co">#&gt; 16507       35      CD4_T_cell    DC_or_Mono  phospho.S6  0.003676138</span></span>
<span><span class="co">#&gt; 16357       35      CD4_T_cell      dn_T_CD3      HLA.DR  0.010487694</span></span>
<span><span class="co">#&gt; 3697        28          B_cell            NK          Na -0.004549435</span></span>
<span><span class="co">#&gt; 85540        3  Keratin_Tumour            DC          Ca -0.013661908</span></span>
<span><span class="co">#&gt; 4741        35          B_cell      dn_T_CD3      HLA.DR  0.009098103</span></span>
<span><span class="co">#&gt; 82078       23  Keratin_Tumour  Unidentified HLA_Class_1  0.003193547</span></span>
<span><span class="co">#&gt; 4885        35          B_cell    DC_or_Mono      HLA.DR  0.009010498</span></span>
<span><span class="co">#&gt; 80974       20  Keratin_Tumour        Tumour HLA_Class_1  0.002896132</span></span>
<span><span class="co">#&gt; 5083        35          B_cell  Other_Immune  phospho.S6  0.004769406</span></span>
<span><span class="co">#&gt; 16363       35      CD4_T_cell      dn_T_CD3  phospho.S6  0.003018899</span></span>
<span><span class="co">#&gt; 81737       21  Keratin_Tumour            DC Pan.Keratin -0.005965789</span></span>
<span><span class="co">#&gt; 4888        35          B_cell    DC_or_Mono      H3K9ac  0.004089079</span></span>
<span><span class="co">#&gt; 4747        35          B_cell      dn_T_CD3  phospho.S6  0.003764374</span></span>
<span><span class="co">#&gt; 5043        35          B_cell  Other_Immune           P  0.008211837</span></span>
<span><span class="co">#&gt;                 tval          pval           fdr</span></span>
<span><span class="co">#&gt; 69468  -1.041435e+15  0.000000e+00  0.000000e+00</span></span>
<span><span class="co">#&gt; 144255 -2.401765e+15  0.000000e+00  0.000000e+00</span></span>
<span><span class="co">#&gt; 16402  -3.582248e+01 2.397650e-230 1.402146e-225</span></span>
<span><span class="co">#&gt; 16498   3.480159e+01 1.271214e-219 5.575546e-215</span></span>
<span><span class="co">#&gt; 4891    3.118797e+01 9.965324e-182 3.496633e-177</span></span>
<span><span class="co">#&gt; 16354   3.060287e+01 7.064297e-177 2.065600e-172</span></span>
<span><span class="co">#&gt; 16507   3.030375e+01 6.283951e-174 1.574938e-169</span></span>
<span><span class="co">#&gt; 16357   2.956638e+01 1.016140e-166 2.228395e-162</span></span>
<span><span class="co">#&gt; 3697   -2.977454e+01 2.086988e-164 4.068235e-160</span></span>
<span><span class="co">#&gt; 85540  -2.929766e+01 6.445385e-160 1.130778e-155</span></span>
<span><span class="co">#&gt; 4741    2.633749e+01 9.689227e-136 1.545344e-131</span></span>
<span><span class="co">#&gt; 82078   2.579280e+01 2.253761e-134 3.294999e-130</span></span>
<span><span class="co">#&gt; 4885    2.611891e+01 9.094602e-134 1.227352e-129</span></span>
<span><span class="co">#&gt; 80974   2.546445e+01 5.006160e-131 6.273434e-127</span></span>
<span><span class="co">#&gt; 5083    2.574695e+01 1.964973e-130 2.298233e-126</span></span>
<span><span class="co">#&gt; 16363   2.547751e+01 1.606492e-128 1.761518e-124</span></span>
<span><span class="co">#&gt; 81737  -2.469446e+01 1.033257e-126 1.066322e-122</span></span>
<span><span class="co">#&gt; 4888    2.531328e+01 1.396961e-126 1.361572e-122</span></span>
<span><span class="co">#&gt; 4747    2.508124e+01 1.548912e-124 1.430216e-120</span></span>
<span><span class="co">#&gt; 5043    2.506522e+01 2.141765e-124 1.878756e-120</span></span></code></pre>
<p>However, this does not mean factoring in contamination into our
linear model was ineffective. In general, cell type specific markers
such as CD68, CD45, and CD20 should not change in cells they are not
specific to. Therefore, relationships detected to be significant
involving these cell type markers are likely false positives and will be
treated as such for the purposes of evaluation.</p>
<p>Plotting the relationship between false positives and true positives,
we’d expect the contamination correction to be greatest in relationships
which are detected to be more significant.</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellTypeMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"CD56"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD68"</span>, <span class="st">"CD45"</span>, <span class="st">"CD20"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">values</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"None"</span>, <span class="st">"Corrected"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>TP <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">stateChanges</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">stateChanges</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"None"</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>TP <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">stateChangesCorrected</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">stateChangesCorrected</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"Corrected"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TP</span>, y <span class="op">=</span> <span class="va">FP</span>, colour <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell state marker"</span>, x <span class="op">=</span> <span class="st">"Cell type marker"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-roc-1.png" width="480"></p>
<p>Here, we zoom in on the ROC curve where the top 100 lowest p values
occur, where we indeed see more true positives than false positives with
contamination correction.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TP</span>, y <span class="op">=</span> <span class="va">FP</span>, colour <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell state marker"</span>, x <span class="op">=</span> <span class="st">"Cell type marker"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomark-roc-zoom-1.png" width="480"></p>
<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>What can we conclude from the above ROC graphs?</li>
</ol>
</div>
</div>
<div class="section level5">
<h5 id="are-these-cell-type---marker-relationships-predictive-or-survival-outcomes">Are these cell type - marker relationships predictive or survival
outcomes?<a class="anchor" aria-label="anchor" href="#are-these-cell-type---marker-relationships-predictive-or-survival-outcomes"></a>
</h5>
<p>Similiar to <code>Kontextual</code>, we can run a similar survival
analysis using our state changes results. Here, <code>prepMatrix</code>
extracts the coefficients, or the <code>coef</code> column of
<code>stateChanges</code> by default. To use the t values instead,
specify <code>column = "tval"</code> in the <code>prepMatrix</code>
function.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Preparing features for Statial</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">stateChanges</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensuring rownames of stateMat match up with rownames of the survival vector</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">stateMat</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.0001</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">stateMat</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;                                          coef se.coef    pval adjPval</span></span>
<span><span class="co">#&gt; Macrophages__CD4_T_cell__CD138            440     110 8.1e-05   0.054</span></span>
<span><span class="co">#&gt; Macrophages__Other_Immune__HLA_Class_1   -500     170 2.9e-03   0.620</span></span>
<span><span class="co">#&gt; Keratin_Tumour__Mesenchymal__dsDNA       -930     330 4.9e-03   0.620</span></span>
<span><span class="co">#&gt; Keratin_Tumour__CD8_T_cell__Keratin6     -220      80 5.2e-03   0.620</span></span>
<span><span class="co">#&gt; Keratin_Tumour__Mono_or_Neu__Pan.Keratin -260      94 6.0e-03   0.620</span></span>
<span><span class="co">#&gt; Macrophages__Other_Immune__CD4           -480     180 7.6e-03   0.620</span></span>
<span><span class="co">#&gt;                                                                           cluster</span></span>
<span><span class="co">#&gt; Macrophages__CD4_T_cell__CD138                     Macrophages__CD4_T_cell__CD138</span></span>
<span><span class="co">#&gt; Macrophages__Other_Immune__HLA_Class_1     Macrophages__Other_Immune__HLA_Class_1</span></span>
<span><span class="co">#&gt; Keratin_Tumour__Mesenchymal__dsDNA             Keratin_Tumour__Mesenchymal__dsDNA</span></span>
<span><span class="co">#&gt; Keratin_Tumour__CD8_T_cell__Keratin6         Keratin_Tumour__CD8_T_cell__Keratin6</span></span>
<span><span class="co">#&gt; Keratin_Tumour__Mono_or_Neu__Pan.Keratin Keratin_Tumour__Mono_or_Neu__Pan.Keratin</span></span>
<span><span class="co">#&gt; Macrophages__Other_Immune__CD4                     Macrophages__Other_Immune__CD4</span></span></code></pre>
<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>Are the survival results the same for both the distance and
abundance metrics?</li>
<li>Do these survival results change with contamination correction?</li>
<li>Why might the survival results look the same/different between
distance and abundance?</li>
</ol>
</div>
<p>For our state changes results,
<code>Keratin_Tumour__Mesenchymal__HLA_Class_1</code> is the most
significant pairwise relationship which contributes to patient survival.
That is, the relationship between HLA class I expression in
keratin/tumour cells and their spatial proximity to mesenchymal cells.
As there is a negative coeffcient associated with this relationship,
which tells us that higher HLA class I expression in keratin/tumour
cells nearby mesenchymal cell populations lead to poorer survival
outcomes for patients.</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting the most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="va">stateMat</span><span class="op">[[</span><span class="st">"Keratin_Tumour__Mesenchymal__HLA_Class_1"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Higher expressed in close cells"</span>, <span class="st">"Lower expressed in close cells"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html" class="external-link">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html" class="external-link">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html" class="external-link">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Keratin_Tumour__Mesenchymal__HLA_Class_1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/spatiomarkSurvPlot-1.png" width="480"></p>
<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>How should these coefficients be interpreted?</li>
<li>Do any of these relationships makes sense?</li>
<li>Could you visualise representative images?</li>
</ol>
</div>
</div>
</div>
<div class="section level4">
<h4 id="how-do-we-generate-a-molecular-representation-for-each-individual">How do we generate a molecular representation for each
individual?<a class="anchor" aria-label="anchor" href="#how-do-we-generate-a-molecular-representation-for-each-individual"></a>
</h4>
<p>The next section of this workshop will be dedicated to scFeatures - a
straightforward package for generating a suite of molecular
representations regarding a patient, taking a matrix of
<code>proteins x cells</code> as input. The molecular representation is
interpretable and hence facilitates downstream analysis of the
individual. In general, <code>scFeatures</code> generates features
across six categories representing different molecular views of cellular
characteristics. These include:</p>
<ol style="list-style-type: lower-roman">
<li>cell type proportions</li>
<li>cell type specific gene expressions</li>
<li>cell type specific pathway expressions</li>
<li>cell type specific cell-cell interaction (CCI) scores</li>
<li>overall aggregated gene expressions</li>
<li>spatial metrics</li>
</ol>
<p>The different types of features constructed enable a more
comprehensive multi-view understanding of each individual from a matrix
of <code>spot x cells</code>. By default, the function will generate a
total of 13 different types of features (feature_types) shown below and
store them in a list. All generated feature types are stored in a matrix
of <code>samples x features</code>.</p>
<pre><code><span><span class="co">##  [1] "proportion_raw"       "proportion_logit"     "proportion_ratio"    </span></span>
<span><span class="co">##  [4] "gene_mean_celltype"   "gene_prop_celltype"   "gene_cor_celltype"   </span></span>
<span><span class="co">##  [7] "gene_mean_bulk"       "gene_prop_bulk"       "gene_cor_bulk"       </span></span>
<span><span class="co">## [10] "L_stats"              "celltype_interaction" "morans_I"            </span></span>
<span><span class="co">## [13] "nn_correlation"</span></span></code></pre>
<p>There are different ways you can use <code>scFeatures</code> to
generate molecular representations for individuals and it requires the
following information for spatial data.</p>
<ul>
<li>data,</li>
<li>sample,</li>
<li>X coordinates,</li>
<li>Y coordinates,</li>
<li>feature_types, and</li>
<li>type</li>
</ul>
<p>There are a total of 13 different types of features (feature_types)
that you can choose to generate. The argument type refers the type of
input data we have. This is either <code>scrna</code> (single-cell
RNA-sequencing data), <code>spatial_p</code> (spatial proteomics data),
or <code>spatial_t</code> (single cell spatial data).</p>
<p>Suppose that we are interested in determining the proportion of each
cell type in each individual within each region. It is necessary to
specify <code>type = spatial_p</code> to reflect that we have spatial
proteomics data and <code>feature_types = proportion_raw</code> to
indicate we intend to calculate cell type proportion for each of the
region-specific sub-cell types.</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"intensities"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellType</span></span>
<span><span class="va">spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scFeatures/man/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span><span class="va">matrix</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span>,</span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, celltype <span class="op">=</span> <span class="va">celltype</span>, spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>,</span>
<span>                                ncores <span class="op">=</span> <span class="fl">32</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"scfeaturesResult.rda"</span>, package <span class="op">=</span> <span class="st">"ScdneySpatial"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;        B_cell  CD4_T_cell CD8_T_cell           DC  DC_or_Mono</span></span>
<span><span class="co">#&gt; 1 0.221985678 0.047029224 0.03348171 0.0003870718 0.034062319</span></span>
<span><span class="co">#&gt; 2 0.000660502 0.006935271 0.08091149 0.0122192867 0.002972259</span></span>
<span><span class="co">#&gt; 3 0.067616785 0.077751386 0.07854315 0.0057007126 0.068725257</span></span>
<span><span class="co">#&gt; 4 0.049525817 0.110341713 0.04576246 0.0003010688 0.068944754</span></span>
<span><span class="co">#&gt; 5 0.000554939 0.047169811 0.09822420 0.0005549390 0.061598224</span></span></code></pre>
<p>The output of scFeatures, <code>scfeatures_result</code>, is a list
of 13 dataframes, which each represent a different molecular feature of
a patient. Below, we give an example visualisation of the raw cell type
proportions for each patient.</p>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span></span>
<span><span class="va">feature</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">feature</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">feature</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient"</span>, <span class="st">"condition"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">feature</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">patient</span> , y <span class="op">=</span> <span class="va">value</span> , fill <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>   </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Proportion raw feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/scFeaturesPlot-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="how-good-are-our-metrics-at-predicting-patient-survival">How good are our metrics at predicting patient survival?<a class="anchor" aria-label="anchor" href="#how-good-are-our-metrics-at-predicting-patient-survival"></a>
</h4>
<p>Finally we demonstrate how we can use the Bioconductor package
<code>ClassifyR</code> to perform patient classification with the
features generated from <code>Statial</code>. In addition to
<code>Kontextual</code>, <code>SpatioMark</code>, and
<code>scFeatures</code> values, we also calculate cell type proportions
and region proportions using the <code>getProp</code> function in
<code>spicyR</code>. Here we perform 5 fold cross validation with 20
repeats, using a CoxPH model for survival classification.</p>
<p>To prepare our metrics for classification, first we must convert our
dataframes into matrix format. This can be accomplished, as we did
earlier, with the <code>prepMatrix</code> function of
<code>Statial</code>. As we generated matrices for
<code>Kontextual</code> and the distance metric of
<code>SpatioMark</code> earlier, we will now be generating matrices for
the <code>SpatioMark</code> corrected distance metric, and the abundance
metrics.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Distances Corrected Matrix</span></span>
<span><span class="va">stateMatCorDist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">stateChangesCorrected</span><span class="op">)</span></span>
<span><span class="va">stateMatCorDist</span> <span class="op">&lt;-</span> <span class="va">stateMatCorDist</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">stateMatCorDist</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">stateMatCorDist</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">stateMatCorDist</span> <span class="op">&lt;-</span> <span class="va">stateMatCorDist</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">stateMatCorDist</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.0001</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"abundances"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">kerenSPE</span>, <span class="st">"abundances"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co">## Abundances</span></span>
<span><span class="va">stateChangesAbund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"abundances"</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateMatAbund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">stateChangesAbund</span><span class="op">)</span></span>
<span><span class="va">stateMatAbund</span> <span class="op">&lt;-</span> <span class="va">stateMatAbund</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">stateMatAbund</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">stateMatAbund</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">stateMatAbund</span> <span class="op">&lt;-</span> <span class="va">stateMatAbund</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">stateMatAbund</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.0001</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Abundances Corrected</span></span>
<span><span class="va">stateChangesCorrectedAbund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html" class="external-link">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"abundances"</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  contamination <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateMatCorAbund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html" class="external-link">prepMatrix</a></span><span class="op">(</span><span class="va">stateChangesCorrectedAbund</span><span class="op">)</span></span>
<span><span class="va">stateMatCorAbund</span> <span class="op">&lt;-</span> <span class="va">stateMatCorAbund</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">stateMatCorAbund</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">stateMatCorAbund</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">stateMatCorAbund</span> <span class="op">&lt;-</span> <span class="va">stateMatCorAbund</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">stateMatCorAbund</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.0001</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">.8</span><span class="op">]</span></span></code></pre></div>
<p>Then we compile these matrices into a list. <code>scFeatures</code>
conveniently provides all its metrics as part of a list of matrices
already, allowing us to easily combine our <code>Statial</code> metrics
with <code>scFeatures</code> into 1 list of matrices. We then input this
list into the <code>crossValidate</code> function of
<code>ClassifyR</code>, specifying our survival object created earlier,
and 5 folds of cross validation with 20 repeats. We will also be using a
feature selection algorithm selecting for 10 features for each
metric.</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">statialFeatureList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>`SpatioMark Distances` <span class="op">=</span> <span class="va">stateMat</span>,</span>
<span>                           `SpatioMark Corrected Distances` <span class="op">=</span> <span class="va">stateMatCorDist</span>,</span>
<span>                           `SpatioMark Abundances` <span class="op">=</span> <span class="va">stateMatAbund</span>,</span>
<span>                           `SpatioMark Corrected Abundances` <span class="op">=</span> <span class="va">stateMatCorAbund</span>,</span>
<span>                           `Kontextual` <span class="op">=</span> <span class="va">kontextMat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="va">featureList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">statialFeatureList</span>, <span class="va">scfeatures_result</span><span class="op">)</span></span>
<span><span class="va">featureList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">featureList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenCV</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">featureList</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">kerenSurv</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"CoxPH"</span>,</span>
<span>  selectionMethod  <span class="op">=</span> <span class="st">"CoxPH"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nFeatures <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Next, we use the <code>performancePlot</code> function to assess the
C-index from each repeat of the 5-fold cross-validation.</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span><span class="va">kerenCV</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">statialFeatureList</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_rect</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">0</span>, xmax <span class="op">=</span> <span class="fl">5.5</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>, ymax <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_rect</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">5.5</span>, xmax <span class="op">=</span> <span class="fl">14.5</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>, ymax <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"blue"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_rect</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">14.5</span>, xmax <span class="op">=</span> <span class="fl">19</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>, ymax <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"green"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/perfPlot-coxph-1.png" width="700"></p>
<p>Unfortunately, our spatial metrics did not outperform some of the
simpler metrics generated by scFeatures, particularly
<code>gene_mean_bulk</code>, which simply produces a matrix describing
the average gene expression of each marker in each patient. To examine
another variable which might be of interest, we used recurrence as
another outcome to determine whether our spatial metrics could uncover
signal.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outcome</span> <span class="op">&lt;-</span> <span class="va">survData</span><span class="op">$</span><span class="va">RECURRENCE_LABEL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">survData</span><span class="op">$</span><span class="va">imageID</span></span>
<span></span>
<span><span class="va">outcome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span></span>
<span></span>
<span><span class="va">featureList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">featureList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenCV_recurrence</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">featureList</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">outcome</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"elasticNetGLM"</span>,</span>
<span>  selectionMethod <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nFeatures <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Again, using <code>performancePlot</code>, this time for recurrence,
we found better performance in select spatial metrics.</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span><span class="va">kerenCV_recurrence</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">statialFeatureList</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_rect</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">0</span>, xmax <span class="op">=</span> <span class="fl">5.5</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>, ymax <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_rect</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">5.5</span>, xmax <span class="op">=</span> <span class="fl">14.5</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>, ymax <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"blue"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_rect</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">14.5</span>, xmax <span class="op">=</span> <span class="fl">19</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>, ymax <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"green"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_material_files/figure-html/perfPlot-recurrence-1.png" width="700"></p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; R Under development (unstable) (2023-11-22 r85609)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] lubridate_1.9.3                 forcats_1.0.0                  </span></span>
<span><span class="co">#&gt;  [3] stringr_1.5.1                   purrr_1.0.2                    </span></span>
<span><span class="co">#&gt;  [5] readr_2.1.4                     tibble_3.2.1                   </span></span>
<span><span class="co">#&gt;  [7] tidyverse_2.0.0                 ttservice_0.4.0                </span></span>
<span><span class="co">#&gt;  [9] tidyr_1.3.0                     dplyr_1.1.4                    </span></span>
<span><span class="co">#&gt; [11] tidySingleCellExperiment_1.13.1 scater_1.31.1                  </span></span>
<span><span class="co">#&gt; [13] scuttle_1.13.0                  reshape_0.8.9                  </span></span>
<span><span class="co">#&gt; [15] plotly_4.10.3                   glmnet_4.1-8                   </span></span>
<span><span class="co">#&gt; [17] Matrix_1.6-4                    ggsurvfit_1.0.0                </span></span>
<span><span class="co">#&gt; [19] ggplot2_3.4.4                   batchelor_1.19.0               </span></span>
<span><span class="co">#&gt; [21] cytomapper_1.15.0               EBImage_4.45.0                 </span></span>
<span><span class="co">#&gt; [23] Statial_1.5.5                   spicyR_1.15.3                  </span></span>
<span><span class="co">#&gt; [25] simpleSeg_1.5.0                 scHOT_1.15.0                   </span></span>
<span><span class="co">#&gt; [27] scFeatures_1.3.0                scClassify_1.15.0              </span></span>
<span><span class="co">#&gt; [29] lisaClust_1.11.1                FuseSOM_1.5.0                  </span></span>
<span><span class="co">#&gt; [31] ClassifyR_3.7.1                 survival_3.5-7                 </span></span>
<span><span class="co">#&gt; [33] BiocParallel_1.37.0             MultiAssayExperiment_1.29.0    </span></span>
<span><span class="co">#&gt; [35] generics_0.1.3                  SpatialDatasets_1.1.0          </span></span>
<span><span class="co">#&gt; [37] BumpyMatrix_1.11.0              STexampleData_1.11.0           </span></span>
<span><span class="co">#&gt; [39] SpatialExperiment_1.13.0        SingleCellExperiment_1.25.0    </span></span>
<span><span class="co">#&gt; [41] SummarizedExperiment_1.33.1     Biobase_2.63.0                 </span></span>
<span><span class="co">#&gt; [43] GenomicRanges_1.55.1            GenomeInfoDb_1.39.1            </span></span>
<span><span class="co">#&gt; [45] IRanges_2.37.0                  S4Vectors_0.41.2               </span></span>
<span><span class="co">#&gt; [47] MatrixGenerics_1.15.0           matrixStats_1.1.0              </span></span>
<span><span class="co">#&gt; [49] ExperimentHub_2.11.0            AnnotationHub_3.11.0           </span></span>
<span><span class="co">#&gt; [51] BiocFileCache_2.11.1            dbplyr_2.4.0                   </span></span>
<span><span class="co">#&gt; [53] BiocGenerics_0.49.1            </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] igraph_1.5.1                  graph_1.81.0                 </span></span>
<span><span class="co">#&gt;   [3] ica_1.0-3                     fpc_2.2-10                   </span></span>
<span><span class="co">#&gt;   [5] zlibbioc_1.49.0               tidyselect_1.2.0             </span></span>
<span><span class="co">#&gt;   [7] bit_4.0.5                     lattice_0.22-5               </span></span>
<span><span class="co">#&gt;   [9] rjson_0.2.21                  blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [11] S4Arrays_1.3.1                parallel_4.4.0               </span></span>
<span><span class="co">#&gt;  [13] png_0.1-8                     ResidualMatrix_1.13.0        </span></span>
<span><span class="co">#&gt;  [15] cli_3.6.1                     ggplotify_0.1.2              </span></span>
<span><span class="co">#&gt;  [17] ProtGenerics_1.35.0           multtest_2.59.0              </span></span>
<span><span class="co">#&gt;  [19] goftest_1.2-3                 pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt;  [21] textshaping_0.3.7             BiocIO_1.13.0                </span></span>
<span><span class="co">#&gt;  [23] kernlab_0.9-32                bluster_1.13.0               </span></span>
<span><span class="co">#&gt;  [25] BiocNeighbors_1.21.1          uwot_0.1.16                  </span></span>
<span><span class="co">#&gt;  [27] curl_5.1.0                    mime_0.12                    </span></span>
<span><span class="co">#&gt;  [29] evaluate_0.23                 tiff_0.1-12                  </span></span>
<span><span class="co">#&gt;  [31] leiden_0.4.3.1                stringi_1.8.2                </span></span>
<span><span class="co">#&gt;  [33] backports_1.4.1               desc_1.4.2                   </span></span>
<span><span class="co">#&gt;  [35] FCPS_1.3.4                    lmerTest_3.1-3               </span></span>
<span><span class="co">#&gt;  [37] XML_3.99-0.16                 httpuv_1.6.12                </span></span>
<span><span class="co">#&gt;  [39] flexmix_2.3-19                AnnotationDbi_1.65.2         </span></span>
<span><span class="co">#&gt;  [41] magrittr_2.0.3                DataVisualizations_1.3.2     </span></span>
<span><span class="co">#&gt;  [43] rappdirs_0.3.3                splines_4.4.0                </span></span>
<span><span class="co">#&gt;  [45] mclust_6.0.1                  jpeg_0.1-10                  </span></span>
<span><span class="co">#&gt;  [47] ggraph_2.1.0                  ggbeeswarm_0.7.2             </span></span>
<span><span class="co">#&gt;  [49] sctransform_0.4.1             terra_1.7-55                 </span></span>
<span><span class="co">#&gt;  [51] DBI_1.1.3                     HDF5Array_1.31.0             </span></span>
<span><span class="co">#&gt;  [53] genefilter_1.85.0             jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt;  [55] withr_2.5.2                   class_7.3-22                 </span></span>
<span><span class="co">#&gt;  [57] systemfonts_1.0.5             rprojroot_2.0.4              </span></span>
<span><span class="co">#&gt;  [59] lmtest_0.9-40                 GSEABase_1.65.0              </span></span>
<span><span class="co">#&gt;  [61] tidygraph_1.2.3               rtracklayer_1.63.0           </span></span>
<span><span class="co">#&gt;  [63] BiocManager_1.30.22           htmlwidgets_1.6.3            </span></span>
<span><span class="co">#&gt;  [65] fs_1.6.3                      biomaRt_2.59.0               </span></span>
<span><span class="co">#&gt;  [67] segmented_2.0-0               ggrepel_0.9.4                </span></span>
<span><span class="co">#&gt;  [69] princurve_2.1.6               labeling_0.4.3               </span></span>
<span><span class="co">#&gt;  [71] Cepo_1.9.0                    SparseArray_1.3.1            </span></span>
<span><span class="co">#&gt;  [73] DEoptimR_1.1-3                ranger_0.16.0                </span></span>
<span><span class="co">#&gt;  [75] SingleCellSignalR_1.15.0      mixtools_2.0.0               </span></span>
<span><span class="co">#&gt;  [77] diptest_0.77-0                annotate_1.81.0              </span></span>
<span><span class="co">#&gt;  [79] reticulate_1.34.0             minpack.lm_1.2-4             </span></span>
<span><span class="co">#&gt;  [81] zoo_1.8-12                    raster_3.6-26                </span></span>
<span><span class="co">#&gt;  [83] XVector_0.43.0                knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [85] nnls_1.5                      AUCell_1.25.0                </span></span>
<span><span class="co">#&gt;  [87] timechange_0.2.0              foreach_1.5.2                </span></span>
<span><span class="co">#&gt;  [89] fansi_1.0.5                   patchwork_1.1.3              </span></span>
<span><span class="co">#&gt;  [91] caTools_1.18.2                grid_4.4.0                   </span></span>
<span><span class="co">#&gt;  [93] data.table_1.14.8             rhdf5_2.47.1                 </span></span>
<span><span class="co">#&gt;  [95] vegan_2.6-4                   R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;  [97] psych_2.3.9                   RSpectra_0.16-1              </span></span>
<span><span class="co">#&gt;  [99] irlba_2.3.5.1                 fastDummies_1.7.3            </span></span>
<span><span class="co">#&gt; [101] gridGraphics_0.5-1            ellipsis_0.3.2               </span></span>
<span><span class="co">#&gt; [103] lazyeval_0.2.2                yaml_2.3.7                   </span></span>
<span><span class="co">#&gt; [105] scattermore_1.2               BiocVersion_3.19.1           </span></span>
<span><span class="co">#&gt; [107] crayon_1.5.2                  RcppAnnoy_0.0.21             </span></span>
<span><span class="co">#&gt; [109] RColorBrewer_1.1-3            progressr_0.14.0             </span></span>
<span><span class="co">#&gt; [111] tweenr_2.0.2                  later_1.3.1                  </span></span>
<span><span class="co">#&gt; [113] ggridges_0.5.4                codetools_0.2-19             </span></span>
<span><span class="co">#&gt; [115] GlobalOptions_0.1.2           Seurat_5.0.1                 </span></span>
<span><span class="co">#&gt; [117] KEGGREST_1.43.0               Rtsne_0.16                   </span></span>
<span><span class="co">#&gt; [119] shape_1.4.6                   limma_3.59.1                 </span></span>
<span><span class="co">#&gt; [121] Rsamtools_2.19.2              filelock_1.0.2               </span></span>
<span><span class="co">#&gt; [123] pkgconfig_2.0.3               xml2_1.3.5                   </span></span>
<span><span class="co">#&gt; [125] ggpubr_0.6.0                  GenomicAlignments_1.39.0     </span></span>
<span><span class="co">#&gt; [127] spatstat.sparse_3.0-3         ape_5.7-1                    </span></span>
<span><span class="co">#&gt; [129] viridisLite_0.4.2             xtable_1.8-4                 </span></span>
<span><span class="co">#&gt; [131] coop_0.6-3                    fastcluster_1.2.3            </span></span>
<span><span class="co">#&gt; [133] highr_0.10                    car_3.1-2                    </span></span>
<span><span class="co">#&gt; [135] plyr_1.8.9                    httr_1.4.7                   </span></span>
<span><span class="co">#&gt; [137] prabclus_2.3-3                tools_4.4.0                  </span></span>
<span><span class="co">#&gt; [139] globals_0.16.2                SeuratObject_5.0.1           </span></span>
<span><span class="co">#&gt; [141] beeswarm_0.4.0                broom_1.0.5                  </span></span>
<span><span class="co">#&gt; [143] nlme_3.1-164                  MatrixModels_0.5-3           </span></span>
<span><span class="co">#&gt; [145] crosstalk_1.2.1               lme4_1.1-35.1                </span></span>
<span><span class="co">#&gt; [147] digest_0.6.33                 permute_0.9-7                </span></span>
<span><span class="co">#&gt; [149] numDeriv_2016.8-1.1           tzdb_0.4.0                   </span></span>
<span><span class="co">#&gt; [151] farver_2.1.1                  AnnotationFilter_1.27.0      </span></span>
<span><span class="co">#&gt; [153] reshape2_1.4.4                yulab.utils_0.1.0            </span></span>
<span><span class="co">#&gt; [155] viridis_0.6.4                 glue_1.6.2                   </span></span>
<span><span class="co">#&gt; [157] cachem_1.0.8                  hopach_2.63.0                </span></span>
<span><span class="co">#&gt; [159] polyclip_1.10-6               proxyC_0.3.4                 </span></span>
<span><span class="co">#&gt; [161] Biostrings_2.71.1             profileModel_0.6.1           </span></span>
<span><span class="co">#&gt; [163] parallelly_1.36.0             mnormt_2.1.1                 </span></span>
<span><span class="co">#&gt; [165] statmod_1.5.0                 concaveman_1.1.0             </span></span>
<span><span class="co">#&gt; [167] RcppHNSW_0.5.0                ragg_1.2.6                   </span></span>
<span><span class="co">#&gt; [169] ScaledMatrix_1.11.0           carData_3.0-5                </span></span>
<span><span class="co">#&gt; [171] minqa_1.2.6                   pbapply_1.7-2                </span></span>
<span><span class="co">#&gt; [173] spam_2.10-0                   dqrng_0.3.2                  </span></span>
<span><span class="co">#&gt; [175] utf8_1.2.4                    graphlayouts_1.0.2           </span></span>
<span><span class="co">#&gt; [177] gtools_3.9.5                  analogue_0.17-6              </span></span>
<span><span class="co">#&gt; [179] ggsignif_0.6.4                gridExtra_2.3                </span></span>
<span><span class="co">#&gt; [181] shiny_1.8.0                   GSVA_1.51.0                  </span></span>
<span><span class="co">#&gt; [183] GenomeInfoDbData_1.2.11       R.utils_2.12.3               </span></span>
<span><span class="co">#&gt; [185] rhdf5filters_1.15.1           RCurl_1.98-1.13              </span></span>
<span><span class="co">#&gt; [187] memoise_2.0.1                 rmarkdown_2.25               </span></span>
<span><span class="co">#&gt; [189] pheatmap_1.0.12               scales_1.3.0                 </span></span>
<span><span class="co">#&gt; [191] R.methodsS3_1.8.2             svglite_2.1.2                </span></span>
<span><span class="co">#&gt; [193] future_1.33.0                 RANN_2.6.1                   </span></span>
<span><span class="co">#&gt; [195] spatstat.data_3.0-3           EnsDb.Mmusculus.v79_2.99.0   </span></span>
<span><span class="co">#&gt; [197] cluster_2.1.6                 msigdbr_7.5.1                </span></span>
<span><span class="co">#&gt; [199] spatstat.utils_3.0-4          hms_1.1.3                    </span></span>
<span><span class="co">#&gt; [201] fitdistrplus_1.1-11           munsell_0.5.0                </span></span>
<span><span class="co">#&gt; [203] cowplot_1.1.1                 scam_1.2-14                  </span></span>
<span><span class="co">#&gt; [205] colorspace_2.1-0              rlang_1.1.2                  </span></span>
<span><span class="co">#&gt; [207] EnsDb.Hsapiens.v79_2.99.0     DelayedMatrixStats_1.25.1    </span></span>
<span><span class="co">#&gt; [209] sparseMatrixStats_1.15.0      shinydashboard_0.7.2         </span></span>
<span><span class="co">#&gt; [211] dotCall64_1.1-1               ggforce_0.4.1                </span></span>
<span><span class="co">#&gt; [213] circlize_0.4.15               mgcv_1.9-0                   </span></span>
<span><span class="co">#&gt; [215] xfun_0.41                     modeltools_0.2-23            </span></span>
<span><span class="co">#&gt; [217] iterators_1.0.14              abind_1.4-5                  </span></span>
<span><span class="co">#&gt; [219] interactiveDisplayBase_1.41.0 Rhdf5lib_1.25.0              </span></span>
<span><span class="co">#&gt; [221] fftwtools_0.9-11              bitops_1.0-7                 </span></span>
<span><span class="co">#&gt; [223] promises_1.2.1                RSQLite_2.3.3                </span></span>
<span><span class="co">#&gt; [225] DelayedArray_0.29.0           proxy_0.4-27                 </span></span>
<span><span class="co">#&gt; [227] compiler_4.4.0                prettyunits_1.2.0            </span></span>
<span><span class="co">#&gt; [229] boot_1.3-28.1                 beachmat_2.19.0              </span></span>
<span><span class="co">#&gt; [231] listenv_0.9.0                 Rcpp_1.0.11                  </span></span>
<span><span class="co">#&gt; [233] edgeR_4.1.2                   BiocSingular_1.19.0          </span></span>
<span><span class="co">#&gt; [235] tensor_1.5                    MASS_7.3-60.1                </span></span>
<span><span class="co">#&gt; [237] progress_1.2.2                ggupset_0.3.0                </span></span>
<span><span class="co">#&gt; [239] babelgene_22.9                spatstat.random_3.2-2        </span></span>
<span><span class="co">#&gt; [241] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">#&gt; [243] rstatix_0.7.2                 vipor_0.4.5                  </span></span>
<span><span class="co">#&gt; [245] ensembldb_2.27.1              ROCR_1.0-11                  </span></span>
<span><span class="co">#&gt; [247] nnet_7.3-19                   rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt; [249] gtable_0.3.4                  KernSmooth_2.23-22           </span></span>
<span><span class="co">#&gt; [251] miniUI_0.1.1.1                deldir_2.0-2                 </span></span>
<span><span class="co">#&gt; [253] htmltools_0.5.7               RcppParallel_5.1.7           </span></span>
<span><span class="co">#&gt; [255] bit64_4.0.5                   spatstat.explore_3.2-5       </span></span>
<span><span class="co">#&gt; [257] lifecycle_1.0.4               nloptr_2.0.3                 </span></span>
<span><span class="co">#&gt; [259] restfulr_0.0.15               sass_0.4.7                   </span></span>
<span><span class="co">#&gt; [261] vctrs_0.6.5                   robustbase_0.99-1            </span></span>
<span><span class="co">#&gt; [263] spatstat.geom_3.2-7           scran_1.31.0                 </span></span>
<span><span class="co">#&gt; [265] sp_2.1-2                      future.apply_1.11.0          </span></span>
<span><span class="co">#&gt; [267] bslib_0.6.1                   pillar_1.9.0                 </span></span>
<span><span class="co">#&gt; [269] GenomicFeatures_1.55.1        gplots_3.1.3                 </span></span>
<span><span class="co">#&gt; [271] magick_2.8.1                  metapod_1.11.0               </span></span>
<span><span class="co">#&gt; [273] locfit_1.5-9.8                svgPanZoom_0.3.4             </span></span>
<span><span class="co">#&gt; [275] jsonlite_1.8.7                brglm_0.7.2</span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Farhan Ameen, Alex Qin, Nick Robertson, Ellis Patrick, Shila Ghazanfar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
